{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Data Preprocessing On Gene Expression Dataset**\n",
    "\n",
    "#### **First Step**\n",
    "*First, load the gene expression data that already process in R. \n",
    "The gene expression data should have a column of gene names and 34 columns of the samples. <br>\n",
    "We expected that there will be missing value and data as well as the duplicated gene names will be appear in the datasets.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <th>GSM509792_E1542N.CEL</th>\n",
       "      <th>GSM509793_E1546N.CEL</th>\n",
       "      <th>GSM509794_E1566N.CEL</th>\n",
       "      <th>GSM509795_E1584N.CEL</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM509811_E1566T.CEL</th>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <th>GSM509813_E1589T.CEL</th>\n",
       "      <th>GSM509814_E1603T.CEL</th>\n",
       "      <th>GSM509815_E1610T.CEL</th>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <th>GSM509817_E1635T.CEL</th>\n",
       "      <th>GSM509818_E1709T.CEL</th>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <th>GSM509820_E2644T.CEL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22272</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5.787397</td>\n",
       "      <td>5.913325</td>\n",
       "      <td>4.450484</td>\n",
       "      <td>5.808386</td>\n",
       "      <td>5.003064</td>\n",
       "      <td>5.974022</td>\n",
       "      <td>4.676522</td>\n",
       "      <td>4.530070</td>\n",
       "      <td>4.892546</td>\n",
       "      <td>...</td>\n",
       "      <td>5.190195</td>\n",
       "      <td>6.952222</td>\n",
       "      <td>5.235334</td>\n",
       "      <td>4.417394</td>\n",
       "      <td>6.691355</td>\n",
       "      <td>5.410509</td>\n",
       "      <td>5.430227</td>\n",
       "      <td>4.272877</td>\n",
       "      <td>4.865624</td>\n",
       "      <td>5.347518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22273</th>\n",
       "      <td>NaN</td>\n",
       "      <td>7.330281</td>\n",
       "      <td>7.202484</td>\n",
       "      <td>4.830335</td>\n",
       "      <td>7.121127</td>\n",
       "      <td>6.077353</td>\n",
       "      <td>7.565568</td>\n",
       "      <td>4.397282</td>\n",
       "      <td>4.689992</td>\n",
       "      <td>5.481145</td>\n",
       "      <td>...</td>\n",
       "      <td>6.372190</td>\n",
       "      <td>8.579120</td>\n",
       "      <td>6.653998</td>\n",
       "      <td>5.090072</td>\n",
       "      <td>8.399972</td>\n",
       "      <td>7.027365</td>\n",
       "      <td>6.948558</td>\n",
       "      <td>4.511476</td>\n",
       "      <td>6.141248</td>\n",
       "      <td>6.940607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22274</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.363339</td>\n",
       "      <td>3.409699</td>\n",
       "      <td>3.294732</td>\n",
       "      <td>3.463503</td>\n",
       "      <td>3.517765</td>\n",
       "      <td>3.371528</td>\n",
       "      <td>3.423060</td>\n",
       "      <td>3.417975</td>\n",
       "      <td>3.423598</td>\n",
       "      <td>...</td>\n",
       "      <td>3.372451</td>\n",
       "      <td>3.273974</td>\n",
       "      <td>3.425394</td>\n",
       "      <td>3.437641</td>\n",
       "      <td>3.488736</td>\n",
       "      <td>3.349208</td>\n",
       "      <td>3.294963</td>\n",
       "      <td>3.414566</td>\n",
       "      <td>3.337863</td>\n",
       "      <td>3.387840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22275</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.768794</td>\n",
       "      <td>3.853740</td>\n",
       "      <td>3.716894</td>\n",
       "      <td>3.802079</td>\n",
       "      <td>3.651687</td>\n",
       "      <td>3.800569</td>\n",
       "      <td>3.727047</td>\n",
       "      <td>3.767205</td>\n",
       "      <td>3.599901</td>\n",
       "      <td>...</td>\n",
       "      <td>3.536484</td>\n",
       "      <td>3.661301</td>\n",
       "      <td>3.743973</td>\n",
       "      <td>3.561133</td>\n",
       "      <td>3.698957</td>\n",
       "      <td>3.623726</td>\n",
       "      <td>3.601925</td>\n",
       "      <td>3.499687</td>\n",
       "      <td>3.598900</td>\n",
       "      <td>3.453938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22276</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.479989</td>\n",
       "      <td>3.491986</td>\n",
       "      <td>3.473315</td>\n",
       "      <td>3.471217</td>\n",
       "      <td>3.564663</td>\n",
       "      <td>3.463013</td>\n",
       "      <td>3.448192</td>\n",
       "      <td>3.398924</td>\n",
       "      <td>3.617687</td>\n",
       "      <td>...</td>\n",
       "      <td>3.374688</td>\n",
       "      <td>3.411003</td>\n",
       "      <td>3.379661</td>\n",
       "      <td>3.478230</td>\n",
       "      <td>3.430541</td>\n",
       "      <td>3.453279</td>\n",
       "      <td>3.339018</td>\n",
       "      <td>3.465621</td>\n",
       "      <td>3.411268</td>\n",
       "      <td>3.407996</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene Symbol  GSM509787_E1507N.CEL  GSM509788_E1520N.CEL  \\\n",
       "22272         NaN              5.787397              5.913325   \n",
       "22273         NaN              7.330281              7.202484   \n",
       "22274         NaN              3.363339              3.409699   \n",
       "22275         NaN              3.768794              3.853740   \n",
       "22276         NaN              3.479989              3.491986   \n",
       "\n",
       "       GSM509789_E1521N.CEL  GSM509790_E1532N.CEL  GSM509791_E1535N.CEL  \\\n",
       "22272              4.450484              5.808386              5.003064   \n",
       "22273              4.830335              7.121127              6.077353   \n",
       "22274              3.294732              3.463503              3.517765   \n",
       "22275              3.716894              3.802079              3.651687   \n",
       "22276              3.473315              3.471217              3.564663   \n",
       "\n",
       "       GSM509792_E1542N.CEL  GSM509793_E1546N.CEL  GSM509794_E1566N.CEL  \\\n",
       "22272              5.974022              4.676522              4.530070   \n",
       "22273              7.565568              4.397282              4.689992   \n",
       "22274              3.371528              3.423060              3.417975   \n",
       "22275              3.800569              3.727047              3.767205   \n",
       "22276              3.463013              3.448192              3.398924   \n",
       "\n",
       "       GSM509795_E1584N.CEL  ...  GSM509811_E1566T.CEL  GSM509812_E1584T.CEL  \\\n",
       "22272              4.892546  ...              5.190195              6.952222   \n",
       "22273              5.481145  ...              6.372190              8.579120   \n",
       "22274              3.423598  ...              3.372451              3.273974   \n",
       "22275              3.599901  ...              3.536484              3.661301   \n",
       "22276              3.617687  ...              3.374688              3.411003   \n",
       "\n",
       "       GSM509813_E1589T.CEL  GSM509814_E1603T.CEL  GSM509815_E1610T.CEL  \\\n",
       "22272              5.235334              4.417394              6.691355   \n",
       "22273              6.653998              5.090072              8.399972   \n",
       "22274              3.425394              3.437641              3.488736   \n",
       "22275              3.743973              3.561133              3.698957   \n",
       "22276              3.379661              3.478230              3.430541   \n",
       "\n",
       "       GSM509816_E1614T.CEL  GSM509817_E1635T.CEL  GSM509818_E1709T.CEL  \\\n",
       "22272              5.410509              5.430227              4.272877   \n",
       "22273              7.027365              6.948558              4.511476   \n",
       "22274              3.349208              3.294963              3.414566   \n",
       "22275              3.623726              3.601925              3.499687   \n",
       "22276              3.453279              3.339018              3.465621   \n",
       "\n",
       "       GSM509819_E1796T.CEL  GSM509820_E2644T.CEL  \n",
       "22272              4.865624              5.347518  \n",
       "22273              6.141248              6.940607  \n",
       "22274              3.337863              3.387840  \n",
       "22275              3.598900              3.453938  \n",
       "22276              3.411268              3.407996  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the Gene Expression Datasets\n",
    "gene_expression_data = pd.read_csv(r\"D:\\UTM\\20222023 SEM 2\\PSM I\\FYP\\dataset\\GeneExpression.csv\")\n",
    "\n",
    "# Display the Last Five Rows of the Gene Expression Dataset\n",
    "gene_expression_data.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Second Step**\n",
    "*A few processes should be done to remove all the missing value and data. Then get the average value of the duplicated genes. <br>\n",
    "The dataset will be used to check whether it still contain the missing value and data as well as the duplicated genes.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Missing Data:  1052\n"
     ]
    }
   ],
   "source": [
    "# Check for the Number of the NaN Gene Symbol\n",
    "missing_data = pd.isna(gene_expression_data[\"Gene Symbol\"]).sum()\n",
    "print(\"The Missing Data: \", missing_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene Symbol              True\n",
       "GSM509787_E1507N.CEL    False\n",
       "GSM509788_E1520N.CEL    False\n",
       "GSM509789_E1521N.CEL    False\n",
       "GSM509790_E1532N.CEL    False\n",
       "GSM509791_E1535N.CEL    False\n",
       "GSM509792_E1542N.CEL    False\n",
       "GSM509793_E1546N.CEL    False\n",
       "GSM509794_E1566N.CEL    False\n",
       "GSM509795_E1584N.CEL    False\n",
       "GSM509796_E1589N.CEL    False\n",
       "GSM509797_E1603N.CEL    False\n",
       "GSM509798_E1610N.CEL    False\n",
       "GSM509799_E1614N.CEL    False\n",
       "GSM509800_E1635N.CEL    False\n",
       "GSM509801_E1709N.CEL    False\n",
       "GSM509802_E1796N.CEL    False\n",
       "GSM509803_E2644N.CEL    False\n",
       "GSM509804_E1507T.CEL    False\n",
       "GSM509805_E1520T.CEL    False\n",
       "GSM509806_E1521T.CEL    False\n",
       "GSM509807_E1532T.CEL    False\n",
       "GSM509808_E1535T.CEL    False\n",
       "GSM509809_E1542T.CEL    False\n",
       "GSM509810_E1546T.CEL    False\n",
       "GSM509811_E1566T.CEL    False\n",
       "GSM509812_E1584T.CEL    False\n",
       "GSM509813_E1589T.CEL    False\n",
       "GSM509814_E1603T.CEL    False\n",
       "GSM509815_E1610T.CEL    False\n",
       "GSM509816_E1614T.CEL    False\n",
       "GSM509817_E1635T.CEL    False\n",
       "GSM509818_E1709T.CEL    False\n",
       "GSM509819_E1796T.CEL    False\n",
       "GSM509820_E2644T.CEL    False\n",
       "dtype: bool"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check is Have Unnamed Columns Exist in Gene Expression Dataset\n",
    "gene_expression_data.isnull().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <th>GSM509792_E1542N.CEL</th>\n",
       "      <th>GSM509793_E1546N.CEL</th>\n",
       "      <th>GSM509794_E1566N.CEL</th>\n",
       "      <th>GSM509795_E1584N.CEL</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM509811_E1566T.CEL</th>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <th>GSM509813_E1589T.CEL</th>\n",
       "      <th>GSM509814_E1603T.CEL</th>\n",
       "      <th>GSM509815_E1610T.CEL</th>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <th>GSM509817_E1635T.CEL</th>\n",
       "      <th>GSM509818_E1709T.CEL</th>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <th>GSM509820_E2644T.CEL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22233</th>\n",
       "      <td>GAPDH</td>\n",
       "      <td>12.949766</td>\n",
       "      <td>12.943392</td>\n",
       "      <td>12.889411</td>\n",
       "      <td>12.898720</td>\n",
       "      <td>12.645978</td>\n",
       "      <td>12.496930</td>\n",
       "      <td>12.840867</td>\n",
       "      <td>12.770401</td>\n",
       "      <td>12.958057</td>\n",
       "      <td>...</td>\n",
       "      <td>12.802725</td>\n",
       "      <td>12.203343</td>\n",
       "      <td>13.712690</td>\n",
       "      <td>12.848983</td>\n",
       "      <td>12.757335</td>\n",
       "      <td>13.190879</td>\n",
       "      <td>13.354656</td>\n",
       "      <td>12.845014</td>\n",
       "      <td>12.964635</td>\n",
       "      <td>13.280742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22234</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>9.484552</td>\n",
       "      <td>9.478317</td>\n",
       "      <td>8.964032</td>\n",
       "      <td>9.931751</td>\n",
       "      <td>8.680998</td>\n",
       "      <td>10.007735</td>\n",
       "      <td>9.546031</td>\n",
       "      <td>9.163568</td>\n",
       "      <td>9.328506</td>\n",
       "      <td>...</td>\n",
       "      <td>10.150800</td>\n",
       "      <td>10.795962</td>\n",
       "      <td>11.478691</td>\n",
       "      <td>9.972291</td>\n",
       "      <td>9.834421</td>\n",
       "      <td>10.964052</td>\n",
       "      <td>9.125182</td>\n",
       "      <td>8.942139</td>\n",
       "      <td>11.721256</td>\n",
       "      <td>10.531963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22235</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>3.628048</td>\n",
       "      <td>3.574983</td>\n",
       "      <td>3.608933</td>\n",
       "      <td>3.636437</td>\n",
       "      <td>3.695790</td>\n",
       "      <td>3.550540</td>\n",
       "      <td>3.676861</td>\n",
       "      <td>3.611349</td>\n",
       "      <td>3.611434</td>\n",
       "      <td>...</td>\n",
       "      <td>3.699795</td>\n",
       "      <td>3.944819</td>\n",
       "      <td>4.595343</td>\n",
       "      <td>3.704882</td>\n",
       "      <td>3.753756</td>\n",
       "      <td>3.909653</td>\n",
       "      <td>3.639074</td>\n",
       "      <td>3.719383</td>\n",
       "      <td>4.276364</td>\n",
       "      <td>3.912584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22236</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>6.333498</td>\n",
       "      <td>6.353891</td>\n",
       "      <td>6.289165</td>\n",
       "      <td>6.305032</td>\n",
       "      <td>6.033114</td>\n",
       "      <td>6.336885</td>\n",
       "      <td>6.455834</td>\n",
       "      <td>6.040308</td>\n",
       "      <td>5.855748</td>\n",
       "      <td>...</td>\n",
       "      <td>6.250628</td>\n",
       "      <td>7.494119</td>\n",
       "      <td>8.187525</td>\n",
       "      <td>6.186974</td>\n",
       "      <td>6.591403</td>\n",
       "      <td>6.835004</td>\n",
       "      <td>6.004906</td>\n",
       "      <td>5.970246</td>\n",
       "      <td>8.576720</td>\n",
       "      <td>6.481528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22237</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>5.004554</td>\n",
       "      <td>4.701656</td>\n",
       "      <td>4.599693</td>\n",
       "      <td>4.796084</td>\n",
       "      <td>4.220579</td>\n",
       "      <td>4.945150</td>\n",
       "      <td>4.793822</td>\n",
       "      <td>4.641824</td>\n",
       "      <td>4.490249</td>\n",
       "      <td>...</td>\n",
       "      <td>5.019220</td>\n",
       "      <td>5.600496</td>\n",
       "      <td>7.534357</td>\n",
       "      <td>4.637498</td>\n",
       "      <td>5.647896</td>\n",
       "      <td>6.348953</td>\n",
       "      <td>5.161385</td>\n",
       "      <td>4.158484</td>\n",
       "      <td>7.867963</td>\n",
       "      <td>5.452178</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene Symbol  GSM509787_E1507N.CEL  GSM509788_E1520N.CEL  \\\n",
       "22233       GAPDH             12.949766             12.943392   \n",
       "22234       STAT1              9.484552              9.478317   \n",
       "22235       STAT1              3.628048              3.574983   \n",
       "22236       STAT1              6.333498              6.353891   \n",
       "22237       STAT1              5.004554              4.701656   \n",
       "\n",
       "       GSM509789_E1521N.CEL  GSM509790_E1532N.CEL  GSM509791_E1535N.CEL  \\\n",
       "22233             12.889411             12.898720             12.645978   \n",
       "22234              8.964032              9.931751              8.680998   \n",
       "22235              3.608933              3.636437              3.695790   \n",
       "22236              6.289165              6.305032              6.033114   \n",
       "22237              4.599693              4.796084              4.220579   \n",
       "\n",
       "       GSM509792_E1542N.CEL  GSM509793_E1546N.CEL  GSM509794_E1566N.CEL  \\\n",
       "22233             12.496930             12.840867             12.770401   \n",
       "22234             10.007735              9.546031              9.163568   \n",
       "22235              3.550540              3.676861              3.611349   \n",
       "22236              6.336885              6.455834              6.040308   \n",
       "22237              4.945150              4.793822              4.641824   \n",
       "\n",
       "       GSM509795_E1584N.CEL  ...  GSM509811_E1566T.CEL  GSM509812_E1584T.CEL  \\\n",
       "22233             12.958057  ...             12.802725             12.203343   \n",
       "22234              9.328506  ...             10.150800             10.795962   \n",
       "22235              3.611434  ...              3.699795              3.944819   \n",
       "22236              5.855748  ...              6.250628              7.494119   \n",
       "22237              4.490249  ...              5.019220              5.600496   \n",
       "\n",
       "       GSM509813_E1589T.CEL  GSM509814_E1603T.CEL  GSM509815_E1610T.CEL  \\\n",
       "22233             13.712690             12.848983             12.757335   \n",
       "22234             11.478691              9.972291              9.834421   \n",
       "22235              4.595343              3.704882              3.753756   \n",
       "22236              8.187525              6.186974              6.591403   \n",
       "22237              7.534357              4.637498              5.647896   \n",
       "\n",
       "       GSM509816_E1614T.CEL  GSM509817_E1635T.CEL  GSM509818_E1709T.CEL  \\\n",
       "22233             13.190879             13.354656             12.845014   \n",
       "22234             10.964052              9.125182              8.942139   \n",
       "22235              3.909653              3.639074              3.719383   \n",
       "22236              6.835004              6.004906              5.970246   \n",
       "22237              6.348953              5.161385              4.158484   \n",
       "\n",
       "       GSM509819_E1796T.CEL  GSM509820_E2644T.CEL  \n",
       "22233             12.964635             13.280742  \n",
       "22234             11.721256             10.531963  \n",
       "22235              4.276364              3.912584  \n",
       "22236              8.576720              6.481528  \n",
       "22237              7.867963              5.452178  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean the Gene Expression Dataset by Dropping the NaN Value in Gene Symbol Column\n",
    "cleaned_gene_expression_data = gene_expression_data.dropna()\n",
    "\n",
    "# Display the Last Five Rows of the Gene Expression Dataset\n",
    "cleaned_gene_expression_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Missing Data:  0\n"
     ]
    }
   ],
   "source": [
    "# To Check If There Still Have Missing Data in Gene Symbol Column\n",
    "missing_data = pd.isna(cleaned_gene_expression_data[\"Gene Symbol\"]).sum()\n",
    "print(\"The Missing Data: \", missing_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The duplicated rows:  7710\n"
     ]
    }
   ],
   "source": [
    "# Check for Duplicated Gene Symbol that Appeared in Gene Expression Dataset\n",
    "duplicated_data = cleaned_gene_expression_data[\"Gene Symbol\"].duplicated().sum()\n",
    "print(\"The duplicated rows: \", duplicated_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <th>GSM509792_E1542N.CEL</th>\n",
       "      <th>GSM509793_E1546N.CEL</th>\n",
       "      <th>GSM509794_E1566N.CEL</th>\n",
       "      <th>GSM509795_E1584N.CEL</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM509811_E1566T.CEL</th>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <th>GSM509813_E1589T.CEL</th>\n",
       "      <th>GSM509814_E1603T.CEL</th>\n",
       "      <th>GSM509815_E1610T.CEL</th>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <th>GSM509817_E1635T.CEL</th>\n",
       "      <th>GSM509818_E1709T.CEL</th>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <th>GSM509820_E2644T.CEL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13510</th>\n",
       "      <td>FAM86B1 /// FAM86B2 /// FAM86C1 /// FAM86DP //...</td>\n",
       "      <td>6.425221</td>\n",
       "      <td>6.604427</td>\n",
       "      <td>6.774870</td>\n",
       "      <td>6.521508</td>\n",
       "      <td>6.163176</td>\n",
       "      <td>6.522643</td>\n",
       "      <td>6.697689</td>\n",
       "      <td>6.832607</td>\n",
       "      <td>6.679620</td>\n",
       "      <td>...</td>\n",
       "      <td>7.161260</td>\n",
       "      <td>6.893163</td>\n",
       "      <td>7.049957</td>\n",
       "      <td>6.222582</td>\n",
       "      <td>7.677644</td>\n",
       "      <td>7.233824</td>\n",
       "      <td>6.205083</td>\n",
       "      <td>7.096407</td>\n",
       "      <td>6.900650</td>\n",
       "      <td>9.378421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13511</th>\n",
       "      <td>SNHG17</td>\n",
       "      <td>6.760667</td>\n",
       "      <td>6.366031</td>\n",
       "      <td>6.876295</td>\n",
       "      <td>6.453466</td>\n",
       "      <td>6.326391</td>\n",
       "      <td>6.391639</td>\n",
       "      <td>6.869734</td>\n",
       "      <td>7.539441</td>\n",
       "      <td>6.858778</td>\n",
       "      <td>...</td>\n",
       "      <td>8.404900</td>\n",
       "      <td>7.084104</td>\n",
       "      <td>7.505765</td>\n",
       "      <td>6.891298</td>\n",
       "      <td>8.250209</td>\n",
       "      <td>6.868656</td>\n",
       "      <td>7.902891</td>\n",
       "      <td>7.263352</td>\n",
       "      <td>7.348683</td>\n",
       "      <td>8.216392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13512</th>\n",
       "      <td>HNRNPUL2 /// HNRNPUL2-BSCL2</td>\n",
       "      <td>3.977508</td>\n",
       "      <td>4.177136</td>\n",
       "      <td>3.862921</td>\n",
       "      <td>4.276351</td>\n",
       "      <td>4.258736</td>\n",
       "      <td>4.107978</td>\n",
       "      <td>3.989460</td>\n",
       "      <td>4.075416</td>\n",
       "      <td>3.958381</td>\n",
       "      <td>...</td>\n",
       "      <td>3.897055</td>\n",
       "      <td>4.117200</td>\n",
       "      <td>4.074250</td>\n",
       "      <td>3.883615</td>\n",
       "      <td>3.986138</td>\n",
       "      <td>3.867989</td>\n",
       "      <td>3.823485</td>\n",
       "      <td>4.095182</td>\n",
       "      <td>3.952813</td>\n",
       "      <td>4.051649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13513</th>\n",
       "      <td>LOC100505915</td>\n",
       "      <td>5.091434</td>\n",
       "      <td>4.993988</td>\n",
       "      <td>5.122977</td>\n",
       "      <td>4.851322</td>\n",
       "      <td>5.552102</td>\n",
       "      <td>4.997737</td>\n",
       "      <td>4.929513</td>\n",
       "      <td>4.773627</td>\n",
       "      <td>5.070867</td>\n",
       "      <td>...</td>\n",
       "      <td>4.952614</td>\n",
       "      <td>5.341808</td>\n",
       "      <td>4.977601</td>\n",
       "      <td>5.143281</td>\n",
       "      <td>5.116671</td>\n",
       "      <td>5.047892</td>\n",
       "      <td>5.027927</td>\n",
       "      <td>5.078444</td>\n",
       "      <td>5.137326</td>\n",
       "      <td>4.556811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13514</th>\n",
       "      <td>NPEPL1</td>\n",
       "      <td>7.231490</td>\n",
       "      <td>7.194917</td>\n",
       "      <td>7.138887</td>\n",
       "      <td>7.339033</td>\n",
       "      <td>7.223029</td>\n",
       "      <td>7.208323</td>\n",
       "      <td>7.143350</td>\n",
       "      <td>7.372182</td>\n",
       "      <td>7.386384</td>\n",
       "      <td>...</td>\n",
       "      <td>7.194305</td>\n",
       "      <td>7.230357</td>\n",
       "      <td>7.175572</td>\n",
       "      <td>7.287106</td>\n",
       "      <td>6.860418</td>\n",
       "      <td>6.692204</td>\n",
       "      <td>7.095290</td>\n",
       "      <td>7.731047</td>\n",
       "      <td>6.983723</td>\n",
       "      <td>7.025979</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Gene Symbol  \\\n",
       "13510  FAM86B1 /// FAM86B2 /// FAM86C1 /// FAM86DP //...   \n",
       "13511                                             SNHG17   \n",
       "13512                        HNRNPUL2 /// HNRNPUL2-BSCL2   \n",
       "13513                                       LOC100505915   \n",
       "13514                                             NPEPL1   \n",
       "\n",
       "       GSM509787_E1507N.CEL  GSM509788_E1520N.CEL  GSM509789_E1521N.CEL  \\\n",
       "13510              6.425221              6.604427              6.774870   \n",
       "13511              6.760667              6.366031              6.876295   \n",
       "13512              3.977508              4.177136              3.862921   \n",
       "13513              5.091434              4.993988              5.122977   \n",
       "13514              7.231490              7.194917              7.138887   \n",
       "\n",
       "       GSM509790_E1532N.CEL  GSM509791_E1535N.CEL  GSM509792_E1542N.CEL  \\\n",
       "13510              6.521508              6.163176              6.522643   \n",
       "13511              6.453466              6.326391              6.391639   \n",
       "13512              4.276351              4.258736              4.107978   \n",
       "13513              4.851322              5.552102              4.997737   \n",
       "13514              7.339033              7.223029              7.208323   \n",
       "\n",
       "       GSM509793_E1546N.CEL  GSM509794_E1566N.CEL  GSM509795_E1584N.CEL  ...  \\\n",
       "13510              6.697689              6.832607              6.679620  ...   \n",
       "13511              6.869734              7.539441              6.858778  ...   \n",
       "13512              3.989460              4.075416              3.958381  ...   \n",
       "13513              4.929513              4.773627              5.070867  ...   \n",
       "13514              7.143350              7.372182              7.386384  ...   \n",
       "\n",
       "       GSM509811_E1566T.CEL  GSM509812_E1584T.CEL  GSM509813_E1589T.CEL  \\\n",
       "13510              7.161260              6.893163              7.049957   \n",
       "13511              8.404900              7.084104              7.505765   \n",
       "13512              3.897055              4.117200              4.074250   \n",
       "13513              4.952614              5.341808              4.977601   \n",
       "13514              7.194305              7.230357              7.175572   \n",
       "\n",
       "       GSM509814_E1603T.CEL  GSM509815_E1610T.CEL  GSM509816_E1614T.CEL  \\\n",
       "13510              6.222582              7.677644              7.233824   \n",
       "13511              6.891298              8.250209              6.868656   \n",
       "13512              3.883615              3.986138              3.867989   \n",
       "13513              5.143281              5.116671              5.047892   \n",
       "13514              7.287106              6.860418              6.692204   \n",
       "\n",
       "       GSM509817_E1635T.CEL  GSM509818_E1709T.CEL  GSM509819_E1796T.CEL  \\\n",
       "13510              6.205083              7.096407              6.900650   \n",
       "13511              7.902891              7.263352              7.348683   \n",
       "13512              3.823485              4.095182              3.952813   \n",
       "13513              5.027927              5.078444              5.137326   \n",
       "13514              7.095290              7.731047              6.983723   \n",
       "\n",
       "       GSM509820_E2644T.CEL  \n",
       "13510              9.378421  \n",
       "13511              8.216392  \n",
       "13512              4.051649  \n",
       "13513              4.556811  \n",
       "13514              7.025979  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Average the Data of the Duplicated Gene Symbol in Gene Expression Dataset\n",
    "average_gene_expression_data = cleaned_gene_expression_data.groupby([\"Gene Symbol\"], sort=False).mean().reset_index()\n",
    "\n",
    "# Display the Last Five Rows of the Gene Expression Dataset\n",
    "average_gene_expression_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The duplicated rows:  0\n",
      "The dimension of gene expression data:  (22277, 35)\n",
      "The dimension of gene expression data after clear the missing data and get the average value of duplicate rows:  (13515, 35)\n"
     ]
    }
   ],
   "source": [
    "# Check Whether It is Still Have Duplicate Gene Symbol in Gene Expression Dataset\n",
    "duplicated_data = average_gene_expression_data[\"Gene Symbol\"].duplicated().sum()\n",
    "print(\"The duplicated rows: \", duplicated_data)\n",
    "\n",
    "# Check the Dimension of the Gene Expression Dataset\n",
    "print(\"The dimension of gene expression data: \" , gene_expression_data.shape)\n",
    "print(\"The dimension of gene expression data after clear the missing data and get the average value of duplicate rows: \" , average_gene_expression_data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Data Preprocessing On Protein-Protein Interaction Data**\n",
    "\n",
    "#### **First Step**\n",
    "*First, load the protein-protein interaction data that had been obtained from STRING website.<br>\n",
    "<br>\n",
    "The data will contained information as below: <br>\n",
    "    1. node1 <br>\t\n",
    "    2. node2 <br>\n",
    "    3. node1_string_id <br>\t\n",
    "    4. node2_string_id <br>\n",
    "    5. neighborhood_on_chromosome <br>\t\n",
    "    6. gene_fusion <br>\n",
    "    7. phylogenetic_cooccurrence <br>\t\n",
    "    8. homology\tcoexpression <br>\t\n",
    "    9. experimentally_determined_interaction <br>\t\n",
    "    10. database_annotated <br>\t\n",
    "    11. automated_textmining <br>\t\n",
    "    12. combined_score  <br>\n",
    "<br>\n",
    "Only Node1, Node2 and Combined Score will be retrieved as Node1 and Node2 refer as Gene Symbol while Combined Score is weighted sum of individual evidence scores.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the PPI Datasets\n",
    "import pandas as pd\n",
    "\n",
    "ppi_data_reactome = pd.read_csv(r\"D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/PPI_Network/human_genes_STRING_interactions.tsv\", sep='\\t')\n",
    "ppi_data_monarch = pd.read_csv(r\"D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/PPI_Network/monarch_string_interactions.tsv\", sep='\\t')\n",
    "ppi_data_kegg = pd.read_csv(r\"D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/PPI_Network/kegg_string_interactions.tsv\", sep='\\t')\n",
    "ppi_data_disease = pd.read_csv(r\"D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/PPI_Network/disease_string_interactions.tsv\", sep='\\t')\n",
    "\n",
    "# Only Retrieved the Gene Symbols that are Connected With Each Other With the Combined Score\n",
    "# Since the combined_score = (w1 × evidence_score1) + (w2 × evidence_score2) + ... + (wn × evidence_scoren)\n",
    "# Hence Evaluate the PPI Using Combined Score\n",
    "ppi_data_reactome_new = ppi_data_reactome[['#node1','node2', 'combined_score']].copy()\n",
    "ppi_data_monarch_new = ppi_data_monarch[['#node1','node2', 'combined_score']].copy()\n",
    "ppi_data_kegg_new = ppi_data_kegg[['#node1','node2', 'combined_score']].copy()\n",
    "ppi_data_disease_new = ppi_data_disease[['#node1','node2', 'combined_score']].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\3410457905.py:2: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  preprocessed_PPI = ppi_data_reactome_new.append(ppi_data_monarch_new)\n",
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\3410457905.py:3: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  preprocessed_PPI = preprocessed_PPI.append(ppi_data_kegg_new)\n",
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\3410457905.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  preprocessed_PPI = preprocessed_PPI.append(ppi_data_disease_new)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>VIP</td>\n",
       "      <td>0.444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>MC2R</td>\n",
       "      <td>0.463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>LIG1</td>\n",
       "      <td>0.497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>POMC</td>\n",
       "      <td>0.566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>POM121</td>\n",
       "      <td>0.600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #node1   node2  combined_score\n",
       "0   AAAS     VIP           0.444\n",
       "1   AAAS    MC2R           0.463\n",
       "2   AAAS    LIG1           0.497\n",
       "3   AAAS    POMC           0.566\n",
       "4   AAAS  POM121           0.600"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combined All PPI Network Together\n",
    "preprocessed_PPI = ppi_data_reactome_new.append(ppi_data_monarch_new)\n",
    "preprocessed_PPI = preprocessed_PPI.append(ppi_data_kegg_new)\n",
    "preprocessed_PPI = preprocessed_PPI.append(ppi_data_disease_new)\n",
    "\n",
    "# Display the First Five Rows of the PPI Network\n",
    "preprocessed_PPI.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Second Step**\n",
    "*Now, we will going to filter the genes from the gene expression data that are included in Protein-protein Interaction Network. <br>*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The genes that involve in the PPI network:  {'FMNL2', 'FXYD1', 'PARP10', 'SOX13', 'PDLIM4', 'HSPA1L', 'UBE2T', 'INSL3', 'RARB', 'NDUFC1', 'CSNK2B', 'SUZ12', 'FAT4', 'ANKRD23', 'AOX1', 'CEL', 'ALG3', 'MYD88', 'PSMB11', 'PSMB10', 'LCN9', 'UBE2S', 'VPS36', 'CPA2', 'UBB', 'LPHN2', 'NDUFS4', 'PLEKHA7', 'C1GALT1C1', 'VPS37D', 'ETFA', 'HSPA8', 'HSP90AA1', 'CACNB2', 'SNW1', 'KL', 'NDUFA8', 'EYA4', 'WDR35', 'ALOX5AP', 'CFAP57', 'MRAP', 'FZD8', 'TAAR2', 'CANX', 'MICB', 'C4orf22', 'COMT', 'ELK1', 'PTK2', 'MAST4', 'RPL5', 'RPS26', 'PLCG1', 'RPGR', 'CASP1', 'C21orf59-TCP10L', 'PSME2', 'WRAP53', 'TMTC1', 'PALB2', 'SDHC', 'GCLC', 'PVRL2', 'TAP2', 'RPL10A', 'AVPR1A', 'GABBR2', 'RPL10L', 'RPL36A', 'ERBB3', 'PCGF2', 'PABPN1', 'CHD7', 'PIEZO1', 'RHOG', 'TDP1', 'ERCC8', 'SLC24A1', 'ENPP3', 'TCEB3CL2', 'FGF2', 'SMAP1', 'DAB1', 'CPAMD8', 'USH2A', 'SLC40A1', 'LNX1', 'CACNA1A', 'STXBP5L', 'CRHR1', 'E2F4', 'CHST14', 'RASSF5', 'THSD7A', 'LOX', 'PLCG2', 'CCDC146', 'ARPC1B', 'POLG', 'PILRA', 'CITED2', 'ENSP00000482816', 'PEBP1', 'KREMEN1', 'C9', 'ARF1', 'CLCC1', 'NACA2', 'DYRK1A', 'POLR2D', 'RYR1', 'TGFBI', 'NRL', 'PARN', 'MSL1', 'UQCC1', 'SPRED1', 'NR4A1', 'GAA', 'NLRP3', 'MGAT4B', 'LARP4B', 'B3GALT6', 'PSMD14', 'S100A12', 'MUC5AC', 'CNGA1', 'SYT1', 'ZFP57', 'VANGL2', 'CREB3', 'HSPA5', 'ITGB6', 'AP1S2', 'LUM', 'ENSP00000481824', 'FAM19A2', 'TGFBR1', 'GTF2A2', 'PRKAR2B', 'DCN', 'SEMA5A', 'CX3CR1', 'TUBB2A', 'PBX2', 'VAX1', 'RPL24', 'C6orf165', 'CACNA1B', 'MIB1', 'ABCD4', 'SORBS3', 'ARMS2', 'FBN2', 'MAPT', 'USP45', 'ADM', 'PSENEN', 'FXR1', 'MAP3K11', 'TBX22', 'BCOR', 'MAPK8', 'PSMA2', 'AGPS', 'WASL', 'PDCD1', 'ACOX3', 'DCAF4', 'GNPTAB', 'RDH11', 'FXYD4', 'SDHAF1', 'RPL14', 'PARP4', 'VWA3B', 'ATP1A4', 'MDM4', 'TUBB3', 'SLC17A5', 'RNF213', 'FZR1', 'ADAM12', 'NOTCH4', 'DPEP1', 'HMGA1', 'TACSTD2', 'RPL7', 'CALR', 'HNMT', 'DVL3', 'RXFP1', 'ERP44', 'ANKHD1', 'SCN3B', 'PSMD2', 'ARHGEF3', 'ADAMTS8', 'RPS27', 'RAF1', 'RPS29', 'MTA2', 'CAPSL', 'BBS2', 'CRYGD', 'RPGRIP1', 'ATXN2', 'TAF2', 'LRP6', 'RPS6', 'MCCC2', 'TAAR9', 'KAT2B', 'ARMC9', 'CRK', 'NCOR1', 'ABCG5', 'PSMB9', 'EGLN3', 'MOCS1', 'SMAD2', 'FANCI', 'GTF2E2', 'CBL', 'KSR1', 'GPC3', 'STRN', 'NHP2', 'RNF185', 'IRX6', 'CCBE1', 'PCMTD1', 'NCF2', 'DRAM2', 'ANO1', 'CNNM3', 'DOCK2', 'SLC19A2', 'STAT2', 'PITX1', 'GPBP1', 'ARRB2', 'BRD4', 'USP8', 'CUX1', 'DYNC1I1', 'ALG13', 'SPPL2A', 'SIAH2', 'ITGB4', 'ALLC', 'SLC26A3', 'SUPT4H1', 'GSK3A', 'CD109', 'FN1', 'IL2RA', 'COG8', 'ABI2', 'LRRFIP1', 'TBL1X', 'COLEC10', 'ITPR3', 'RPIA', 'PLXND1', 'SMC2', 'C8orf37', 'AHR', 'VIP', 'WDR64', 'DAK', 'POLR2G', 'C15orf26', 'CFB', 'ACACA', 'GDF6', 'CTSA', 'TPM3', 'ARPC1A', 'NUPL1', 'TMEM56', 'EXOC3L4', 'CISD2', 'PMF1', 'COX6B2', 'OPLAH', 'CD86', 'CHCHD6', 'UQCRB', 'MYH2', 'GNS', 'RHO', 'GNG13', 'AADAC', 'OMG', 'SLC44A4', 'KMT2D', 'LRPAP1', 'EPS15L1', 'IL1RL1', 'HSF4', 'OPN3', 'NUP205', 'TBC1D24', 'CNN2', 'CYP2R1', 'IRS1', 'CHMP1A', 'LGALS4', 'NAA10', 'BFSP2', 'FGR', 'ATP13A2', 'NUP133', 'HERC2', 'RPS20', 'SLC1A3', 'THOC2', 'D2HGDH', 'EIF1AX', 'GRM5', 'LRP1', 'NT5C3A', 'UBAP1', 'SND1', 'TBC1D2', 'TP63', 'GPC1', 'MAML2', 'F11', 'MYH6', 'FTCD', 'HLA-B', 'ENO3', 'AGGF1', 'TNC', 'GPS2', 'HRH2', 'PLAGL1', 'GBAS', 'TBK1', 'CD36', 'RICTOR', 'MC3R', 'TRIM27', 'HLA-DQA2', 'F2', 'ITGB8', 'TICAM1', 'BTD', 'C1QB', 'HCK', 'NPSR1', 'ESF1', 'ZIC2', 'RAB5A', 'PSMA3', 'CHRDL1', 'PSIP1', 'SCYL1', 'CRYBA4', 'TRIM32', 'HEATR3', 'TBC1D32', 'ANTXR2', 'FXYD7', 'TSPAN8', 'CDH11', 'IGHV3-11', 'CLRN1', 'RPS7', 'SKP1', 'FYN', 'RHBDF2', 'MYC', 'HLA-DRA', 'POM121C', 'PGM2L1', 'PIK3CD', 'SKI', 'PARP9', 'TMEM30C', 'GLIS3', 'CAPN2', 'IRF6', 'NDUFA6', 'ARID3A', 'DKK4', 'ASTN2', 'VPS25', 'PPP3R2', 'FOXP1', 'EXTL3', 'CAMK2A', 'SLCO5A1', 'VPS37A', 'ADAMTS18', 'TUBB8B', 'PSMA6', 'REV3L', 'NRXN3', 'RBFOX1', 'PAFAH1B1', 'DHX38', 'APP', 'MBD4', 'ABL1', 'SLC12A3', 'UBA52', 'CCER1', 'PRCD', 'FLNA', 'COL3A1', 'ENSP00000262283', 'PIAS1', 'DEPDC5', 'PSMD12', 'SMG6', 'NFKBIA', 'SFTPB', 'ATP1A1', 'PKLR', 'SCAND1', 'BECN1', 'PTCH1', 'C11orf30', 'RBBP4', 'SLC27A4', 'HLA-DPB1', 'CDK9', 'ST6GAL1', 'BEST1', 'MFAP5', 'C1QBP', 'F8', 'KIF6', 'BLK', 'BTRC', 'VSX1', 'ABCA3', 'SRD5A3', 'NDFIP1', 'FSHR', 'PRDX1', 'EXO1', 'PSMC2', 'GTF2H3', 'KLF11', 'KIF3B', 'KLC4', 'RPS23', 'BDKRB2', 'ASB10', 'IL12RB2', 'SEH1L', 'POMGNT1', 'SLIT3', 'SLC35A3', 'LY96', 'PRKCSH', 'AP2B1', 'GALNS', 'DYNLL1', 'ADCY8', 'BRE', 'PRKAG2', 'MYO18B', 'AKT1', 'RAD51B', 'CHST5', 'RCC1', 'JDP2', 'MUC5B', 'CLPTM1L', 'AKT2', 'UQCRC1', 'FANCE', 'NUMA1', 'C10orf11', 'NYX', 'TLR4', 'ADNP', 'NDUFA4', 'CSPG5', 'TNR', 'CD14', 'CYP11A1', 'BABAM1', 'TTC39B', 'BIN1', 'ITGB1', 'SLC6A19', 'TDRKH', 'FAT3', 'GNB3', 'GPBAR1', 'WWP2', 'OPA3', 'IL17A', 'DYNLL2', 'VCL', 'TBCC', 'SLC16A12', 'DERL3', 'JUP', 'MPI', 'CARF', 'GPATCH3', 'STARD13', 'IKBKG', 'PDPK1', 'ESPN', 'KAT6B', 'MS4A6A', 'CCDC141', 'HMG20B', 'KIAA1217', 'NMT2', 'SMC1A', 'CDC16', 'LAMC2', 'GVQW3', 'FGF16', 'SLC25A6', 'TSPAN12', 'PRELP', 'ADAMTS13', 'ADK', 'CDHR1', 'FOXO1', 'CELA2B', 'TGM3', 'NFKBIL1', 'DIBD1', 'CAMK2D', 'PDGFRA', 'PYY', 'FGF9', 'SPTA1', 'PGA4', 'RGR', 'BPTF', 'LINGO2', 'NDUFC2', 'DMPK', 'PTPN12', 'GSDMA', 'IL7R', 'FGFR4', 'ING5', 'AEBP1', 'HIST1H2AH', 'SGSH', 'ALS2CR11', 'PCCA', 'CEP290', 'RPLP1', 'CDK8', 'VIPR2', 'SLITRK4', 'SNRNP200', 'TNFRSF1A', 'EXD1', 'STON1-GTF2A1L', 'TAAR1', 'ZDHHC24', 'CTNND1', 'AGPAT3', 'GPR98', 'POLE2', 'JAK3', 'NDUFB4', 'PIK3CA', 'WNT10B', 'ADAMTS2', 'AKT1S1', 'GTF2F2', 'RAD51C', 'SOD2', 'TRAF3IP2', 'PIK3R4', 'PRSS56', 'SYK', 'STAT1', 'NRAS', 'HIST3H2BB', 'HTR7', 'C12orf57', 'ERCC3', 'CUL5', 'POLR2L', 'SAG', 'ZNF513', 'RAB28', 'ZNF804B', 'ABCA4', 'DHODH', 'ENPP1', 'NDUFS2', 'CTBP1', 'ERAP2', 'FGF17', 'TUBAL3', 'GTF2H1', 'SDHA', 'LMBR1', 'ATP5G2', 'ATF4', 'CD96', 'MRC1', 'PTCHD1', 'MAP2K3', 'BBS7', 'NEIL1', 'GALNT1', 'HEPH', 'TSHR', 'NCK1', 'PRKAR1A', 'ALG11', 'TUBGCP2', 'TLR3', 'RPN1', 'ABCB4', 'MTG2', 'RPS13', 'PSMC5', 'ATP5F1', 'RIPK4', 'PTGER4', 'CREB3L1', 'C5orf56', 'ADAMTS20', 'DERA', 'KIAA1549', 'JPH3', 'FLI1', 'SUOX', 'AGTRAP', 'EXOSC2', 'MMADHC', 'KIF5B', 'UQCRHL', 'PHOX2A', 'DNMT3A', 'IL18', 'LZTFL1', 'NBEA', 'LRP5', 'CCL5', 'FBN1', 'HADHB', 'VDAC2', 'PTPRJ', 'WDR36', 'RBM10', 'GLDC', 'ZBTB16', 'CRYGS', 'DRD5', 'FGF19', 'TMEM67', 'PAPSS1', 'GPX4', 'SLC16A1', 'CREB3L4', 'MCM8', 'ACOXL', 'CYP11B1', 'ZAK', 'CDK5', 'LMF1', 'TRNT1', 'ARAP2', 'CAV3', 'GRM6', 'CHMP4C', 'BBS9', 'ELOVL4', 'SRBD1', 'SFSWAP', 'PRKG2', 'PLOD1', 'SNF8', 'FDX1L', 'KRT8', 'CC2D2A', 'COG7', 'PRKCD', 'GTF2H2', 'NCAN', 'CCDC122', 'TBCK', 'IFNGR2', 'NDUFA9', 'TTC25', 'WDR72', 'GUCY2F', 'GTF2B', 'ATP5B', 'TACR2', 'APOE', 'TNKS', 'PIK3R5', 'NRG2', 'MUC17', 'NPLOC4', 'CHMP3', 'SUGCT', 'ESR2', 'ACAN', 'MTA3', 'NECAB2', 'USH1G', 'PLG', 'TWIST1', 'MAGI2', 'CDK7', 'DPM2', 'SPC24', 'STAC', 'LAS1L', 'NFKB2', 'BET1', 'ABI1', 'FBXL7', 'MCCC1', 'F10', 'GTF2F1', 'ZEB2', 'GSS', 'JUN', 'STK4', 'TAS2R5', 'XRCC6', 'TP53', 'UNC50', 'SV2A', 'FKBP1A', 'MCU', 'CCDC149', 'MNX1', 'CKAP2L', 'GCLM', 'ST6GALNAC4', 'RIPK2', 'KPNA4', 'RPS18', 'GALT', 'CSMD1', 'LMNB1', 'MTRR', 'COL6A1', 'PACS2', 'DHDDS', 'MT-ND4', 'TLR5', 'MAK', 'GDAP1', 'PAX4', 'SLC6A3', 'TIRAP', 'ATRX', 'ATXN3', 'TLR1', 'HHAT', 'RPL13', 'TNFRSF11A', 'SLC7A7', 'SLC10A7', 'RPL38', 'RPS3', 'XRCC2', 'UGT1A8', 'ADAT3', 'CD4', 'CRYGB', 'MTFMT', 'ZC3HAV1', 'SEC13', 'SUMO1', 'ALDOB', 'ANAPC1', 'CACNA1S', 'LSS', 'DUSP6', 'ZBTB4', 'MANBA', 'DNAH6', 'AK5', 'CLCN6', 'SH3PXD2B', 'TIMMDC1', 'MPRIP', 'PROM1', 'CYP2B6', 'LITAF', 'JAK1', 'PRKACA', 'PAX6', 'TBC1D20', 'RBP3', 'HIST1H2BL', 'GADD45G', 'TCF4', 'APOC1', 'ATG5', 'DHH', 'PTGER2', 'CCNT2', 'GPANK1', 'HLA-C', 'IL4', 'PRPS1', 'APAF1', 'IRF4', 'NIPBL', 'INPP5E', 'MAGI1', 'CASP9', 'PTF1A', 'SORD', 'CSNK2A3', 'KIF17', 'NEU1', 'SLC4A4', 'SNX24', 'CCNI2', 'PSMA8', 'CSPG4', 'BCL6', 'RAX', 'KIF5C', 'NPHP1', 'CHD4', 'CTSL', 'ARPC3', 'PVRL4', 'FAR1', 'SH2B3', 'NCAM1', 'UNC13D', 'HOXB6', 'GRB2', 'MTA1', 'FMN2', 'C8B', 'SLX4', 'ST14', 'CAST', 'PDCD6IP', 'IL6ST', 'UQCRFS1', 'RPS15A', 'C6orf10', 'ZFHX4', 'CD9', 'FGF12', 'GNGT2', 'FIBCD1', 'ITPR1', 'ARL6', 'FGF1', 'TANGO2', 'ITPR2', 'C1orf127', 'MAPK1', 'HMCN1', 'MED25', 'ADAMTS4', 'NOTCH1', 'ILDR1', 'BBS4', 'EPDR1', 'HNF1B', 'GNGT1', 'PSMD10', 'SCAF8', 'SLC26A4', 'P4HA2', 'CDH13', 'SORCS2', 'MVB12B', 'AKR7A3', 'NDUFS1', 'DCHS1', 'NUP107', 'YES1', 'TFAP2D', 'BZW1', 'SAMD9', 'HCCS', 'C2', 'POU2F1', 'ZNF71', 'HOXD13', 'PSEN2', 'AP1S1', 'KERA', 'FHIT', 'UBC', 'PNLIPRP3', 'NPS', 'TYRP1', 'GGT1', 'FADS2', 'LACRT', 'ZNF469', 'SPON2', 'PHB', 'LIG4', 'SENP7', 'ANK3', 'MT-CO2', 'CRIM1', 'HIST1H2BB', 'ESCO2', 'TAF13', 'MFRP', 'MT-ATP6', 'CD247', 'NMT1', 'RPS11', 'TRIM29', 'GOLGA4', 'SLC12A1', 'FMO3', 'FLT3', 'NEK2', 'ROCK1', 'PSMB3', 'RPL19', 'RAP1A', 'CNTN3', 'MRPL11', 'OCRL', 'REN', 'MED13L', 'CDK2', 'TRAK1', 'IGF1R', 'FEZ2', 'EVI5L', 'SEL1L', 'ADAMTS7', 'VDAC3', 'MYO5A', 'C3AR1', 'COL6A2', 'AHNAK', 'MYH3', 'NUP210', 'SLC4A1', 'RPS19', 'SYNGR1', 'ATP1B3', 'EVI2A', 'C3', 'SLC3A1', 'MUC15', 'SLC29A3', 'BMP7', 'ARFGEF1', 'MLLT10', 'DUSP10', 'PLCL1', 'MAPK11', 'GNAT2', 'OGG1', 'MAPK9', 'EP300', 'NDUFV1', 'FBXO31', 'BAIAP2', 'RYR2', 'NUP43', 'EXOC2', 'PSMC1', 'OS9', 'PHF19', 'MMP1', 'MUC22', 'RPS4X', 'LRIG1', 'TRAF1', 'NDUFB3', 'UNC45B', 'NDUFA1', 'BRAF', 'GPR27', 'MYO10', 'BRMS1', 'RPL34', 'BCL11A', 'ARID4A', 'CALCRL', 'SYNE1', 'ROBO3', 'SPOCK1', 'EPYC', 'CACNA2D2', 'RPL23A', 'GBE1', 'AP1G1', 'CAV2', 'TTC27', 'GTF2E1', 'CCDC60', 'DOCK3', 'CEP63', 'STAM2', 'CHML', 'TAF1L', 'BBIP1', 'UQCR10', 'CHST6', 'RPS27L', 'TSHB', 'KIAA0196', 'MRVI1', 'PDE4DIP', 'STK19', 'GAN', 'LOXL1', 'SLC22A12', 'WISP1', 'POM121', 'HADHA', 'GAB1', 'DERL1', 'COX6C', 'FCGR1A', 'ALMS1', 'KCNH7', 'LAMC1', 'NDUFA11', 'EYS', 'MYO9B', 'HIST1H2BD', 'COA4', 'COX7B2', 'ERAP1', 'CSNK1A1', 'ANKHD1-EIF4EBP3', 'RASSF8', 'TGFB3', 'IKZF3', 'ABI3', 'CFHR4', 'ERCC2', 'GIF', 'CNKSR3', 'MEGF9', 'ADCY6', 'COMP', 'HCN4', 'SLC5A1', 'LAMA3', 'DSP', 'EPB41L3', 'CABP4', 'MAOA', 'SDHB', 'DYRK1B', 'DOCK9', 'HMGA2', 'SFTPC', 'CCDC91', 'PARVB', 'PPIA', 'FAM19A5', 'AHI1', 'FGF4', 'RPN2', 'CLMP', 'PRPSAP1', 'RXFP2', 'LOXHD1', 'ATP5G3', 'CREB5', 'FURIN', 'RCOR1', 'NCOR2', 'SERPING1', 'DPAGT1', 'GRIN3A', 'GYPC', 'SLC16A7', 'TRAT1', 'HIST1H2BA', 'ATP1B2', 'HNF4A', 'COX4I2', 'IPCEF1', 'MAP2K6', 'XYLT2', 'P4HB', 'FCHSD2', 'MCIDAS', 'NEURL1B', 'DVL1', 'DPM1', 'STRA6', 'GFER', 'RIPK1', 'PAPSS2', 'NDUFB10', 'ABCC9', 'IGSF23', 'PHF6', 'LMCD1', 'ARMC1', 'KIFAP3', 'RORC', 'FANCF', 'RPGRIP1L', 'HNF1A', 'SLC39A11', 'SNTN', 'GLTSCR1', 'POLR1C', 'MAP2K4', 'GPNMB', 'CLIC2', 'SIN3A', 'ELMO2', 'FUT6', 'HGS', 'SLC1A4', 'CDC27', 'DPP6', 'CASP8AP2', 'RSPO1', 'SKIV2L', 'MET', 'RCC2', 'ADAMTS19', 'IDH1', 'ERLEC1', 'RIPPLY2', 'PRRC2C', 'WWOX', 'E2F3', 'KPNA1', 'POMC', 'BUD13', 'OAT', 'FGFR1OP', 'PPCDC', 'SLC6A17', 'MED12L', 'FLG', 'TGFB2', 'COL10A1', 'GATAD2B', 'HLA-A', 'ENSP00000423550', 'SHF', 'MKKS', 'ADARB2', 'EDIL3', 'TBL1XR1', 'DKK1', 'ANTXR1', 'ATP5E', 'PLK2', 'COL14A1', 'SHH', 'ALDH2', 'FDXR', 'SKP2', 'ABCB11', 'GPR179', 'KCNV2', 'MITF', 'KDELR2', 'DAG1', 'CALCB', 'TXN', 'STAB2', 'P2RX4', 'FRY', 'NUPL2', 'CACNB1', 'TLL1', 'PSMF1', 'SNCA', 'CDC25B', 'LDLRAP1', 'NDUFA7', 'SLC11A2', 'GALNT3', 'PPP3CA', 'CYP1B1', 'COL13A1', 'HIST2H2AB', 'UBE2I', 'CENPK', 'MAML3', 'SLC7A5', 'PML', 'ADORA2A', 'APOB', 'ADAMTSL3', 'RPSA', 'RPL11', 'FGF8', 'GNAT3', 'NEDD4L', 'IDH3B', 'UROD', 'LHFP', 'OR51B5', 'GDF5', 'NEUROG2', 'MECOM', 'GANAB', 'MPP4', 'CALN1', 'CREBBP', 'TIMP2', 'MUC21', 'HDAC4', 'GPR180', 'CHMP2A', 'TRIM21', 'ARHGAP42', 'KCTD1', 'LRRC23', 'PACRG', 'PPP2R3C', 'SIM2', 'SLC35A2', 'HBEGF', 'PSMD11', 'PRSS36', 'INS', 'DSE', 'MUC20', 'CBX1', 'ATP5H', 'CPVL', 'NDP', 'UQCRQ', 'FXYD3', 'GNG2', 'WDR19', 'PSMD8', 'SOST', 'SIX1', 'IMPG2', 'TFEB', 'RPS24', 'EPGN', 'PIEZO2', 'NDC1', 'AEBP2', 'NDUFA4L2', 'CTBP2', 'OPN1LW', 'PAX2', 'PTCHD3', 'APH1B', 'GATA6', 'IMPDH2', 'MSRA', 'RPL13A', 'CCNK', 'ADAMTS1', 'STT3A', 'APBB1IP', 'THSD1', 'BHLHA9', 'TREM2', 'NEDD9', 'ATP5D', 'HEXA', 'CPM', 'MC1R', 'PTGFR', 'CWC27', 'TAF7L', 'PRUNE2', 'CRYGC', 'STX1A', 'AMN', 'FRS2', 'ERLIN2', 'IRAK1BP1', 'TRIM24', 'CTC1', 'GREB1', 'CCDC85A', 'SPSB2', 'PRMT2', 'POMT2', 'SEMA4A', 'DYNLT1', 'MTCH2', 'ELMO1', 'HIST1H2AC', 'RAMP1', 'CNGB1', 'DMWD', 'CCDC22', 'HSPA1B', 'IL1R1', 'MAP1LC3B', 'BCL2L11', 'IFT27', 'COTL1', 'PIK3C2A', 'TCEB3C', 'NOXO1', 'LEMD3', 'GNAI3', 'PSMA7', 'GPR150', 'COL9A2', 'DLL3', 'CLUAP1', 'CYP27B1', 'MTOR', 'GTF2H5', 'SLC4A11', 'ITIH2', 'COG3', 'FOXO4', 'TUBB', 'SLC14A2', 'PCCB', 'UBE2L3', 'CNGA3', 'ATP1A3', 'SEC24B', 'MGST1', 'RAB3GAP1', 'NMNAT1', 'CRB1', 'ADAMTSL5', 'CHRNG', 'CALCR', 'GREM1', 'RPL12', 'SIAH3', 'MED12', 'KIF3A', 'UQCR11', 'CYP19A1', 'PPP2R5D', 'SPECC1L', 'INSRR', 'NDUFV3', 'LHFPL3', 'ATP5O', 'GUF1', 'CDH12', 'GOLGB1', 'PSTPIP1', 'ENSP00000381234', 'RBM8A', 'SCTR', 'AP2A2', 'GNG4', 'ATP6V1B2', 'MSMO1', 'NCF1', 'MUCL1', 'KCNA4', 'NOTUM', 'ATP11A', 'ACTB', 'ZFP64', 'WNT16', 'PARP14', 'SMO', 'IL19', 'TRPM1', 'IDS', 'EGF', 'CUX2', 'MSH5', 'SLC5A2', 'HOXB3', 'PSMD4', 'TULP1', 'GATA4', 'SEPN1', 'PSMC3', 'ACTC1', 'RPL22L1', 'GCKR', 'NOP10', 'PAX3', 'NEGR1', 'FGFR1', 'ADAMTS16', 'THEGL', 'AK6', 'FANCD2', 'MT-ND5', 'MAD1L1', 'NREP', 'SOS1', 'NRROS', 'DDOST', 'GLS2', 'SLC5A5', 'TTN', 'DUSP8', 'ABCA1', 'VPS4A', 'GP9', 'AIPL1', 'WAC', 'CHMP4A', 'WRB', 'IGF1', 'SH3GL3', 'VWA7', 'LHCGR', 'NEK9', 'VPS4B', 'TPO', 'KCNU1', 'KPNA3', 'COX7A2L', 'PPP1R16B', 'PIK3R1', 'SH3GL1', 'SLC6A14', 'BCR', 'NCSTN', 'ENPP6', 'NCKIPSD', 'BAD', 'VIPR1', 'SLC38A8', 'GPC5', 'GTPBP1', 'PRPF4', 'FRK', 'RPS8', 'BTC', 'GNAS', 'LHX3', 'NAPEPLD', 'FBXO7', 'CD46', 'PPP1CB', 'PITX2', 'ZNF276', 'PLD1', 'TMCO1', 'RCVRN', 'BBS10', 'SI', 'PDE6G', 'DNAJA3', 'ERCC4', 'DDX5', 'CDKN2A', 'APOBEC3G', 'GJB2', 'TBX4', 'MGAT4C', 'FAM46A', 'PSMC6', 'DLEU1', 'COLQ', 'MYT1L', 'NOTCH2', 'ADCY7', 'SDHD', 'LY86', 'CHMP6', 'FAM155A', 'PRNP', 'KCNIP4', 'NUP37', 'KIF26B', 'KIAA1109', 'DNAH14', 'MAPKAPK3', 'ADCY9', 'TNXB', 'UNC5D', 'PTH', 'RGS9', 'MSH2', 'MT-CYB', 'LCT', 'ANKRD13A', 'EFEMP2', 'SALL1', 'IFNAR2', 'FKBP4', 'AAR2', 'ATP5A1', 'RHBDF1', 'XYLT1', 'EPB41', 'CLPB', 'GATA1', 'ARSB', 'HDAC9', 'HIST2H2BE', 'FCGR3A', 'SPTB', 'IL6R', 'AP1M2', 'SPRED2', 'RNF144A', 'MAPK13', 'CREB1', 'TGFBR2', 'IL4R', 'TNIP1', 'ATOH7', 'RFWD3', 'HIST1H2AE', 'FEN1', 'ATF6B', 'LCK', 'AP2S1', 'GRIP1', 'BIN2', 'TRPS1', 'MT-ND1', 'FGF22', 'PSMD7', 'MUC3A', 'CBLL1', 'NEIL3', 'SOX2', 'NUP88', 'CC2D2B', 'TUBB4A', 'ANAPC5', 'GCNT2', 'HDAC3', 'CYFIP2', 'RPS16', 'SLCO1B1', 'PTH2R', 'MGAT4A', 'MSH6', 'FOXC1', 'ABCC8', 'SEMA4B', 'RAX2', 'TUBA4A', 'EVA1C', 'CNTN4', 'RGS13', 'CHMP5', 'GATA3', 'PIGW', 'CLNK', 'DENND1B', 'PHGDH', 'AVPR2', 'WDR49', 'PSMB4', 'PSMG4', 'TOMM40', 'NAV3', 'BBS12', 'GIP', 'SSPN', 'AXIN1', 'PKD2', 'FMR1', 'CD28', 'STX1B', 'ENSP00000469970', 'HIST1H2AA', 'NELFCD', 'CYSLTR2', 'ARPC5', 'DCBLD1', 'GALNT12', 'TXNRD1', 'BBS1', 'RBMS3', 'ENTPD5', 'TMEFF2', 'VIM', 'CRAMP1L', 'PSMD1', 'BNC2', 'SLC6A5', 'CD8B', 'KIF21A', 'WASF1', 'MUC7', 'GJA1', 'CREB3L3', 'GYG2', 'NLGN1', 'PDE6H', 'ACAD10', 'GRIN2B', 'EXOC3L2', 'CKAP5', 'PCNXL2', 'EPB41L1', 'RPL9', 'GPHB5', 'TCF7L2', 'ELL', 'NCF4', 'ENTPD1', 'HDAC6', 'NDUFB6', 'CCND1', 'CREB3L2', 'ABCA12', 'CAPNS1', 'HIST1H2AJ', 'TLR2', 'CACNA1D', 'CSMD2', 'RPL36', 'NELFA', 'COX7B', 'DNAH2', 'ZBTB18', 'MC2R', 'FADS1', 'DDR2', 'CFHR5', 'L3MBTL3', 'TBC1D21', 'TRDN', 'OGN', 'RDH5', 'HIST1H2AI', 'TNRC6B', 'HK1', 'SLC25A5', 'MRPL33', 'MACROD2', 'RPL37', 'TRIM28', 'SALL2', 'NCKAP1', 'IQGAP2', 'TDRD7', 'GYG1', 'DOCK1', 'LTB4R', 'CASZ1', 'ZKSCAN5', 'MUT', 'RGS9BP', 'SCAPER', 'CELF2', 'SLC7A14', 'APPL1', 'WSCD2', 'BRIP1', 'PTPRF', 'NUP188', 'LTBP3', 'TRIP11', 'PRDX2', 'MAPK12', 'DECR2', 'MIB2', 'RPL28', 'ARHGAP15', 'DLG2', 'EFHC1', 'JHDM3C', 'SPATA7', 'AP1B1', 'PTPRB', 'CPSF4', 'DCXR', 'PPP2CB', 'KCNT2', 'MMP2', 'EFNB1', 'C6', 'PRPH2', 'PTPRN2', 'ATP6V1G2', 'NUP85', 'G6PC', 'BRF1', 'KHDRBS2', 'HIST2H2BF', 'P2RX7', 'BMS1', 'NDUFB7', 'PIFO', 'RPS6KA2', 'ADAMTS9', 'ARHGEF18', 'DLL1', 'CFP', 'CRYBB1', 'F5', 'MARK3', 'RAD21', 'SLC37A4', 'NOX1', 'CDC42', 'ENSP00000472316', 'LAMA1', 'MOGS', 'CAPN5', 'HDAC1', 'GJA3', 'ETV6', 'TMEM232', 'CCAR2', 'BANF1', 'THSD4', 'RLN3', 'CYP20A1', 'AMER1', 'ESRP1', 'RPS2', 'CHD6', 'SNAP91', 'ADAM10', 'PRKX', 'ADAMTSL2', 'CDH1', 'GNAT1', 'G6PC3', 'USB1', 'ATP6V1H', 'SLC22A16', 'EXT2', 'FLVCR1', 'CRISPLD2', 'CETP', 'OPN1SW', 'CDKN2B', 'FRAS1', 'CRYBA1', 'DNAJB11', 'RAMP2', 'ATM', 'B3GNT1', 'SLC7A9', 'BACH2', 'FZD5', 'GNG7', 'NUP153', 'SELO', 'XRCC4', 'PKD1', 'MRAS', 'POU4F1', 'SV2B', 'ARHGAP31', 'C1QTNF7', 'GIPR', 'EOGT', 'FREM2', 'RPL21', 'BANP', 'TLR6', 'LZTR1', 'ARRDC1', 'RBBP7', 'RHOA', 'PITX3', 'IL5', 'CASP3', 'MOBP', 'LPIN1', 'PIGX', 'SELP', 'CD19', 'GINS2', 'IQCK', 'NDUFS8', 'ARHGEF4', 'LAMB3', 'HIST1H2BM', 'CD8A', 'PHACTR1', 'BRCA2', 'GAPDH', 'RHAG', 'RPL29', 'RORA', 'SLC34A1', 'RBP1', 'WISP3', 'ANAPC7', 'SCT', 'C14orf37', 'ADCY2', 'RNGTT', 'TPPP', 'GLB1', 'FANCM', 'QTRT1', 'HDAC5', 'C8G', 'TMEM98', 'NFS1', 'CAPN1', 'BMP3', 'PGM3', 'NISCH', 'SRC', 'ESRRG', 'NDUFS5', 'HLA-DOB', 'KLHL1', 'NRG4', 'DKC1', 'C1orf168', 'NDUFV2', 'FLT3LG', 'HMGXB3', 'KANK1', 'NCKAP1L', 'PARK2', 'LONP1', 'RP1L1', 'CDC23', 'NDUFA5', 'SLC9A9', 'YWHAB', 'PCYT1A', 'RPL27A', 'NOXA1', 'PSMC4', 'GNB4', 'NBEAL1', 'MYH11', 'LARP7', 'PIK3R3', 'ABCC2', 'PCDH15', 'PDE6C', 'B4GALT7', 'CDKN1A', 'HAP1', 'PGF', 'FAM105A', 'PDE7B', 'ATP1B1', 'F13B', 'SLC9A2', 'SLCO2A1', 'HSP90AB1', 'CD55', 'CR1', 'CLDN19', 'USF2', 'HEMGN', 'MROH8', 'KRT12', 'ETS1', 'FXYD2', 'THSD7B', 'IL10', 'ITGB3', 'LACTB2', 'TOX2', 'SLC12A6', 'SSRP1', 'SMARCA4', 'COL4A4', 'SSB', 'PDE6B', 'SSBP2', 'RPS9', 'MYPN', 'HOXD11', 'TTR', 'EPHB4', 'GSK3B', 'INS-IGF2', 'LRRIQ4', 'SLC9A6', 'ANAPC15', 'SLC2A2', 'CDH3', 'ABCB6', 'RAB3GAP2', 'CLTC', 'TUSC3', 'CRYAB', 'KCNN3', 'TTC28', 'TRAK2', 'FCGR2A', 'INVS', 'C8A', 'GJB6', 'COX4I1', 'SEMA3E', 'LMBRD1', 'AMBRA1', 'MT-ND6', 'FNDC3B', 'HEY1', 'NEK10', 'C1GALT1', 'ROR2', 'GPR39', 'RAB23', 'GPR37L1', 'DLG5', 'ENSP00000480348', 'FRMD7', 'SLC9A4', 'TNNI3K', 'HDAC10', 'CAMK2B', 'ATP2B1', 'GCFC2', 'PLP1', 'MYH7B', 'NHS', 'CDK13', 'MMP14', 'FAF1', 'PVRL1', 'TNFSF4', 'EPG5', 'LIPC', 'NR3C1', 'ST3GAL2', 'SCO2', 'TUBA1C', 'VSX2', 'EGFR', 'TBX5', 'NAA15', 'CSNK1E', 'GLI3', 'DPEP2', 'RS1', 'RPL39', 'PDLIM5', 'GTF2A1', 'RBP4', 'P2RY11', 'POLR3B', 'GABRG2', 'ADCY3', 'TMEM258', 'PAMR1', 'TAAR8', 'SLC39A4', 'LIM2', 'C1orf95', 'CDC42EP4', 'ACER3', 'MT-ATP8', 'VPS37C', 'RPL35', 'IGFALS', 'SYN2', 'CCND2', 'MIP', 'COX7C', 'ATP5G1', 'ADRB3', 'MYO15A', 'PDE6A', 'FGA', 'RFT1', 'TAAR5', 'ITGA2B', 'CFHR3', 'NFKB1', 'PTPRD', 'PRKAR1B', 'TRIM59', 'MLLT1', 'ERBB2', 'STAT5A', 'PLA2G15', 'VCAN', 'MT-ND4L', 'BEST4', 'KPNB1', 'CORO1A', 'RPLP0', 'IL12A', 'SLC2A9', 'LDLRAD4', 'ORC1', 'CYC1', 'COX8C', 'DVL2', 'TEK', 'KLB', 'IKBKB', 'WDR27', 'FAM174B', 'TAZ', 'CTSG', 'TAF12', 'IHH', 'IL18R1', 'KIAA2026', 'RAB7A', 'NCALD', 'FILIP1L', 'MVB12A', 'PSMA4', 'PSMD13', 'B4GALT1', 'STAT3', 'STAT5B', 'FGFR1OP2', 'TBC1D5', 'COL17A1', 'GPATCH2', 'PXDNL', 'NDUFA12', 'VPS11', 'XKR9', 'AHCYL1', 'COL8A1', 'MAN1B1', 'FGG', 'PARD3B', 'EZH2', 'NUP98', 'NRG1', 'EREG', 'GLP2R', 'MAP3K7', 'UBAC2', 'WNT10A', 'RAC1', 'S1PR1', 'HIST3H2A', 'ADAD1', 'C9orf92', 'KCNMA1', 'INIP', 'WASF3', 'PCSK5', 'EIF1B', 'KAT2A', 'MNAT1', 'RPL37A', 'MAPKAP1', 'CERKL', 'SMYD2', 'YWHAE', 'UNC93B1', 'GALNT2', 'FERMT1', 'ABHD5', 'PTH1R', 'AQP4', 'VPS28', 'STAT4', 'CYP27A1', 'FRMD4A', 'PDGFRB', 'GPR83', 'KDM1A', 'MC4R', 'PSME4', 'SLC1A1', 'TACR1', 'UGT1A4', 'TYK2', 'CAMK2G', 'PCSK6', 'PIK3R2', 'TGFB1', 'PPP2R5C', 'TOPORS', 'C5', 'NDUFB11', 'FGF10', 'PDE6D', 'B3GAT3', 'CYP26C1', 'DLD', 'GRIN2D', 'SHANK2', 'RASA1', 'PITPNC1', 'IGBP1', 'TUBB2B', 'KRT5', 'MDM2', 'NUP62', 'TUBB6', 'IFT122', 'PPARG', 'RNF39', 'COG6', 'SCLT1', 'CHMP2B', 'FANCG', 'UBE2D1', 'TTLL5', 'LMNA', 'COX8A', 'TFRC', 'ICOS', 'CYBA', 'TAC1', 'RMI2', 'TAF3', 'PIGL', 'RPL35A', 'KIF5A', 'TERT', 'TNS3', 'SIX3', 'NCBP1', 'SLC44A1', 'STRADB', 'GPC6', 'IL33', 'SRPK2', 'SPON1', 'COL5A2', 'SDK1', 'PQBP1', 'XRCC5', 'PCSK9', 'GCG', 'FGD1', 'SIK3', 'NUP93', 'MAN1C1', 'CALM3', 'RPL18A', 'DNMBP', 'STK24', 'DAXX', 'HDAC8', 'RFX3', 'TAF15', 'FRMPD1', 'BCAN', 'RAN', 'ADCY4', 'HIVEP3', 'POU5F1', 'GP1BA', 'C1QC', 'COX5A', 'CELF1', 'BOLL', 'CFHR1', 'PGK1', 'FUZ', 'AP3B1', 'NDRG1', 'TENM4', 'EED', 'RPS14', 'GIGYF2', 'KCNJ11', 'PTPRT', 'ABCA2', 'MMP3', 'HPSE2', 'CRHR2', 'RNMT', 'PCLO', 'EAF1', 'ZNF408', 'EGR1', 'ANAPC10', 'RPL27', 'TENM3', 'TUBA1A', 'FOXE3', 'CST3', 'UROS', 'LRRC4C', 'ELP4', 'C3orf20', 'BMPR1B', 'EFNA5', 'HECW1', 'ERCC5', 'MUC6', 'GP5', 'CYP24A1', 'GRK7', 'TBP', 'MYT1', 'GNB5', 'FIP1L1', 'CSF2RB', 'AGK', 'GNAI2', 'POLR1D', 'COX5B', 'CCR5', 'ADAMTS17', 'SEC24D', 'RPL6', 'TAF5', 'GNB1', 'RBPJ', 'NCBP2', 'EEF2', 'UNC119', 'TSG101', 'TAF9B', 'GRIN3B', 'FOXL2', 'ANKS1B', 'AP1S3', 'LEMD2', 'TBCD', 'CHM', 'BRCA1', 'SLC35F3', 'FGF23', 'IPO5', 'ARAF', 'TBXAS1', 'NUP50', 'SV2C', 'TET2', 'PTDSS1', 'TBX3', 'AFAP1', 'C1QA', 'DLC1', 'RASGRF2', 'DKK2', 'FHOD3', 'GUSB', 'EFCAB5', 'CEP250', 'LIG1', 'COL25A1', 'KANK4', 'CYP17A1', 'LARGE', 'MUC4', 'PTPRK', 'TAF4', 'HBG2', 'RPE65', 'PI4K2B', 'ERBB4', 'PISD', 'NCOA6', 'ZCRB1', 'MYO5B', 'ALDOA', 'KCNQ5', 'NELFB', 'ROM1', 'ANAPC16', 'BEST3', 'RTEL1', 'TMEM39A', 'IL32', 'NADSYN1', 'COL18A1', 'HIST1H4C', 'KLHL7', 'CLU', 'UVRAG', 'CNKSR1', 'ACTR3', 'HDAC2', 'GNAO1', 'GLIS1', 'NGF', 'VPS33B', 'KCNJ18', 'ALG14', 'PBK', 'HAUS6', 'SNAP25', 'MESP2', 'CFH', 'POLR2I', 'ATP5J', 'CNGB3', 'PIK3AP1', 'NOTCH3', 'REEP6', 'HES7', 'RPS15', 'STON1', 'RYR3', 'ADAMTS3', 'PCSK1', 'PDGFB', 'FLNB', 'GPR158', 'FSTL5', 'GABRG3', 'DHCR7', 'ASPM', 'MMP9', 'F9', 'PSMB2', 'SH3GL2', 'PMS2', 'TMX3', 'MAP2K2', 'B2M', 'IL6', 'LYN', 'MBTPS2', 'FGF5', 'SLC26A2', 'LACC1', 'SFTPD', 'C6orf48', 'FOXA3', 'CAMK4', 'ENSP00000458866', 'POC1B', 'ADCYAP1R1', 'TMEM38B', 'LRAT', 'AGBL1', 'ELAVL4', 'CCNE1', 'SH3KBP1', 'RARG', 'CA4', 'METTL22', 'MAT1A', 'TMPRSS11F', 'TEKT1', 'TCEB2', 'C11orf65', 'ELOVL1', 'HES5', 'ITIH4', 'UBN1', 'BASP1', 'GJA8', 'HIST1H2BH', 'RTL1', 'MAF', 'ADRB2', 'BMP2', 'CIB2', 'ABCC6', 'NDUFB5', 'KCNG1', 'OSBPL1A', 'COL7A1', 'IQCB1', 'SLC36A2', 'TUBA1B', 'WFS1', 'SCN5A', 'HDAC7', 'TPR', 'EHD4', 'PNPLA6', 'ZDHHC19', 'PSEN1', 'TCP10L', 'PPP2R1B', 'MC5R', 'AREG', 'NDUFA2', 'GP1BB', 'CNTN1', 'PSMB8', 'AP2A1', 'PRKACG', 'EPHA2', 'IFT172', 'PARP6', 'UBE2C', 'PLEKHM1', 'TCEB1', 'HIST1H3D', 'SLC33A1', 'HIST2H2AC', 'POLR2A', 'USP9X', 'ZNF341', 'RERE', 'MDN1', 'IRF1', 'RPL32', 'RIC1', 'PNPLA2', 'SYNGAP1', 'PRTN3', 'PLEC', 'FANCL', 'SMC3', 'EFEMP1', 'MGAT1', 'PLCB1', 'SEMA5B', 'ATP6V1E1', 'ACY1', 'TSLP', 'PLCXD2', 'DYNC1H1', 'ARL2', 'RFTN2', 'MTAP', 'VPS37B', 'IGFBP4', 'LRP4', 'KLC2', 'GNE', 'TCTN3', 'SF3B4', 'CNNM4', 'GFPT1', 'KLKB1', 'SLC20A2', 'NOG', 'HIST2H2AA', 'PARP1', 'SLC2A1', 'GUCY2C', 'POP1', 'BSCL2', 'CLVS1', 'PPP1R14C', 'SLCO3A1', 'SLC6A18', 'CACNA2D4', 'TCEB3B', 'ADORA2B', 'COL1A1', 'PI16', 'SDC1', 'CYCS', 'FZD4', 'PITPNM3', 'OCA2', 'PDE11A', 'UBIAD1', 'CHUK', 'CASP8', 'ACTN2', 'ADM2', 'CHSY1', 'ARFRP1', 'KDM3B', 'CALCA', 'FDX1', 'KCNJ13', 'CENPC', 'MYOC', 'CTLA4', 'RGS7', 'FCER1G', 'OPTN', 'TIMP3', 'MYH9', 'DPP10', 'GSDMB', 'GNG10', 'FGF7', 'GUCA1B', 'IFNA2', 'IDUA', 'USHBP1', 'IRS2', 'TRAPPC12', 'UBE2E1', 'MUC12', 'COL8A2', 'SAP30L', 'C3orf17', 'FAM185A', 'IRF2', 'CFTR', 'CCNH', 'GUCA1A', 'ZNF670', 'FSCN2', 'ZNF644', 'GATAD2A', 'PC', 'CACNA1F', 'PSMB5', 'ALG12', 'ASPH', 'LAMA2', 'DOLK', 'KIT', 'MTX1', 'VANGL1', 'MAPK14', 'NIN', 'HEY2', 'MAML1', 'BRAP', 'PRR5', 'ANO9', 'BSG', 'NXN', 'PARP8', 'MEGF8', 'AGFG1', 'TMEM233', 'FGFR3', 'NEURL1', 'KAT6A', 'RPL17', 'SLC5A7', 'FBXO25', 'RP9', 'PORCN', 'TLR7', 'SDC2', 'SUPT5H', 'CD80', 'POMT1', 'GPR143', 'MMP24', 'ITGA4', 'C1orf94', 'SLMAP', 'GGCX', 'MAPK8IP1', 'DYNC1LI1', 'VEGFA', 'GNPAT', 'COL22A1', 'GUCY2D', 'ENSP00000449026', 'TUBA3D', 'KLC1', 'PPP1CC', 'ACE2', 'PPIF', 'BAALC', 'ARPC4', 'DLX5', 'CAPNS2', 'GALK1', 'APLP1', 'IFT74', 'HIST1H2AD', 'NAB1', 'GTF2IRD1', 'HSPA2', 'NEB', 'GRK1', 'KNG1', 'TNFRSF14', 'ROBO1', 'PKN3', 'BDNF', 'IL16', 'SDCCAG8', 'CSNK2A2', 'CTSH', 'TAAR6', 'RP1', 'DPM3', 'RPL22', 'THBS2', 'CRB2', 'LPP', 'PHLPP2', 'SLC17A8', 'CNBD2', 'RNF166', 'RB1', 'RPS28', 'FZD7', 'LFNG', 'KHK', 'PIM1', 'ATP1A2', 'TMPRSS2', 'KCNK3', 'C2orf71', 'RTTN', 'RUNX3', 'SLC34A2', 'DOCK6', 'EPHB1', 'DNMT3B', 'ACTR2', 'KCNH2', 'WNT3A', 'BRF2', 'MUTYH', 'ATG7', 'CDK4', 'POLR2H', 'SIGMAR1', 'RPS6KB2', 'CACNB4', 'ZMIZ1', 'CLTA', 'PCNT', 'SLC35C1', 'SERPINB7', 'ERMP1', 'HBE1', 'LRRC34', 'DNM1', 'PRKAR2A', 'FGF3', 'GLP1R', 'GNAZ', 'DYNC1LI2', 'PPP2CA', 'NEIL2', 'TNF', 'EPM2A', 'NKX2-5', 'AOAH', 'DGKQ', 'VAV3', 'TRAF3IP1', 'SPIB', 'SUPT16H', 'TAF4B', 'SALL4', 'APH1A', 'GPR84', 'AKAP9', 'ADAM9', 'SLC35F4', 'PEX7', 'PRPF3', 'RPL30', 'LRP1B', 'CRX', 'JAG1', 'PDX1', 'NDUFS3', 'RAD51', 'SLC24A5', 'OR5V1', 'RGS5', 'VTA1', 'ERCC1', 'ACTG1', 'OMD', 'DLL4', 'TPST2', 'RIT2', 'HNRNPK', 'JAG2', 'PIP4K2A', 'SERTAD2', 'ALG2', 'NUS1', 'SLCO1B3', 'WIPF2', 'GUK1', 'NT5E', 'RPS5', 'IARS2', 'GIPC2', 'MUC16', 'DRD1', 'CCL11', 'SHC1', 'AGBL5', 'ADAMTS14', 'COMMD1', 'ATAD2B', 'NHLRC1', 'CYP7B1', 'DPYD', 'CCNT1', 'CPS1', 'CYP26B1', 'GOLGA2', 'MLH1', 'TNFAIP3', 'NEUROD1', 'FADD', 'ALK', 'PSMA1', 'SERPINI2', 'VAMP1', 'RAVER2', 'RPL18', 'BAG4', 'NDUFA13', 'NUP54', 'CYP4F22', 'HLA-DQB1', 'LRP2', 'HRAS', 'PHLDB2', 'PTGIR', 'PDGFA', 'ATP5C1', 'FYCO1', 'NOX4', 'CEBPD', 'FAM114A2', 'COL5A1', 'OR51Q1', 'DNAH17', 'TNS1', 'PSMA5', 'CFDP1', 'RAMP3', 'REST', 'ADCY1', 'MVK', 'GPR45', 'TLE1', 'MLANA', 'SLC34A3', 'TAF1', 'NOS2', 'TSC2', 'RNF114', 'CLCF1', 'CEP19', 'ST3GAL3', 'RDX', 'FAM86C1', 'TNFSF15', 'HEXB', 'C9orf172', 'RBL1', 'EIF2AK4', 'CDC25A', 'SLC6A2', 'TXNIP', 'PPP1R3C', 'KITLG', 'REEP3', 'EIF2S1', 'RNF5', 'HMGB3', 'MGAT2', 'SHOC2', 'GPC4', 'CALD1', 'ABCF1', 'HLA-DQA1', 'LRRC3C', 'MT-ND2', 'ARMC5', 'NDUFB1', 'GHRH', 'ABCD1', 'NDUFAF6', 'EDEM2', 'KDR', 'NF1', 'FREM1', 'POLR2C', 'ACSL3', 'IRF5', 'PIGV', 'SLC22A5', 'JAZF1', 'HRH1', 'ABCA8', 'DYNLT3', 'WFDC1', 'KPNA5', 'LTA', 'GHR', 'COL16A1', 'TMCO4', 'RLBP1', 'ALB', 'CDK6', 'VAPA', 'SESTD1', 'ADAMTSL4', 'WT1', 'RPL23', 'ARSG', 'KCNN4', 'PDE4B', 'COX7A1', 'ZBP1', 'PHF21A', 'SBSPON', 'KCNQ1', 'FAM131B', 'GNG5', 'DOCK7', 'RPL3', 'PSMB6', 'NDUFC2-KCTD14', 'CTNNA2', 'PSMB7', 'MTR', 'TUBA3E', 'NPM1', 'HYAL3', 'CEP128', 'KIAA1614', 'TRAF3', 'NAV2', 'XPO1', 'SMOC1', 'VDAC1', 'KIAA1751', 'WDR12', 'AMPD3', 'MADD', 'TRPV6', 'BGN', 'ENSP00000451560', 'GNG8', 'IL13', 'ST6GALNAC3', 'SPINT2', 'COL4A1', 'VHL', 'RDH12', 'DDIT3', 'PRKCB', 'IL1RL2', 'CDKAL1', 'CUL1', 'FUT8', 'KIAA0195', 'SLC16A10', 'IRAK4', 'ARHGAP21', 'SCNN1A', 'PIGU', 'RPL36AL', 'KIAA1731NL', 'PTH2', 'SCN2B', 'PTPRR', 'CRBN', 'WASF2', 'CACNA1C', 'DNAJC6', 'PRPF31', 'SLC16A8', 'MYO1C', 'TLN1', 'ROCK2', 'SEMA3A', 'NAGLU', 'GNG11', 'EIF4G3', 'SLC35A1', 'CD3G', 'YAP1', 'RTP4', 'ELN', 'RGS22', 'SPAST', 'WDHD1', 'TRAPPC9', 'PACS1', 'HYAL1', 'CLDN1', 'ELFN1', 'ST3GAL4', 'ALDH1A3', 'SPTBN1', 'GGT5', 'PLEKHA1', 'STIP1', 'FGF6', 'MAP2K1', 'CHST3', 'GYS1', 'FLAD1', 'IAPP', 'CYP4V2', 'GTF2H4', 'ELAVL2', 'P2RY12', 'TESPA1', 'CADM2', 'GPR20', 'CDH23', 'KPNA7', 'ZFYVE9', 'NDUFAB1', 'GRIN2A', 'GRIN2C', 'CHD3', 'OPRM1', 'CDK5R1', 'KIZ', 'LTF', 'POLR2K', 'VMP1', 'MERTK', 'MT-CO3', 'IMPDH1', 'GNA15', 'CAMK1D', 'CP', 'BAP1', 'KLC3', 'B3GALTL', 'MMACHC', 'RNF43', 'SYVN1', 'LAMB4', 'SRCIN1', 'PSMD5', 'SUGT1', 'MT-CO1', 'LTBP2', 'PLXDC2', 'RPL8', 'EPB41L4A', 'TALDO1', 'PTPRC', 'AGBL2', 'TG', 'SCN10A', 'FGFR2', 'PTPN22', 'PPARGC1B', 'POLR2F', 'ST6GALNAC2', 'SOD1', 'TYR', 'SAP18', 'TRADD', 'EIF2AK3', 'SLC35D1', 'GNG12', 'RBFOX2', 'ADCYAP1', 'IFT140', 'THAP4', 'CCND3', 'CD58', 'MGAT5', 'ABCG8', 'PSME1', 'CHMP7', 'USH1C', 'ALG1', 'ENO1', 'SPIN1', 'RBBP8', 'FASLG', 'CSF2RA', 'MMAA', 'PDZD7', 'COX7A2', 'PRPF6', 'VAV2', 'ESR1', 'PRDM5', 'ZMYM2', 'SLC1A7', 'CTNNB1', 'PPOX', 'CHKB', 'MIF4GD', 'ADAMTS6', 'WDR48', 'FAM102A', 'TBX20', 'NDUFB2', 'TAF7', 'IER5L', 'TRIM26', 'HIF3A', 'PYCARD', 'CYBB', 'MTHFR', 'FAM180A', 'FIBP', 'BRK1', 'NTHL1', 'MYO18A', 'MICA', 'ERBB2IP', 'ZIC1', 'NELFE', 'MYO7A', 'RGS10', 'MUC13', 'KRAS', 'PSMD3', 'HAL', 'ICA1L', 'ZNF385D', 'CCNE2', 'ABO', 'ALG8', 'C7', 'RPL26L1', 'GDF7', 'ARID4B', 'TAF6', 'TAF10', 'AKT3', 'GNAQ', 'HMX1', 'RBX1', 'GNAI1', 'MYRF', 'UGT1A10', 'ERF', 'ABHD12', 'CAV1', 'DNAJC3', 'HDAC11', 'MAD2L2', 'WDPCP', 'IL1RN', 'PSME3', 'STXBP6', 'SFMBT1', 'GNA11', 'SNX3', 'TRPM3', 'TPST1', 'SYN3', 'TCOF1', 'ARR3', 'TCEB3CL', 'ACVR2A', 'IL1B', 'KCNH1', 'FBXW7', 'WAS', 'CHORDC1', 'OVOL1', 'WIPF1', 'RPS17', 'SDC4', 'LHB', 'TRAF2', 'NDUFS6', 'CFI', 'KSR2', 'CCDC136', 'CEMIP', 'CUBN', 'TTC8', 'TUBB4B', 'LCA5', 'TANK', 'ATF2', 'HIST1H2BC', 'GDF3', 'VCP', 'TNFRSF10A', 'HTRA1', 'PARP16', 'ALS2CR12', 'HGF', 'RAC2', 'RPLP2', 'SUDS3', 'TECTA', 'SDC3', 'MGAM', 'RD3', 'SMAD4', 'DUSP7', 'RPS25', 'STAM', 'ADAMTS5', 'NTF4', 'BACH1', 'ARMC4', 'TFAP2A', 'HGSNAT', 'DAD1', 'COL2A1', 'FGF18', 'TAF11', 'GYS2', 'CYSLTR1', 'PDZD3', 'TRIM10', 'RALY', 'CDKN1B', 'PGM1', 'ATP2A2', 'RPL39L', 'PRIMPOL', 'MED1', 'RANGAP1', 'LRIT3', 'CDC26', 'MAPK3', 'ADAM17', 'TAB2', 'MICALL2', 'SLC6A20', 'SPINK4', 'PPP3CB', 'SYT2', 'DERL2', 'FDFT1', 'FAM126B', 'TFDP1', 'BTK', 'CHRND', 'VPS13B', 'PAPD7', 'CDKN1C', 'SURF1', 'PTGDR', 'PTHLH', 'SRPX', 'KDM7A', 'MUC1', 'CYP11B2', 'FGB', 'NRG3', 'AAAS', 'WDFY4', 'HEYL', 'CRYBA2', 'RP2', 'OTUD7A', 'MEFV', 'KIF7', 'VAMP2', 'MYO16', 'GNB2', 'PMF1-BGLAP', 'SYNE3', 'IL1A', 'ITK', 'FZD6', 'NUP160', 'NIT2', 'MECP2', 'EBP', 'ANKRD11', 'C7orf72', 'EEFSEC', 'FGF20', 'RANBP2', 'CDC25C', 'CSK', 'DIRC3', 'RPL31', 'FOXO6', 'PRKACB', 'WNT7A', 'HHIP', 'RAB5B', 'OTX2', 'MBD3', 'TUBGCP5', 'POLR2E', 'GAP43', 'SFTA3', 'CRYBB2', 'FAM120B', 'ISG15', 'COX6B1', 'STAT6', 'PSORS1C1', 'POU2AF1', 'CHN1', 'GPHA2', 'TINF2', 'HIST1H2BN', 'ALG6', 'CRYBB3', 'ANKRD22', 'TTC37', 'CRH', 'TLR10', 'TSHZ3', 'HIST1H2BK', 'MAP2', 'C10orf2', 'FTL', 'ADAMTSL1', 'HLA-DRB1', 'BFSP1', 'CCNC', 'ZNF28', 'PIK3CB', 'PHYH', 'COL11A1', 'UQCRH', 'SLC25A31', 'NDUFAF7', 'HOXA11', 'MAN2A1', 'SLC25A17', 'DPEP3', 'AGRN', 'HIST1H2BJ', 'FARS2', 'TLR9', 'CDC37', 'RUNX1', 'NDUFS7', 'APOA1', 'MEDAG', 'ANAPC4', 'ERCC6', 'SFPQ', 'MARK4', 'MSH3', 'WNT5A', 'FANCA', 'RPS3A', 'ADAMTS12', 'SPRY3', 'MUSK', 'IFNAR1', 'SLC2A10', 'E2F2', 'MMAB', 'RPL10', 'RPL7A', 'NDUFA3', 'CD320', 'PLD3', 'SLC25A4', 'RECQL4', 'EIF2AK2', 'MYL2', 'TTC39C', 'MAP2K7', 'UQCRC2', 'NAALADL2', 'GPR176', 'ADCY5', 'ERLIN1', 'CPSF6', 'RDH10', 'CYP21A2', 'PIKFYVE', 'RPL3L', 'GRSF1', 'HOXD3', 'NR2E3', 'WIPF3', 'NEK1', 'MAMLD1', 'FANCC', 'THBS1', 'CHRNA3', 'PPP3CC', 'PSPH', 'ARRB1', 'NUP155', 'FZD2', 'WDR17', 'CNTRL', 'LPPR4', 'AP1M1', 'ADAMTS15', 'HARS', 'TUBB8', 'FOXO3', 'CNOT1', 'SLC39A5', 'SMOX', 'DFNB31', 'DMD', 'PMM2', 'RRP36', 'TPMT', 'EHMT2', 'ADAMTS10', 'TAMM41', 'ZBTB7B', 'HSPG2', 'VDR', 'DUSP9', 'MPDU1', 'TBX1', 'F12', 'H2AFY', 'CTSS', 'SFTPA1', 'LTBP1', 'ANGPT1', 'VAV1', 'ADRM1', 'COL4A2', 'GLS', 'BEST2', 'GCK', 'BMP4', 'SMAD3', 'COL4A3', 'FBLN5', 'ELOVL2', 'DLX6', 'GOLPH3L', 'PLXNA4', 'GSTM5', 'TUBA8', 'RANBP1', 'TEAD1', 'MAPK10', 'ADSS', 'MT-ND3', 'CXCR4', 'SLC4A3', 'CRTAP', 'EHBP1', 'OVOL2', 'PSMD9', 'SLC27A6', 'CHMP4B', 'KPNA2', 'OFD1', 'TNKS2', 'ARPC2', 'FAM161A', 'MKS1', 'SLC24A4', 'COG5', 'APC', 'DYNC1I2', 'HTR4', 'CTNND2', 'SPATA32', 'UTP23', 'TCEB3', 'NDUFB8', 'KREMEN2', 'BCL2L1', 'RPS12', 'GAL3ST2', 'RLN2', 'NOD2', 'GPR25', 'DP2', 'GNG3', 'MLST8', 'GAB2', 'RAP1B', 'ALDH1A2', 'PAH', 'CPA6', 'IL21R', 'AHCY', 'IFNG', 'POLR2J', 'RSPRY1', 'USP10', 'SLC3A2', 'FSHB', 'SAP30', 'CTTNBP2', 'PAK2', 'POLR2B', 'E2F1', 'EMR1', 'CAMKK1', 'RELA', 'ZEB1', 'AP2M1', 'SKAP1', 'PCGF1', 'PSMB1', 'SCAF4', 'VWF', 'GPR15', 'AKR1B1', 'CYFIP1', 'HTR6', 'ESYT2', 'SOX5', 'PHF20', 'THADA', 'NFIB', 'RPS10', 'ST3GAL1', 'CACNA1H', 'COX6A1', 'PFKFB3', 'SACS', 'FMOD', 'CDC20', 'STK32C', 'FAM126A', 'CTDP1', 'OARD1', 'GMDS', 'ZFHX3', 'CTNNA3', 'AVP', 'DEF8', 'PTEN', 'SLC45A2', 'CNKSR2', 'PTPN11', 'CEP78', 'DLGAP4', 'AVPR1B', 'MAGT1', 'TPM2', 'EXT1', 'DZANK1', 'PPP2R1A', 'RIMS1', 'SPRED3', 'PPP2R5B', 'GAS7', 'KRT3', 'CD163', 'QKI', 'GPC2', 'IFNGR1', 'RAE1', 'COX6A2', 'GRID2', 'DUSP16', 'PIK3C3', 'PRPF8', 'PPP3R1', 'RPS21', 'KCNJ8', 'TNFRSF8', 'BARD1', 'C6orf174', 'HLCS', 'COL1A2', 'CSNK2A1', 'JAK2', 'HES1', 'CLEC16A', 'IFT43', 'IGSF21', 'GALE', 'NUP214', 'PARD3', 'TUBB1', 'RNF150', 'PPP2R5A', 'BAX', 'NCAM2', 'MDGA2', 'SPATA5', 'TCEA1', 'GRM8', 'IQGAP1', 'EPS15', 'GRIN1', 'NDUFB9', 'NTN1', 'NDUFA10', 'HSPA6', 'MAB21L2', 'HIST1H2BO', 'ADRB1', 'CCRL2', 'SLC22A18', 'GCDH', 'SIX6', 'CPNE8', 'MYH7', 'RPL26', 'IGHV4-38-2', 'BBS5', 'TGFA', 'MYCN', 'OPCML', 'RPS27A', 'NPHP4', 'CYP2U1', 'SFTPA2', 'GHRHR', 'HIST2H3D', 'KIAA1279', 'FANCB', 'IFT52', 'ARL3', 'TNFAIP6', 'RPL15', 'BICD2', 'RGL1', 'PIGO', 'PYDC1', 'LPPR5', 'RPL4', 'NUP35', 'ANAPC2', 'HSPB4', 'PPP2R5E', 'INSR', 'TCN2', 'ZBTB20', 'TSPAN10', 'PTGES3', 'PSD'}\n",
      "\n",
      "The number of genes that involved in the PPI network:  3506\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <th>GSM509792_E1542N.CEL</th>\n",
       "      <th>GSM509793_E1546N.CEL</th>\n",
       "      <th>GSM509794_E1566N.CEL</th>\n",
       "      <th>GSM509795_E1584N.CEL</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM509811_E1566T.CEL</th>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <th>GSM509813_E1589T.CEL</th>\n",
       "      <th>GSM509814_E1603T.CEL</th>\n",
       "      <th>GSM509815_E1610T.CEL</th>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <th>GSM509817_E1635T.CEL</th>\n",
       "      <th>GSM509818_E1709T.CEL</th>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <th>GSM509820_E2644T.CEL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HSPA6</td>\n",
       "      <td>5.305487</td>\n",
       "      <td>5.608065</td>\n",
       "      <td>5.433042</td>\n",
       "      <td>5.556593</td>\n",
       "      <td>5.826820</td>\n",
       "      <td>5.715470</td>\n",
       "      <td>5.882963</td>\n",
       "      <td>4.937010</td>\n",
       "      <td>5.357001</td>\n",
       "      <td>...</td>\n",
       "      <td>5.997560</td>\n",
       "      <td>6.227560</td>\n",
       "      <td>7.082567</td>\n",
       "      <td>5.475410</td>\n",
       "      <td>6.327058</td>\n",
       "      <td>5.176684</td>\n",
       "      <td>6.929422</td>\n",
       "      <td>5.429512</td>\n",
       "      <td>5.950225</td>\n",
       "      <td>5.097301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>4.165863</td>\n",
       "      <td>4.085270</td>\n",
       "      <td>3.955792</td>\n",
       "      <td>4.104240</td>\n",
       "      <td>4.364118</td>\n",
       "      <td>4.092763</td>\n",
       "      <td>4.167443</td>\n",
       "      <td>4.097351</td>\n",
       "      <td>4.167891</td>\n",
       "      <td>...</td>\n",
       "      <td>4.241746</td>\n",
       "      <td>4.235995</td>\n",
       "      <td>4.182475</td>\n",
       "      <td>4.198627</td>\n",
       "      <td>3.964533</td>\n",
       "      <td>4.035028</td>\n",
       "      <td>4.121584</td>\n",
       "      <td>4.091635</td>\n",
       "      <td>4.136175</td>\n",
       "      <td>4.212867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCL5</td>\n",
       "      <td>7.191147</td>\n",
       "      <td>7.422020</td>\n",
       "      <td>7.854530</td>\n",
       "      <td>6.542158</td>\n",
       "      <td>7.561083</td>\n",
       "      <td>7.087193</td>\n",
       "      <td>7.176613</td>\n",
       "      <td>6.465548</td>\n",
       "      <td>6.708016</td>\n",
       "      <td>...</td>\n",
       "      <td>6.889683</td>\n",
       "      <td>7.939135</td>\n",
       "      <td>6.702354</td>\n",
       "      <td>6.604700</td>\n",
       "      <td>5.867164</td>\n",
       "      <td>6.923530</td>\n",
       "      <td>4.982753</td>\n",
       "      <td>5.449454</td>\n",
       "      <td>9.759467</td>\n",
       "      <td>8.277225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MMP14</td>\n",
       "      <td>6.839935</td>\n",
       "      <td>6.682461</td>\n",
       "      <td>6.629862</td>\n",
       "      <td>6.754795</td>\n",
       "      <td>6.874246</td>\n",
       "      <td>6.682161</td>\n",
       "      <td>6.987072</td>\n",
       "      <td>6.679945</td>\n",
       "      <td>6.839048</td>\n",
       "      <td>...</td>\n",
       "      <td>6.741775</td>\n",
       "      <td>6.766413</td>\n",
       "      <td>6.688555</td>\n",
       "      <td>6.763015</td>\n",
       "      <td>6.457795</td>\n",
       "      <td>6.672757</td>\n",
       "      <td>6.736853</td>\n",
       "      <td>6.946902</td>\n",
       "      <td>6.624468</td>\n",
       "      <td>6.802759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRADD</td>\n",
       "      <td>6.915792</td>\n",
       "      <td>6.779200</td>\n",
       "      <td>6.876954</td>\n",
       "      <td>7.094586</td>\n",
       "      <td>6.923319</td>\n",
       "      <td>6.763277</td>\n",
       "      <td>7.000070</td>\n",
       "      <td>6.933333</td>\n",
       "      <td>6.802136</td>\n",
       "      <td>...</td>\n",
       "      <td>6.731204</td>\n",
       "      <td>6.764289</td>\n",
       "      <td>7.222791</td>\n",
       "      <td>6.803186</td>\n",
       "      <td>6.329895</td>\n",
       "      <td>6.579684</td>\n",
       "      <td>6.801551</td>\n",
       "      <td>6.802755</td>\n",
       "      <td>6.855176</td>\n",
       "      <td>6.822479</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol  GSM509787_E1507N.CEL  GSM509788_E1520N.CEL  \\\n",
       "0       HSPA6              5.305487              5.608065   \n",
       "1      GUCA1A              4.165863              4.085270   \n",
       "2        CCL5              7.191147              7.422020   \n",
       "3       MMP14              6.839935              6.682461   \n",
       "4       TRADD              6.915792              6.779200   \n",
       "\n",
       "   GSM509789_E1521N.CEL  GSM509790_E1532N.CEL  GSM509791_E1535N.CEL  \\\n",
       "0              5.433042              5.556593              5.826820   \n",
       "1              3.955792              4.104240              4.364118   \n",
       "2              7.854530              6.542158              7.561083   \n",
       "3              6.629862              6.754795              6.874246   \n",
       "4              6.876954              7.094586              6.923319   \n",
       "\n",
       "   GSM509792_E1542N.CEL  GSM509793_E1546N.CEL  GSM509794_E1566N.CEL  \\\n",
       "0              5.715470              5.882963              4.937010   \n",
       "1              4.092763              4.167443              4.097351   \n",
       "2              7.087193              7.176613              6.465548   \n",
       "3              6.682161              6.987072              6.679945   \n",
       "4              6.763277              7.000070              6.933333   \n",
       "\n",
       "   GSM509795_E1584N.CEL  ...  GSM509811_E1566T.CEL  GSM509812_E1584T.CEL  \\\n",
       "0              5.357001  ...              5.997560              6.227560   \n",
       "1              4.167891  ...              4.241746              4.235995   \n",
       "2              6.708016  ...              6.889683              7.939135   \n",
       "3              6.839048  ...              6.741775              6.766413   \n",
       "4              6.802136  ...              6.731204              6.764289   \n",
       "\n",
       "   GSM509813_E1589T.CEL  GSM509814_E1603T.CEL  GSM509815_E1610T.CEL  \\\n",
       "0              7.082567              5.475410              6.327058   \n",
       "1              4.182475              4.198627              3.964533   \n",
       "2              6.702354              6.604700              5.867164   \n",
       "3              6.688555              6.763015              6.457795   \n",
       "4              7.222791              6.803186              6.329895   \n",
       "\n",
       "   GSM509816_E1614T.CEL  GSM509817_E1635T.CEL  GSM509818_E1709T.CEL  \\\n",
       "0              5.176684              6.929422              5.429512   \n",
       "1              4.035028              4.121584              4.091635   \n",
       "2              6.923530              4.982753              5.449454   \n",
       "3              6.672757              6.736853              6.946902   \n",
       "4              6.579684              6.801551              6.802755   \n",
       "\n",
       "   GSM509819_E1796T.CEL  GSM509820_E2644T.CEL  \n",
       "0              5.950225              5.097301  \n",
       "1              4.136175              4.212867  \n",
       "2              9.759467              8.277225  \n",
       "3              6.624468              6.802759  \n",
       "4              6.855176              6.822479  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sort the PPI Data by Gene in Alphabet Order\n",
    "preprocessed_gene_expression_data = preprocessed_PPI.sort_values(by='#node1')\n",
    "\n",
    "# Extract the Genes' Name from PPI Data\n",
    "# Get All the Genes that Occurred in the PPI Data\n",
    "ppi_genes = set(preprocessed_gene_expression_data['#node1']).union(set(preprocessed_gene_expression_data['node2']))\n",
    "print(\"The genes that involve in the PPI network: \" , ppi_genes)\n",
    "print(\"\\nThe number of genes that involved in the PPI network: \", len(ppi_genes))\n",
    "\n",
    "# Filter Gene Expression Data to Include Only Genes in PPI Data\n",
    "filter_gene_expression_data = average_gene_expression_data[average_gene_expression_data['Gene Symbol'].isin(ppi_genes)]\n",
    "filter_gene_expression_data = filter_gene_expression_data.reset_index(drop=True)\n",
    "\n",
    "# Display the First Five Rows of the Gene Expression Dataset\n",
    "filter_gene_expression_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*We already finished the process of preprocessing data.<br>\n",
    "The dataset we obtained are the data which the gene are occurred in Gene Expression Data and PPI Data.*\n",
    "<br><br>\n",
    "*Now, the dataset is ready to be used for the biclustering algorithm.*\n",
    "<br><br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_gene_expression_data.to_csv('D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/Filter Gene Expression Data.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Compare between the Involvement of PPI data**\n",
    "*This step is to identify is the involvement of PPI data able to achieve better result than just only using Gene Expression Dataset.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Given String of Sample Names that Sufferred from Cancer\n",
    "sample_names_str = \"GSM509804_E1507T.CEL GSM509805_E1520T.CEL GSM509806_E1521T.CEL GSM509807_E1532T.CEL GSM509808_E1535T.CEL GSM509809_E1542T.CEL GSM509810_E1546T.CEL GSM509811_E1566T.CEL GSM509812_E1584T.CEL GSM509813_E1589T.CEL GSM509814_E1603T.CEL GSM509815_E1610T.CEL GSM509816_E1614T.CEL GSM509817_E1635T.CEL GSM509818_E1709T.CEL GSM509819_E1796T.CEL GSM509820_E2644T.CEL\"\n",
    "\n",
    "# Split the String into Individual Sample Names\n",
    "sample_names = sample_names_str.split()\n",
    "\n",
    "# Convert the List of Sample Names to an Array\n",
    "sample_names_array = np.array(sample_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\876160023.py:6: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  average_gene_expression_data_ts['Target'] = pd.Series(pd.np.where(average_gene_expression_data_ts.index.isin(sample_names_array), 1, 0), index=average_gene_expression_data_ts.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>DDR1 /// MIR4640</th>\n",
       "      <th>RFC2</th>\n",
       "      <th>HSPA6</th>\n",
       "      <th>PAX8</th>\n",
       "      <th>GUCA1A</th>\n",
       "      <th>MIR5193 /// UBA7</th>\n",
       "      <th>THRA</th>\n",
       "      <th>PTPN21</th>\n",
       "      <th>CCL5</th>\n",
       "      <th>CYP2E1</th>\n",
       "      <th>...</th>\n",
       "      <th>RBM12B</th>\n",
       "      <th>LIME1 /// ZGPAT</th>\n",
       "      <th>LOC102725263 /// LOC389906</th>\n",
       "      <th>C1orf50 /// LOC100129924</th>\n",
       "      <th>FAM86B1 /// FAM86B2 /// FAM86C1 /// FAM86DP /// FAM86FP</th>\n",
       "      <th>SNHG17</th>\n",
       "      <th>HNRNPUL2 /// HNRNPUL2-BSCL2</th>\n",
       "      <th>LOC100505915</th>\n",
       "      <th>NPEPL1</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>9.811079</td>\n",
       "      <td>7.010666</td>\n",
       "      <td>5.305487</td>\n",
       "      <td>5.508012</td>\n",
       "      <td>4.165863</td>\n",
       "      <td>7.693826</td>\n",
       "      <td>5.408280</td>\n",
       "      <td>4.835314</td>\n",
       "      <td>7.191147</td>\n",
       "      <td>6.658615</td>\n",
       "      <td>...</td>\n",
       "      <td>5.802191</td>\n",
       "      <td>6.998560</td>\n",
       "      <td>5.327010</td>\n",
       "      <td>6.905935</td>\n",
       "      <td>6.425221</td>\n",
       "      <td>6.760667</td>\n",
       "      <td>3.977508</td>\n",
       "      <td>5.091434</td>\n",
       "      <td>7.231490</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>9.587092</td>\n",
       "      <td>6.710686</td>\n",
       "      <td>5.608065</td>\n",
       "      <td>5.616255</td>\n",
       "      <td>4.085270</td>\n",
       "      <td>7.462041</td>\n",
       "      <td>5.501202</td>\n",
       "      <td>4.986851</td>\n",
       "      <td>7.422020</td>\n",
       "      <td>5.766358</td>\n",
       "      <td>...</td>\n",
       "      <td>6.196255</td>\n",
       "      <td>6.767065</td>\n",
       "      <td>6.038634</td>\n",
       "      <td>6.843098</td>\n",
       "      <td>6.604427</td>\n",
       "      <td>6.366031</td>\n",
       "      <td>4.177136</td>\n",
       "      <td>4.993988</td>\n",
       "      <td>7.194917</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>9.363137</td>\n",
       "      <td>6.899260</td>\n",
       "      <td>5.433042</td>\n",
       "      <td>5.637902</td>\n",
       "      <td>3.955792</td>\n",
       "      <td>7.261501</td>\n",
       "      <td>5.369323</td>\n",
       "      <td>4.930880</td>\n",
       "      <td>7.854530</td>\n",
       "      <td>6.516132</td>\n",
       "      <td>...</td>\n",
       "      <td>5.643841</td>\n",
       "      <td>7.064645</td>\n",
       "      <td>5.525738</td>\n",
       "      <td>6.808874</td>\n",
       "      <td>6.774870</td>\n",
       "      <td>6.876295</td>\n",
       "      <td>3.862921</td>\n",
       "      <td>5.122977</td>\n",
       "      <td>7.138887</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>9.632889</td>\n",
       "      <td>6.776456</td>\n",
       "      <td>5.556593</td>\n",
       "      <td>5.639477</td>\n",
       "      <td>4.104240</td>\n",
       "      <td>7.509573</td>\n",
       "      <td>5.390338</td>\n",
       "      <td>5.138890</td>\n",
       "      <td>6.542158</td>\n",
       "      <td>7.329259</td>\n",
       "      <td>...</td>\n",
       "      <td>5.990205</td>\n",
       "      <td>7.300676</td>\n",
       "      <td>5.200480</td>\n",
       "      <td>7.009491</td>\n",
       "      <td>6.521508</td>\n",
       "      <td>6.453466</td>\n",
       "      <td>4.276351</td>\n",
       "      <td>4.851322</td>\n",
       "      <td>7.339033</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>9.381932</td>\n",
       "      <td>6.565620</td>\n",
       "      <td>5.826820</td>\n",
       "      <td>5.629071</td>\n",
       "      <td>4.364118</td>\n",
       "      <td>7.438766</td>\n",
       "      <td>5.515332</td>\n",
       "      <td>4.800900</td>\n",
       "      <td>7.561083</td>\n",
       "      <td>6.212731</td>\n",
       "      <td>...</td>\n",
       "      <td>6.401825</td>\n",
       "      <td>7.181522</td>\n",
       "      <td>5.380979</td>\n",
       "      <td>6.576949</td>\n",
       "      <td>6.163176</td>\n",
       "      <td>6.326391</td>\n",
       "      <td>4.258736</td>\n",
       "      <td>5.552102</td>\n",
       "      <td>7.223029</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 13516 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol           DDR1 /// MIR4640      RFC2     HSPA6      PAX8  \\\n",
       "GSM509787_E1507N.CEL          9.811079  7.010666  5.305487  5.508012   \n",
       "GSM509788_E1520N.CEL          9.587092  6.710686  5.608065  5.616255   \n",
       "GSM509789_E1521N.CEL          9.363137  6.899260  5.433042  5.637902   \n",
       "GSM509790_E1532N.CEL          9.632889  6.776456  5.556593  5.639477   \n",
       "GSM509791_E1535N.CEL          9.381932  6.565620  5.826820  5.629071   \n",
       "\n",
       "Gene Symbol             GUCA1A  MIR5193 /// UBA7      THRA    PTPN21  \\\n",
       "GSM509787_E1507N.CEL  4.165863          7.693826  5.408280  4.835314   \n",
       "GSM509788_E1520N.CEL  4.085270          7.462041  5.501202  4.986851   \n",
       "GSM509789_E1521N.CEL  3.955792          7.261501  5.369323  4.930880   \n",
       "GSM509790_E1532N.CEL  4.104240          7.509573  5.390338  5.138890   \n",
       "GSM509791_E1535N.CEL  4.364118          7.438766  5.515332  4.800900   \n",
       "\n",
       "Gene Symbol               CCL5    CYP2E1  ...    RBM12B  LIME1 /// ZGPAT  \\\n",
       "GSM509787_E1507N.CEL  7.191147  6.658615  ...  5.802191         6.998560   \n",
       "GSM509788_E1520N.CEL  7.422020  5.766358  ...  6.196255         6.767065   \n",
       "GSM509789_E1521N.CEL  7.854530  6.516132  ...  5.643841         7.064645   \n",
       "GSM509790_E1532N.CEL  6.542158  7.329259  ...  5.990205         7.300676   \n",
       "GSM509791_E1535N.CEL  7.561083  6.212731  ...  6.401825         7.181522   \n",
       "\n",
       "Gene Symbol           LOC102725263 /// LOC389906  C1orf50 /// LOC100129924  \\\n",
       "GSM509787_E1507N.CEL                    5.327010                  6.905935   \n",
       "GSM509788_E1520N.CEL                    6.038634                  6.843098   \n",
       "GSM509789_E1521N.CEL                    5.525738                  6.808874   \n",
       "GSM509790_E1532N.CEL                    5.200480                  7.009491   \n",
       "GSM509791_E1535N.CEL                    5.380979                  6.576949   \n",
       "\n",
       "Gene Symbol           FAM86B1 /// FAM86B2 /// FAM86C1 /// FAM86DP /// FAM86FP  \\\n",
       "GSM509787_E1507N.CEL                                           6.425221         \n",
       "GSM509788_E1520N.CEL                                           6.604427         \n",
       "GSM509789_E1521N.CEL                                           6.774870         \n",
       "GSM509790_E1532N.CEL                                           6.521508         \n",
       "GSM509791_E1535N.CEL                                           6.163176         \n",
       "\n",
       "Gene Symbol             SNHG17  HNRNPUL2 /// HNRNPUL2-BSCL2  LOC100505915  \\\n",
       "GSM509787_E1507N.CEL  6.760667                     3.977508      5.091434   \n",
       "GSM509788_E1520N.CEL  6.366031                     4.177136      4.993988   \n",
       "GSM509789_E1521N.CEL  6.876295                     3.862921      5.122977   \n",
       "GSM509790_E1532N.CEL  6.453466                     4.276351      4.851322   \n",
       "GSM509791_E1535N.CEL  6.326391                     4.258736      5.552102   \n",
       "\n",
       "Gene Symbol             NPEPL1  Target  \n",
       "GSM509787_E1507N.CEL  7.231490       0  \n",
       "GSM509788_E1520N.CEL  7.194917       0  \n",
       "GSM509789_E1521N.CEL  7.138887       0  \n",
       "GSM509790_E1532N.CEL  7.339033       0  \n",
       "GSM509791_E1535N.CEL  7.223029       0  \n",
       "\n",
       "[5 rows x 13516 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only Involved Gene Expression Dataset\n",
    "\n",
    "average_gene_expression_data_ts = average_gene_expression_data.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "average_gene_expression_data_ts['Target'] = pd.Series(pd.np.where(average_gene_expression_data_ts.index.isin(sample_names_array), 1, 0), index=average_gene_expression_data_ts.index)\n",
    "\n",
    "average_gene_expression_data_ts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test size: 0.1, Accuracy: 1.0\n",
      "test size: 0.2, Accuracy: 0.8571428571428571\n",
      "test size: 0.3, Accuracy: 1.0\n",
      "test size: 0.4, Accuracy: 0.9285714285714286\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import KFold, cross_val_score, cross_val_predict\n",
    "from sklearn import svm\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import confusion_matrix, roc_curve, auc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Select the first 285 columns as features (X)\n",
    "X = average_gene_expression_data_ts.iloc[:, :-1]\n",
    "\n",
    "# Select the last column as the target variable (y)\n",
    "y = average_gene_expression_data_ts.iloc[:, -1]\n",
    "\n",
    "ts = [0.1, 0.2, 0.3, 0.4]\n",
    "\n",
    "for ts in ts:\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=ts, stratify=y)\n",
    "    clf = svm.SVC(kernel='linear', probability=True)\n",
    "    clf.fit(X_train, y_train)\n",
    "    y_pred = clf.predict(X_test)\n",
    "    accuracy = np.mean(y_pred == y_test)\n",
    "    print(f\"test size: {ts}, Accuracy: {accuracy}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\2741390801.py:6: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  filter_gene_expression_data_ts['Target'] = pd.Series(pd.np.where(filter_gene_expression_data_ts.index.isin(sample_names_array), 1, 0), index=filter_gene_expression_data_ts.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>HSPA6</th>\n",
       "      <th>GUCA1A</th>\n",
       "      <th>CCL5</th>\n",
       "      <th>MMP14</th>\n",
       "      <th>TRADD</th>\n",
       "      <th>PLD1</th>\n",
       "      <th>BAD</th>\n",
       "      <th>PRPF8</th>\n",
       "      <th>CAPNS1</th>\n",
       "      <th>RPL35</th>\n",
       "      <th>...</th>\n",
       "      <th>TBC1D2</th>\n",
       "      <th>B4GALT7</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>NACA2</th>\n",
       "      <th>TRDN</th>\n",
       "      <th>SCAF4</th>\n",
       "      <th>LAMA1</th>\n",
       "      <th>FBXO31</th>\n",
       "      <th>SLC44A1</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>5.305487</td>\n",
       "      <td>4.165863</td>\n",
       "      <td>7.191147</td>\n",
       "      <td>6.839935</td>\n",
       "      <td>6.915792</td>\n",
       "      <td>6.905594</td>\n",
       "      <td>6.924447</td>\n",
       "      <td>9.266126</td>\n",
       "      <td>11.424412</td>\n",
       "      <td>12.999451</td>\n",
       "      <td>...</td>\n",
       "      <td>7.765013</td>\n",
       "      <td>5.949763</td>\n",
       "      <td>7.054392</td>\n",
       "      <td>3.451194</td>\n",
       "      <td>3.389328</td>\n",
       "      <td>6.060083</td>\n",
       "      <td>4.591470</td>\n",
       "      <td>6.167321</td>\n",
       "      <td>6.157059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>5.608065</td>\n",
       "      <td>4.085270</td>\n",
       "      <td>7.422020</td>\n",
       "      <td>6.682461</td>\n",
       "      <td>6.779200</td>\n",
       "      <td>6.819023</td>\n",
       "      <td>7.031198</td>\n",
       "      <td>9.223339</td>\n",
       "      <td>11.364217</td>\n",
       "      <td>12.781515</td>\n",
       "      <td>...</td>\n",
       "      <td>7.297379</td>\n",
       "      <td>5.828337</td>\n",
       "      <td>6.838957</td>\n",
       "      <td>3.677271</td>\n",
       "      <td>3.281392</td>\n",
       "      <td>5.524395</td>\n",
       "      <td>4.637121</td>\n",
       "      <td>6.635602</td>\n",
       "      <td>6.376391</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>5.433042</td>\n",
       "      <td>3.955792</td>\n",
       "      <td>7.854530</td>\n",
       "      <td>6.629862</td>\n",
       "      <td>6.876954</td>\n",
       "      <td>6.570245</td>\n",
       "      <td>6.699416</td>\n",
       "      <td>9.333200</td>\n",
       "      <td>10.941655</td>\n",
       "      <td>12.811271</td>\n",
       "      <td>...</td>\n",
       "      <td>7.498661</td>\n",
       "      <td>5.811177</td>\n",
       "      <td>7.229369</td>\n",
       "      <td>3.575415</td>\n",
       "      <td>3.498269</td>\n",
       "      <td>6.353812</td>\n",
       "      <td>4.658779</td>\n",
       "      <td>6.359670</td>\n",
       "      <td>5.889888</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>5.556593</td>\n",
       "      <td>4.104240</td>\n",
       "      <td>6.542158</td>\n",
       "      <td>6.754795</td>\n",
       "      <td>7.094586</td>\n",
       "      <td>7.091794</td>\n",
       "      <td>6.703973</td>\n",
       "      <td>9.285872</td>\n",
       "      <td>11.279280</td>\n",
       "      <td>12.830638</td>\n",
       "      <td>...</td>\n",
       "      <td>7.672872</td>\n",
       "      <td>5.988737</td>\n",
       "      <td>7.151789</td>\n",
       "      <td>3.647739</td>\n",
       "      <td>3.266002</td>\n",
       "      <td>6.034053</td>\n",
       "      <td>4.482420</td>\n",
       "      <td>6.408367</td>\n",
       "      <td>6.167377</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>5.826820</td>\n",
       "      <td>4.364118</td>\n",
       "      <td>7.561083</td>\n",
       "      <td>6.874246</td>\n",
       "      <td>6.923319</td>\n",
       "      <td>6.670091</td>\n",
       "      <td>6.198851</td>\n",
       "      <td>8.994933</td>\n",
       "      <td>11.086393</td>\n",
       "      <td>12.665346</td>\n",
       "      <td>...</td>\n",
       "      <td>7.479567</td>\n",
       "      <td>5.652972</td>\n",
       "      <td>6.965022</td>\n",
       "      <td>3.517448</td>\n",
       "      <td>3.465066</td>\n",
       "      <td>6.884892</td>\n",
       "      <td>4.618102</td>\n",
       "      <td>6.108703</td>\n",
       "      <td>6.840216</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2736 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              HSPA6    GUCA1A      CCL5     MMP14     TRADD  \\\n",
       "GSM509787_E1507N.CEL  5.305487  4.165863  7.191147  6.839935  6.915792   \n",
       "GSM509788_E1520N.CEL  5.608065  4.085270  7.422020  6.682461  6.779200   \n",
       "GSM509789_E1521N.CEL  5.433042  3.955792  7.854530  6.629862  6.876954   \n",
       "GSM509790_E1532N.CEL  5.556593  4.104240  6.542158  6.754795  7.094586   \n",
       "GSM509791_E1535N.CEL  5.826820  4.364118  7.561083  6.874246  6.923319   \n",
       "\n",
       "Gene Symbol               PLD1       BAD     PRPF8     CAPNS1      RPL35  ...  \\\n",
       "GSM509787_E1507N.CEL  6.905594  6.924447  9.266126  11.424412  12.999451  ...   \n",
       "GSM509788_E1520N.CEL  6.819023  7.031198  9.223339  11.364217  12.781515  ...   \n",
       "GSM509789_E1521N.CEL  6.570245  6.699416  9.333200  10.941655  12.811271  ...   \n",
       "GSM509790_E1532N.CEL  7.091794  6.703973  9.285872  11.279280  12.830638  ...   \n",
       "GSM509791_E1535N.CEL  6.670091  6.198851  8.994933  11.086393  12.665346  ...   \n",
       "\n",
       "Gene Symbol             TBC1D2   B4GALT7  CASP8AP2     NACA2      TRDN  \\\n",
       "GSM509787_E1507N.CEL  7.765013  5.949763  7.054392  3.451194  3.389328   \n",
       "GSM509788_E1520N.CEL  7.297379  5.828337  6.838957  3.677271  3.281392   \n",
       "GSM509789_E1521N.CEL  7.498661  5.811177  7.229369  3.575415  3.498269   \n",
       "GSM509790_E1532N.CEL  7.672872  5.988737  7.151789  3.647739  3.266002   \n",
       "GSM509791_E1535N.CEL  7.479567  5.652972  6.965022  3.517448  3.465066   \n",
       "\n",
       "Gene Symbol              SCAF4     LAMA1    FBXO31   SLC44A1  Target  \n",
       "GSM509787_E1507N.CEL  6.060083  4.591470  6.167321  6.157059       0  \n",
       "GSM509788_E1520N.CEL  5.524395  4.637121  6.635602  6.376391       0  \n",
       "GSM509789_E1521N.CEL  6.353812  4.658779  6.359670  5.889888       0  \n",
       "GSM509790_E1532N.CEL  6.034053  4.482420  6.408367  6.167377       0  \n",
       "GSM509791_E1535N.CEL  6.884892  4.618102  6.108703  6.840216       0  \n",
       "\n",
       "[5 rows x 2736 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Involved Gene Expression Dataset and PPI data\n",
    "\n",
    "filter_gene_expression_data_ts = filter_gene_expression_data.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "filter_gene_expression_data_ts['Target'] = pd.Series(pd.np.where(filter_gene_expression_data_ts.index.isin(sample_names_array), 1, 0), index=filter_gene_expression_data_ts.index)\n",
    "\n",
    "filter_gene_expression_data_ts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test size: 0.1, Accuracy: 1.0\n",
      "test size: 0.2, Accuracy: 1.0\n",
      "test size: 0.3, Accuracy: 0.9090909090909091\n",
      "test size: 0.4, Accuracy: 0.9285714285714286\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import KFold, cross_val_score, cross_val_predict\n",
    "from sklearn import svm\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import confusion_matrix, roc_curve, auc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Select the first 285 columns as features (X)\n",
    "X = filter_gene_expression_data_ts.iloc[:, :-1]\n",
    "\n",
    "# Select the last column as the target variable (y)\n",
    "y = filter_gene_expression_data_ts.iloc[:, -1]\n",
    "\n",
    "ts = [0.1, 0.2, 0.3, 0.4]\n",
    "\n",
    "for ts in ts:\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=ts, stratify=y)\n",
    "    clf = svm.SVC(kernel='linear', probability=True)\n",
    "    clf.fit(X_train, y_train)\n",
    "    y_pred = clf.predict(X_test)\n",
    "    accuracy = np.mean(y_pred == y_test)\n",
    "    print(f\"test size: {ts}, Accuracy: {accuracy}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test size: 0.1\n",
      "  Random State: 5, Accuracy: 1.0\n",
      "  Random State: 10, Accuracy: 0.75\n",
      "  Random State: 15, Accuracy: 0.75\n",
      "  Random State: 20, Accuracy: 1.0\n",
      "  Random State: 25, Accuracy: 0.75\n",
      "  Random State: 30, Accuracy: 1.0\n",
      "  Random State: 35, Accuracy: 1.0\n",
      "  Random State: 40, Accuracy: 1.0\n",
      "  Random State: 45, Accuracy: 1.0\n",
      "  Random State: 50, Accuracy: 0.75\n",
      "Mean Value of Accuracy:  0.9\n",
      "\n",
      "\n",
      "Test size: 0.2\n",
      "  Random State: 5, Accuracy: 1.0\n",
      "  Random State: 10, Accuracy: 0.8571428571428571\n",
      "  Random State: 15, Accuracy: 0.8571428571428571\n",
      "  Random State: 20, Accuracy: 1.0\n",
      "  Random State: 25, Accuracy: 1.0\n",
      "  Random State: 30, Accuracy: 0.8571428571428571\n",
      "  Random State: 35, Accuracy: 1.0\n",
      "  Random State: 40, Accuracy: 0.8571428571428571\n",
      "  Random State: 45, Accuracy: 1.0\n",
      "  Random State: 50, Accuracy: 1.0\n",
      "Mean Value of Accuracy:  0.9428571428571428\n",
      "\n",
      "\n",
      "Test size: 0.3\n",
      "  Random State: 5, Accuracy: 1.0\n",
      "  Random State: 10, Accuracy: 0.9090909090909091\n",
      "  Random State: 15, Accuracy: 0.9090909090909091\n",
      "  Random State: 20, Accuracy: 1.0\n",
      "  Random State: 25, Accuracy: 1.0\n",
      "  Random State: 30, Accuracy: 0.9090909090909091\n",
      "  Random State: 35, Accuracy: 1.0\n",
      "  Random State: 40, Accuracy: 0.9090909090909091\n",
      "  Random State: 45, Accuracy: 1.0\n",
      "  Random State: 50, Accuracy: 1.0\n",
      "Mean Value of Accuracy:  0.9636363636363636\n",
      "\n",
      "\n",
      "Test size: 0.4\n",
      "  Random State: 5, Accuracy: 1.0\n",
      "  Random State: 10, Accuracy: 0.9285714285714286\n",
      "  Random State: 15, Accuracy: 0.9285714285714286\n",
      "  Random State: 20, Accuracy: 1.0\n",
      "  Random State: 25, Accuracy: 1.0\n",
      "  Random State: 30, Accuracy: 0.9285714285714286\n",
      "  Random State: 35, Accuracy: 1.0\n",
      "  Random State: 40, Accuracy: 0.9285714285714286\n",
      "  Random State: 45, Accuracy: 1.0\n",
      "  Random State: 50, Accuracy: 0.9285714285714286\n",
      "Mean Value of Accuracy:  0.9642857142857144\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import svm\n",
    "from sklearn.metrics import confusion_matrix, roc_curve, auc\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# Select the first 285 columns as features (X)\n",
    "X = filter_gene_expression_data_ts.iloc[:, :-1]\n",
    "\n",
    "# Select the last column as the target variable (y)\n",
    "y = filter_gene_expression_data_ts.iloc[:, -1]\n",
    "\n",
    "ts_list = [0.1, 0.2, 0.3, 0.4]\n",
    "rs_list = [5, 10, 15, 20, 25, 30, 35, 40, 45, 50]\n",
    "\n",
    "for ts in ts_list:\n",
    "    score = []\n",
    "    print(f\"Test size: {ts}\")\n",
    "    for rs in rs_list:\n",
    "        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=ts, random_state=rs, stratify=y)\n",
    "        clf = svm.SVC(kernel='linear', probability=True)\n",
    "        clf.fit(X_train, y_train)\n",
    "        y_pred = clf.predict(X_test)\n",
    "        accuracy = np.mean(y_pred == y_test)\n",
    "        score.append(accuracy)\n",
    "        print(f\"  Random State: {rs}, Accuracy: {accuracy}\")\n",
    "    print(\"Mean Value of Accuracy: \", np.mean(score))\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Elbow Method**\n",
    "*The performance of plaid model will be vary when different parameters is used. <br> \n",
    "To identify the optimal number of clusters, Elbow Method had been applied.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlUAAAGwCAYAAACAZ5AeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAABI7klEQVR4nO3deXhU5cH+8XsyyWQjK0hCSAhBdghrAKOiVVNSX2qlWhdealGgtRoUxFqx/SlUqyzWVlEUl7dKrRu2xQUESgGhKrIEg2GLLAHCEhazEsg28/z+CBkZQUnIwJlJvp/rmgvmnGfO3CcIc3vOM+fYjDFGAAAAaJIAqwMAAAA0B5QqAAAAL6BUAQAAeAGlCgAAwAsoVQAAAF5AqQIAAPACShUAAIAXBFodoLlwuVw6cOCAIiIiZLPZrI4DAAAawBij8vJyJSQkKCCgaceaKFVecuDAASUlJVkdAwAAnIOCggIlJiY2aRuUKi+JiIiQVPeHEhkZaXEaAADQEGVlZUpKSnJ/jjcFpcpL6k/5RUZGUqoAAPAz3pi6w0R1AAAAL6BUAQAAeAGlCgAAwAsoVQAAAF5AqQIAAPACShUAAIAXUKoAAAC8gFIFAADgBZQqAAAAL6BUAQAAeAGlCgAAwAsoVQAAAF5AqfJhNU6XcgpKVF3rsjoKAAA4C0qVD7v6qY81Yvan+nJfidVRAADAWVCqfFjPdpGSpLW7iyxOAgAAzoZS5cMGdYyVJK3Lp1QBAODrKFU+bHBKXalav6dYTpexOA0AAPg+lCof1rNdpMIddpVX1iqvsNzqOAAA4HtQqnxYoD1AA5JjJEnr93AKEAAAX0ap8nH186rWMq8KAACfRqnyce7J6ruLZAzzqgAA8FWUKh/Xv0O0guw2HSqrUkHRCavjAACA70Cp8nEhQXalto+SxPWqAADwZZQqPzAohetVAQDg6yhVfmDwKfOqAACAb6JU+YG05FjZbNKuoxU6Ul5ldRwAAHAGlCo/EBUWpG5xEZKk9RytAgDAJ1Gq/IT7elWUKgAAfBKlyk+4J6tTqgAA8EmUKj9RP1l9y4EylVfWWJwGAAB8G6XKT8RHhSgpNlQuI23YW2J1HAAA8C2UKj/ivmUN16sCAMDnUKr8yGAmqwMA4LMoVX6kfrJ6TkGJqmqdFqcBAACnolT5kU5twtU63KHqWpdy95VaHQcAAJyCUuVHbDab0jrGSOIUIAAAvoZS5WeYrA4AgG+iVPmZwSfnVa3fUyyny1icBgAA1KNU+Zme7SIV7rCrvLJWeYXlVscBAAAnUar8TKA9QAOS6+ZVccsaAAB8B6XKD7nnVVGqAADwGZQqP3RqqTKGeVUAAPgCSpUf6t8hWkF2mw6VVamg6ITVcQAAgChVfikkyK7U9lGSuF4VAAC+glLlp+pvWcP1qgAA8A2UKj81mMnqAAD4FEqVn0pLjpXNJu06WqEj5VVWxwEAoMWjVPmpqLAgdYuLkCSt52gVAACWo1T5sfpLKzBZHQAA61Gq/Jh7sjqlCgAAy1Gq/Fj9ZPUtB8pUXlljcRoAAFo2SpUfi48KUVJsqFxG2rC3xOo4AAC0aJQqP+e+ZQ3XqwIAwFKUKj83mMnqAAD4BEqVn6ufrJ5TUKKqWqfFaQAAaLkoVX6uU5twtWnlUHWtS7n7Sq2OAwBAi0Wp8nM2m01pyZwCBADAapSqZoCbKwMAYD1KVTNQP1l9/Z5iOV3G4jQAALRMlKpmoEe7CIU77CqvrFVeYbnVcQAAaJEoVc1AoD1AA5JjJHHLGgAArGJpqXI6nXr44YeVkpKi0NBQXXzxxXrsscdkzDensIwxeuSRR9SuXTuFhoYqIyND27dv99hOUVGRRo0apcjISEVHR2vs2LE6duyYx5gvv/xSQ4cOVUhIiJKSkjRz5szT8rz77rvq3r27QkJClJqaqo8++uj87Ph5wM2VAQCwlqWlasaMGXrhhRf03HPPaevWrZoxY4ZmzpypZ5991j1m5syZmjVrlubMmaM1a9YoPDxcmZmZqqysdI8ZNWqUNm/erKVLl2rBggVatWqVfvWrX7nXl5WVadiwYUpOTlZ2draefPJJTZ06VS+99JJ7zGeffaaRI0dq7Nix+uKLLzRixAiNGDFCmzZtujA/jCY69crqp5ZSAABwgRgLDR8+3IwZM8Zj2Q033GBGjRpljDHG5XKZ+Ph48+STT7rXl5SUmODgYPPWW28ZY4zZsmWLkWTWrVvnHrNo0SJjs9nM/v37jTHGPP/88yYmJsZUVVW5xzz44IOmW7du7uc333yzGT58uEeWIUOGmDvvvLNB+1JaWmokmdLS0gaN97YT1bWm8+8WmuQHF5g9RyssyQAAgL/x5ue3pUeqLr30Ui1btkxfffWVJGnjxo365JNPdO2110qS8vPzVVhYqIyMDPdroqKiNGTIEK1evVqStHr1akVHRystLc09JiMjQwEBAVqzZo17zBVXXCGHw+Eek5mZqby8PBUXF7vHnPo+9WPq3+fbqqqqVFZW5vGwUkiQXantoyRxChAAACtYWqomT56sW2+9Vd27d1dQUJD69++viRMnatSoUZKkwsJCSVJcXJzH6+Li4tzrCgsL1bZtW4/1gYGBio2N9Rhzpm2c+h7fNaZ+/bdNmzZNUVFR7kdSUlKj99/buF4VAADWsbRUzZs3T2+88YbefPNNbdiwQXPnztWf/vQnzZ0718pYDfLQQw+ptLTU/SgoKLA6kvt6VXwDEACACy/Qyjd/4IEH3EerJCk1NVV79uzRtGnTNHr0aMXHx0uSDh06pHbt2rlfd+jQIfXr10+SFB8fr8OHD3tst7a2VkVFRe7Xx8fH69ChQx5j6p+fbUz9+m8LDg5WcHDwuez2eZOWHCubTdp1tEJHyqt0UYRv5QMAoDmz9EjV8ePHFRDgGcFut8vlckmSUlJSFB8fr2XLlrnXl5WVac2aNUpPT5ckpaenq6SkRNnZ2e4xy5cvl8vl0pAhQ9xjVq1apZqaGveYpUuXqlu3boqJiXGPOfV96sfUv48/iAoLUre4CEnSeo5WAQBwQVlaqq677jo9/vjjWrhwoXbv3q358+frz3/+s376059KqrtZ8MSJE/XHP/5RH3zwgXJzc/WLX/xCCQkJGjFihCSpR48e+tGPfqRf/vKXWrt2rT799FONHz9et956qxISEiRJ//u//yuHw6GxY8dq8+bNeuedd/TMM89o0qRJ7iwTJkzQ4sWL9dRTT2nbtm2aOnWq1q9fr/Hjx1/wn0tTcL0qAAAs4oVvI56zsrIyM2HCBNOhQwcTEhJiOnXqZH7/+997XPrA5XKZhx9+2MTFxZng4GBzzTXXmLy8PI/tfP3112bkyJGmVatWJjIy0txxxx2mvLzcY8zGjRvN5ZdfboKDg0379u3N9OnTT8szb94807VrV+NwOEyvXr3MwoULG7wvVl9Sod77OftN8oMLzPBZqyzNAQCAP/Dm57fNGK4U6Q1lZWWKiopSaWmpIiMjLctRWFqpS6YtU4BN2jhlmCJCgizLAgCAr/Pm5zf3/mtm4qNClBQbKpeRNuwtsToOAAAtBqWqGTr1ljUAAODCoFQ1Q4OZrA4AwAVHqWqG6q+snlNQoqpap8VpAABoGShVzVCnNuFq08qh6lqXcveVWh0HAIAWgVLVDNlsNqUlcwoQAIALiVLVTHFzZQAALixKVTNVP1l9/Z5iOV1cigwAgPONUtVM9WgXoXCHXeWVtcorLLc6DgAAzR6lqpkKtAdoQHLdzaLXMa8KAIDzjlLVjHFzZQAALhxKVTN26pXVucUjAADnF6WqGevfIVpBdpsOl1dpb9Fxq+MAANCsUaqasZAgu1LbR0mS1u0utjgNAADNG6WqmeN6VQAAXBiUqmau/npVfAMQAIDzi1LVzKUlx8pmk3YdrdCR8iqr4wAA0GxRqpq5qLAgdYuLkCSt52gVAADnDaWqBeB6VQAAnH+UqhbAPVmdUgUAwHlDqWoB6ierbzlQpvLKGovTAADQPFGqWoD4qBAlxYbKZaQNe0usjgMAQLNEqWohTr1lDQAA8D5KVQsxmMnqAACcV5SqFqJ+snpOQYmqap0WpwEAoPmhVLUQndqEq00rh6prXcrdV2p1HAAAmh1KVQths9mUlswpQAAAzhdKVQvCzZUBADh/KFUtSP1k9fV7iuV0GYvTAADQvFCqWpAe7SIU7rCrvLJWeYXlVscBAKBZoVS1IIH2AA1IjpHELWsAAPA2SlULw/WqAAA4PyhVLcypk9WNYV4VAADeQqlqYfolRSvIbtPh8irtLTpudRwAAJoNSlULExJkV2r7KEnSWi6tAACA11CqWiD3KUDmVQEA4DWUqhbIfb2q3cUWJwEAoPmgVLVAacmxstmkXUcrdKS8yuo4AAA0C5SqFigqLEjd4iIkSes5BQgAgFdQqlqoQVyvCgAAr6JUtVBMVgcAwLsoVS1U/WT1LQfKVF5ZY3EaAAD8H6WqhYqPClFSbKhcRtqwt8TqOAAA+D1KVQtWP69qHRcBBQCgyShVLRg3VwYAwHsoVS1Y/WT1nIISVdU6LU4DAIB/o1S1YJ3ahKtNK4eqa13K3VdqdRwAAPwapaoFs9lsSkvmFCAAAN5AqWrh3NerYrI6AABNQqlq4dw3V95TLKfLWJwGAAD/Ralq4Xq0i1C4w67yylrlFZZbHQcAAL9FqWrhAu0BGpAcI4lb1gAA0BSUKnC9KgAAvIBSBY/J6sYwrwoAgHNBqYL6JUUryG7T4fIq7S06bnUcAAD8EqUKCgmyq09itCRpLZdWAADgnFCqIElK68hkdQAAmoJSBUnfTFZft7vY4iQAAPgnShUkSWnJsbLZpPyjFTpSXmV1HAAA/A6lCpKkqLAgdYuLkCSt5xQgAACNRqmC2yCuVwUAwDmjVMHNfb0qShUAAI1GqYJb/WT1LQfKVF5ZY3EaAAD8C6UKbvFRIUqKDZXLSBv2llgdBwAAv0Kpgof6eVXruAgoAACNQqmCB26uDADAuaFUwUP9ZPWcghJV1TotTgMAgP+wvFTt379fP//5z9W6dWuFhoYqNTVV69evd683xuiRRx5Ru3btFBoaqoyMDG3fvt1jG0VFRRo1apQiIyMVHR2tsWPH6tixYx5jvvzySw0dOlQhISFKSkrSzJkzT8vy7rvvqnv37goJCVFqaqo++uij87PTPqxTm3C1aeVQda1LuftKrY4DAIDfsLRUFRcX67LLLlNQUJAWLVqkLVu26KmnnlJMTIx7zMyZMzVr1izNmTNHa9asUXh4uDIzM1VZWekeM2rUKG3evFlLly7VggULtGrVKv3qV79yry8rK9OwYcOUnJys7OxsPfnkk5o6dapeeukl95jPPvtMI0eO1NixY/XFF19oxIgRGjFihDZt2nRhfhg+wmazKS2ZU4AAADSasdCDDz5oLr/88u9c73K5THx8vHnyySfdy0pKSkxwcLB56623jDHGbNmyxUgy69atc49ZtGiRsdlsZv/+/cYYY55//nkTExNjqqqqPN67W7du7uc333yzGT58uMf7DxkyxNx5550N2pfS0lIjyZSWljZovC975b+7TPKDC8ztf11jdRQAAM4rb35+W3qk6oMPPlBaWppuuukmtW3bVv3799fLL7/sXp+fn6/CwkJlZGS4l0VFRWnIkCFavXq1JGn16tWKjo5WWlqae0xGRoYCAgK0Zs0a95grrrhCDofDPSYzM1N5eXkqLi52jzn1ferH1L/Pt1VVVamsrMzj0VzUT1Zfv6dYTpexOA0AAP7B0lK1a9cuvfDCC+rSpYuWLFmiu+66S/fee6/mzp0rSSosLJQkxcXFebwuLi7Ova6wsFBt27b1WB8YGKjY2FiPMWfaxqnv8V1j6td/27Rp0xQVFeV+JCUlNXr/fVWPdhEKd9hVXlmrvMJyq+MAAOAXLC1VLpdLAwYM0BNPPKH+/fvrV7/6lX75y19qzpw5VsZqkIceekilpaXuR0FBgdWRvCbQHqAByXXz2rhlDQAADWNpqWrXrp169uzpsaxHjx7au3evJCk+Pl6SdOjQIY8xhw4dcq+Lj4/X4cOHPdbX1taqqKjIY8yZtnHqe3zXmPr13xYcHKzIyEiPR3PC9aoAAGgcS0vVZZddpry8PI9lX331lZKTkyVJKSkpio+P17Jly9zry8rKtGbNGqWnp0uS0tPTVVJSouzsbPeY5cuXy+VyaciQIe4xq1atUk3NN/ezW7p0qbp16+b+pmF6errH+9SPqX+flsZ9c+X8IhnDvCoAAM6q6fPmz93atWtNYGCgefzxx8327dvNG2+8YcLCwszf//5395jp06eb6Oho8/7775svv/zSXH/99SYlJcWcOHHCPeZHP/qR6d+/v1mzZo355JNPTJcuXczIkSPd60tKSkxcXJy57bbbzKZNm8zbb79twsLCzIsvvuge8+mnn5rAwEDzpz/9yWzdutVMmTLFBAUFmdzc3AbtS3P69p8xxpyorjWdf7fQJD+4wOw+eszqOAAAnBfe/Py2tFQZY8yHH35oevfubYKDg0337t3NSy+95LHe5XKZhx9+2MTFxZng4GBzzTXXmLy8PI8xX3/9tRk5cqRp1aqViYyMNHfccYcpLy/3GLNx40Zz+eWXm+DgYNO+fXszffr007LMmzfPdO3a1TgcDtOrVy+zcOHCBu9HcytVxhhzw/OfmuQHF5h56/ZaHQUAgPPCm5/fNmM4t+MNZWVlioqKUmlpabOZXzV90TbNWblTN6claubP+lodBwAAr/Pm57flt6mB7xqcUv8NwGKLkwAA4PsoVfhOAzvEymaT8o9W6HB55dlfAABAC0apwneKCgtSt7gISdJ6jlYBAPC9KFX4XoNOXq+Ki4ACAPD9KFX4Xu7rVVGqAAD4XpQqfK/6K6tvOVCm8sqas4wGAKDlolThe8VHhSgpNlQuI23YW2J1HAAAfBalCmflnleVzylAAAC+C6UKZ8XNlQEAODtKFc6qfrJ6TkGJqmqdFqcBAMA3UapwVp3ahKtNK4eqa13K3VdqdRwAAHwSpQpnZbPZlJbMKUAAAL4PpQoN4r5eFZPVAQA4I0oVGqR+svr6PcVyuozFaQAA8D2UKjRIj3YRCnfYVV5Zq7zCcqvjAADgcyhVaJBAe4AGJMdI4pY1AACcCaUKDcb1qgAA+G6UKjTYqZPVjWFeFQAAp6JUocH6JUUryG7T4fIq7S06bnUcAAB8CqUKDRYSZFefxGhJ0lourQAAgAdKFRrFfXNl5lUBAOCBUoVGGZxS/w3AYouTAADgWxpVqtauXSun87tvqFtVVaV58+Y1ORR818DkWNlsUv7RCh0ur7Q6DgAAPqNRpSo9PV1ff/21+3lkZKR27drlfl5SUqKRI0d6Lx18TlRokLrFRUiS1nO0CgAAt0aVqm9/jf5MX6vnq/bNX/28KiarAwDwDa/PqbLZbN7eJHyM+3pVTFYHAMCNiepotPorq289WKbyyhqL0wAA4BsCG/uCLVu2qLCwUFLdqb5t27bp2LFjkqSjR496Nx18UnxUiJJiQ1VQdEIb9pboyq4XWR0JAADLNbpUXXPNNR7zpn784x9LqjvtZ4zh9F8LMahjrAqK9mtdfhGlCgAANbJU5efnn68c8DODO8bqXxv2c3NlAABOalSpSk5OPl854GfqJ6vnFJSoqtap4EC7xYkAALBWoyaqHz16VHv27PFYtnnzZt1xxx26+eab9eabb3o1HHxXpzbhatPKoepal3L3lVodBwAAyzWqVN1zzz2aNWuW+/nhw4c1dOhQrVu3TlVVVbr99tv1+uuvez0kfI/NZlNa8snrVXEKEACAxpWqzz//XD/5yU/cz//2t78pNjZWOTk5ev/99/XEE09o9uzZXg8J3+S+XhUXAQUAoHGlqrCwUB07dnQ/X758uW644QYFBtZNzfrJT36i7du3ezUgfFf99arW7ymW08WV9AEALVujSlVkZKRKSkrcz9euXashQ4a4n9tsNlVVVXktHHxbj3YRCnfYVV5Zq7zCcqvjAABgqUaVqksuuUSzZs2Sy+XSP/7xD5WXl+vqq692r//qq6+UlJTk9ZDwTYH2AA1IjpHELWsAAGhUqXr00Uf1wQcfKDQ0VLfccot++9vfKiYmxr3+7bff1pVXXun1kPBd9acAmawOAGjpGnWdqr59+2rr1q369NNPFR8f73HqT5JuvfVW9ezZ06sB4dtOnazOFfUBAC1Zo45UrV69Wp9//rmuv/56d6H629/+ppSUFLVt21bvv/++EhISzktQ+KZ+SdEKstt0uLxKe4uOWx0HAADLNPr03+bNm93Pc3NzNXbsWGVkZGjy5Mn68MMPNW3aNK+HhO8KCbKrT2K0JGktl1YAALRgjSpVOTk5uuaaa9zP3377bQ0ZMkQvv/yyJk2apFmzZmnevHleDwnfNujkvComqwMAWrJGlari4mLFxcW5n69cuVLXXnut+/mgQYNUUFDgvXTwC4NT6r8BWGxxEgAArNOoUhUXF6f8/HxJUnV1tTZs2KBLLrnEvb68vFxBQUHeTQifNzA5VjablH+0QofLK62OAwCAJRpVqv7nf/5HkydP1n//+1899NBDCgsL09ChQ93rv/zyS1188cVeDwnfFhUapG5xEZKk9RytAgC0UI0qVY899pgCAwN15ZVX6uWXX9bLL78sh8PhXv/Xv/5Vw4YN83pI+L7BJy+twGR1AEBL1ajrVLVp00arVq1SaWmpWrVqJbvd7rH+3XffVatWrbwaEP5hUMdY/W31HiarAwBarEaVqnpRUVFnXB4bG9ukMPBf9d8A3HqwTOWVNYoIYW4dAKBladTpP+C7xEeFKCk2VC4jbdhbYnUcAAAuOEoVvMZ9vSrmVQEAWiBKFbyGmysDAFoyShW8pv7myjkFJaqqdVqcBgCAC4tSBa/p1CZcbVo5VF3rUu6+UqvjAABwQVGq4DU2m01pyZwCBAC0TJQqeFX9KUAmqwMAWhpKFbyqfrL6+j3FcrqMxWkAALhwKFXwqh7tIhTusKu8slZ5heVWxwEA4IKhVMGrAu0BGpAcI0ncsgYA0KJQquB1XK8KANASUargdadOVjeGeVUAgJaBUgWv65cUrSC7TYfLq7S36LjVcQAAuCAoVfC6kCC7+iRGS5LWcmkFAEALQanCeeG+uTLzqgAALQSlCufF4JT6bwAWW5wEAIALg1KF82JgcqxsNin/aIUOl1daHQcAgPOOUoXzIio0SN3iIiRJ6zlaBQBoAShVOG8Gn7y0ApPVAQAtgc+UqunTp8tms2nixInuZZWVlcrKylLr1q3VqlUr3XjjjTp06JDH6/bu3avhw4crLCxMbdu21QMPPKDa2lqPMR9//LEGDBig4OBgde7cWa+99tpp7z979mx17NhRISEhGjJkiNauXXs+drNFYbI6AKAl8YlStW7dOr344ovq06ePx/L77rtPH374od59912tXLlSBw4c0A033OBe73Q6NXz4cFVXV+uzzz7T3Llz9dprr+mRRx5xj8nPz9fw4cN11VVXKScnRxMnTtS4ceO0ZMkS95h33nlHkyZN0pQpU7Rhwwb17dtXmZmZOnz48Pnf+Was/kjV1oNlKq+ssTgNAADnmbFYeXm56dKli1m6dKm58sorzYQJE4wxxpSUlJigoCDz7rvvusdu3brVSDKrV682xhjz0UcfmYCAAFNYWOge88ILL5jIyEhTVVVljDHmt7/9renVq5fHe95yyy0mMzPT/Xzw4MEmKyvL/dzpdJqEhAQzbdq0Bu9HaWmpkWRKS0sbvvMtwNAZy03ygwvMim2HrI4CAMBpvPn5bfmRqqysLA0fPlwZGRkey7Ozs1VTU+OxvHv37urQoYNWr14tSVq9erVSU1MVFxfnHpOZmamysjJt3rzZPebb287MzHRvo7q6WtnZ2R5jAgIClJGR4R5zJlVVVSorK/N44HRpHbm5MgCgZbC0VL399tvasGGDpk2bdtq6wsJCORwORUdHeyyPi4tTYWGhe8yphap+ff267xtTVlamEydO6OjRo3I6nWccU7+NM5k2bZqioqLcj6SkpIbtdAsz2D2vim8AAgCaN8tKVUFBgSZMmKA33nhDISEhVsU4Zw899JBKS0vdj4KCAqsj+aT6myvnFJSoqtZpcRoAAM4fy0pVdna2Dh8+rAEDBigwMFCBgYFauXKlZs2apcDAQMXFxam6ulolJSUerzt06JDi4+MlSfHx8ad9G7D++dnGREZGKjQ0VG3atJHdbj/jmPptnElwcLAiIyM9HjhdpzbhatPKoepal3L3lVodBwCA88ayUnXNNdcoNzdXOTk57kdaWppGjRrl/n1QUJCWLVvmfk1eXp727t2r9PR0SVJ6erpyc3M9vqW3dOlSRUZGqmfPnu4xp26jfkz9NhwOhwYOHOgxxuVyadmyZe4xOHc2m01pySevV8W8KgBAMxZo1RtHRESod+/eHsvCw8PVunVr9/KxY8dq0qRJio2NVWRkpO655x6lp6frkksukSQNGzZMPXv21G233aaZM2eqsLBQ/+///T9lZWUpODhYkvTrX/9azz33nH77299qzJgxWr58uebNm6eFCxe633fSpEkaPXq00tLSNHjwYD399NOqqKjQHXfccYF+Gs3boJRYLd5cqHX5RdIPrE4DAMD5YVmpaoi//OUvCggI0I033qiqqiplZmbq+eefd6+32+1asGCB7rrrLqWnpys8PFyjR4/Wo48+6h6TkpKihQsX6r777tMzzzyjxMREvfLKK8rMzHSPueWWW3TkyBE98sgjKiwsVL9+/bR48eLTJq/j3NRPVl+/p1hOl5E9wGZxIgAAvM9mjDFWh2gOysrKFBUVpdLSUuZXfUut06W+f/i3Kqqd+ujeoeqZwM8HAOAbvPn5bfl1qtD8BdoDNCCZ61UBAJo3ShUuiPpTgExWBwA0V5QqXBD116tal18kzjgDAJojShUuiH5J0Qqy23S4vEp7i45bHQcAAK+jVOGCCAmyq09itCRpbT6nAAEAzQ+lChfMIPd9AClVAIDmh1KFC2ZwSv03ALm5MgCg+aFU4YIZmBwre4BN+Ucr9M/sfVbHAQDAqyhVuGCiQoN015UXS5Ie+lcupwEBAM0KpQoX1KQfdtWPesWr2unSna9nq4BvAgIAmglKFS6ogACb/nxLX/VuH6miimqNnbtO5ZU1VscCAKDJKFW44MIcgXrlF4PUNiJYXx06pnve+kK1TpfVsQAAaBJKFSwRHxWiV0anKSQoQB/nHdHjH221OhIAAE1CqYJl+iRG688395Mkvfrpbr2xZo+1gQAAaAJKFSz1P6ntdP8Pu0qSHnl/sz7bcdTiRAAAnBtKFSw3/urOGtEvQU6X0a//nq1dR45ZHQkAgEajVMFyNptN02/sowEdolVWWauxc9er5Hi11bEAAGgUShV8QkiQXS/elqb20aHKP1qhu9/YoBq+EQgA8COUKviMiyKC9croNIU77Pps59d65P3NMsZYHQsAgAahVMGn9GgXqWdu7S+bTXpr7V799dPdVkcCAKBBKFXwORk94/S7a3tIkh5fuEUrth22OBEAAGdHqYJPGjc0RbekJcllpHve+kJ5heVWRwIA4HtRquCTbDabHhvRW0NSYnWsqlZjXluno8eqrI4FAMB3olTBZzkCAzTn5wPVsXWY9pec0J2vZ6uq1ml1LAAAzohSBZ8WE+7QK6MHKSIkUNl7ivXQP3P5RiAAwCdRquDzOrdtpRdGDZQ9wKZ/fbFfz3+80+pIAACchlIFv3B5lzaa+pNekqQnl+Rp8aaDFicCAMATpQp+47ZLknX7pR0lSfe9s1Gb9pdaGwgAgFNQquBX/t/wHrqi60U6UePU2LnrdKis0upIAABIolTBzwTaA/Tc//ZX57atdKisSuPmrteJar4RCACwHqUKficyJEj/NzpNMWFByt1fqvvfzZHLxTcCAQDWolTBLyW3DteLt6UpyG7TR7mF+st/vrI6EgCghaNUwW8NTonVEz9NlSQ9u3yH3vtiv8WJAAAtGaUKfu2mtCTdeWUnSdJv//mlsvcUW5wIANBSUarg9x7M7K4f9oxTda1Ld76+XvuKj1sdCQDQAlGq4PcCAmx6+pZ+6tEuUkePVWvc3PU6VlVrdSwAQAtDqUKzEB4cqP8bnaaLIoK1rbBcE976Qk6+EQgAuIAoVWg2EqJD9fIv0hQcGKBl2w5r+qKtVkcCALQglCo0K/2SovWnm/pKkl7+b77eWbfX4kQAgJaCUoVm57q+CZqY0UWS9Pv5m7R659cWJwIAtASUKjRLE67pouv6JqjWZXTXG9nafbTC6kgAgGaOUoVmyWaz6cmf9VHfpGiVHK/RmLnrVHq8xupYAIBmjFKFZiskyK6XbxuodlEh2nWkQllvblCN02V1LABAM0WpQrPWNjJEr4xOU5jDrk92HNWjH26xOhIAoJmiVKHZ65UQpadv6SebTXr98z2a+9luqyMBAJohShVahGG94vXgj7pLkv7w4Wat/OqIxYkAAM0NpQotxp1XdNJNAxPlMtL4NzZo+6FyqyMBAJoRShVaDJvNpsd/mqrBHWNVXlWrsXPXq6ii2upYAIBmglKFFsURGKA5tw1Uh9gw7S06rl+/nq2qWqfVsQAAzQClCi1ObLhD/zc6TRHBgVq7u0i/n79JxnDzZQBA01Cq0CJ1iYvQc6MGKMAm/SN7n15ctcvqSAAAP0epQot1ZdeL9MiPe0qSZizepiWbCy1OBADwZ5QqtGijL+2on1/SQcZI972To80HSq2OBADwU5QqtGg2m01Truulyzu30fFqp345d70Ol1daHQsA4IcoVWjxguwBmj1qgDpdFK4DpZX65d+yVVnDNwIBAI1DqQIkRYUG6a+jByk6LEgbC0r0wD++5BuBAIBGoVQBJ3VsE64XRg1UYIBNH248oGeWbbc6EgDAj1CqgFOkX9xaj/+0tyTp6f9s14cbD1icCADgLyhVwLfcMqiDfjk0RZL0m3c3KqegxNpAAAC/QKkCzmDytT10Tfe2qqp1adzc9TpQcsLqSAAAH0epAs7AHmDTMyP7q3t8hI4eq9LYuetVUVVrdSwAgA+jVAHfoVVwoF4ZnaY2rRzaerBME9/JkcvFNwIBAGdGqQK+R2JMmF68LU2OwAAt3XJIM5fkWR0JAOCjKFXAWQxMjtHMG/tIkuas3Kl31xdYnAgA4IsoVUADjOjfXvdc3VmS9Lv5uVqbX2RxIgCAr6FUAQ10X0ZX/U9qvGqcRne+vl57vz5udSQAgA+hVAENFBBg01M39VOfxCgVH6/RmLnrVFZZY3UsAICPoFQBjRDqsOvlX6QpPjJEOw4f0/g3v1Ct02V1LACAD7C0VE2bNk2DBg1SRESE2rZtqxEjRigvz/PbVZWVlcrKylLr1q3VqlUr3XjjjTp06JDHmL1792r48OEKCwtT27Zt9cADD6i21vOaQh9//LEGDBig4OBgde7cWa+99tppeWbPnq2OHTsqJCREQ4YM0dq1a72+z/B/cZEhemV0mkKD7Fr11RH9ceFWqyMBAHyApaVq5cqVysrK0ueff66lS5eqpqZGw4YNU0VFhXvMfffdpw8//FDvvvuuVq5cqQMHDuiGG25wr3c6nRo+fLiqq6v12Wefae7cuXrttdf0yCOPuMfk5+dr+PDhuuqqq5STk6OJEydq3LhxWrJkiXvMO++8o0mTJmnKlCnasGGD+vbtq8zMTB0+fPjC/DDgV3q3j9JfbukrSXrts93689KvdKLaaXEqAICVbMYYn7ma4ZEjR9S2bVutXLlSV1xxhUpLS3XRRRfpzTff1M9+9jNJ0rZt29SjRw+tXr1al1xyiRYtWqQf//jHOnDggOLi4iRJc+bM0YMPPqgjR47I4XDowQcf1MKFC7Vp0yb3e916660qKSnR4sWLJUlDhgzRoEGD9Nxzz0mSXC6XkpKSdM8992jy5MmnZa2qqlJVVZX7eVlZmZKSklRaWqrIyMjz9jOCb5m9YoeePHntqtbhDo0b2kk/v6SDIkKCLE4GAGiIsrIyRUVFeeXz26fmVJWWlkqSYmNjJUnZ2dmqqalRRkaGe0z37t3VoUMHrV69WpK0evVqpaamuguVJGVmZqqsrEybN292jzl1G/Vj6rdRXV2t7OxsjzEBAQHKyMhwj/m2adOmKSoqyv1ISkpq6u7DD939g4s188Y+SowJ1dcV1ZqxeJsun7FCT//nK5Ucr7Y6HgDgAvKZUuVyuTRx4kRddtll6t27tySpsLBQDodD0dHRHmPj4uJUWFjoHnNqoapfX7/u+8aUlZXpxIkTOnr0qJxO5xnH1G/j2x566CGVlpa6HwUFXBCyJbLZbLp5UJJW/OYHeuqmvup0UbhKT9To6f9s1+UzVmj6om06eqzq7BsCAPi9QKsD1MvKytKmTZv0ySefWB2lQYKDgxUcHGx1DPiIIHuAbhyYqBH922vxpkI9u3y7thWWa87KnXrts3yNHNxBv7qik9pFhVodFQBwnvjEkarx48drwYIFWrFihRITE93L4+PjVV1drZKSEo/xhw4dUnx8vHvMt78NWP/8bGMiIyMVGhqqNm3ayG63n3FM/TaAhrAH2DS8TzstmjBUr/wiTX2TolVZ49Krn+7WlTM/1kP/yuWioQDQTFlaqowxGj9+vObPn6/ly5crJSXFY/3AgQMVFBSkZcuWuZfl5eVp7969Sk9PlySlp6crNzfX41t6S5cuVWRkpHr27Okec+o26sfUb8PhcGjgwIEeY1wul5YtW+YeAzSGzWZTRs84vXf3pXp97GANSYlVtdOlt9bu1VVPfaxJ83K04/Axq2MCALzI0m//3X333XrzzTf1/vvvq1u3bu7lUVFRCg2tO01y11136aOPPtJrr72myMhI3XPPPZKkzz77TFLdJRX69eunhIQEzZw5U4WFhbrttts0btw4PfHEE5LqLqnQu3dvZWVlacyYMVq+fLnuvfdeLVy4UJmZmZLqLqkwevRovfjiixo8eLCefvppzZs3T9u2bTttrtWZePPbA2ie1uYX6bkVO7TqqyOSJJtN+p/Udsr6QWf1TOC/GQCwglc/v42FJJ3x8eqrr7rHnDhxwtx9990mJibGhIWFmZ/+9Kfm4MGDHtvZvXu3ufbaa01oaKhp06aNuf/++01NTY3HmBUrVph+/foZh8NhOnXq5PEe9Z599lnToUMH43A4zODBg83nn3/e4H0pLS01kkxpaWmjfgZoeXL2Fptxc9eZ5AcXuB9jX1trvthbbHU0AGhxvPn57VPXqfJnHKlCY20rLNPsFTu14MsDqv9bOLRLG42/qrOGdGptbTgAaCG8+flNqfISShXO1c4jx/TCxzs1/4v9crrq/joO6hij8Vd30RVd2shms1mcEACaL0qVD6JUoakKio5rzsqdenf9PlWfvElzn8Qojb+qszJ6xCkggHIFAN5GqfJBlCp4S2FppV7+7y69sWaPKmvqylX3+AjdfVVnDU9tJzvlCgC8hlLlgyhV8Lajx6r010/y9bfVe3SsqlaS1KlNuO76wcUa0b+9guw+cZk5APBrlCofRKnC+VJ6vEZzV+/WXz/NV8nxGklS++hQ3fWDi/WzgYkKCbJbnBAA/BelygdRqnC+Hauq1Ruf79HL/92lo8fqbtYcFxmsXw7tpP8d0kFhDp+56xQA+A1KlQ+iVOFCqaxx6u21e/Xiql06WFopSYoNd2js5Sn6RXqyIkKCLE4IAP6DUuWDKFW40KprXfrXhn16/uOd2ltUdz/ByJBA3X5Ziu64tKNiwh0WJwQA30ep8kGUKlil1unSh18e0HPLd2jnkQpJUpjDrtsuSdbYoSlqGxFicUIA8F2UKh9EqYLVXC6jxZsL9dzyHdpysEySFBwYoJGDO+hXV3RSQnSoxQkBwPdQqnwQpQq+whijFXmHNWvZDuUUlEiSguw2/Wxgon595cVKbh1ubUAA8CGUKh9EqYKvMcbos51f69nl2/X5riJJUoBNur5fe939g4vVJS7C4oQAYD1KlQ+iVMGXrd9dpOdW7NDHeUckSTabdG3veGVd1Vm9EqIsTgcA1qFU+SBKFfxB7r5SPbdiu5ZsPuRednX3thp/dWcN6BBjYTIAsAalygdRquBP8grLNXvFDi348oBcJ/8FuKxza42/qosu6RQrm437CwJoGShVPohSBX+Uf7RCL3y8Q//asF+1J9vVwOQYjb+qs4Z2aaNA7i8IoJmjVPkgShX82b7i43px5S69s75A1bUuSVJIUIB6JUQptX2U+iTW/drpolayB3AUC0DzQanyQZQqNAeHyyr18n936e11BSqvrD1tfZjDrt4JUUpNrCtavdtHKaV1uAIoWgD8FKXKB1Gq0Jy4XEb5X1cod1+pvtxXqtz9Jdp8oEzHq52njW0VHKje7SPVJzFaqe3rjmgltw5jXhYAv0Cp8kGUKjR3TpfRriPHTpasUn25r0RbDpapssZ12tjIkEClJkYptX20+9RhYkwoRQuAz6FU+SBKFVqiWqdLO+qL1r5Sfbm/VFsPlrnnZZ0qOizolPlZ0UpNjFJCVAhFC4ClKFU+iFIF1KlxuvTVoXJ3ycrdV6pthWWqcZ7+T03rcEfd/Kz2UUpNrDuqFRfJDaABXDiUKh9EqQK+W1WtU18VHtOX+0vc87S+OlTuvozDqS6KCD5Zsr6ZDN82gqIF4PygVPkgShXQOJU1Tm0rLFfuvhL3PK3th4/JeYaiFR8ZcsoRrbo5Wq1bBVuQGkBzQ6nyQZQqoOlOVDu15WBZXdHaX6pNJ4vWmf6Vah8dWvdtw8RvrqMVHea48KEB+DVKlQ+iVAHnR0VVrbYcLDs5Gb6ubOUfrThj0UqKDVWfk5Pg+7SPUq/2UYoKDbrwoQH4DUqVD6JUARdOeWWNNh8oO2UyfIl2f338jGM7tg5TamK0eiVEqmPrMCXGhKlD6zBFhlC2AFCqfBKlCrBW6Ykabd7/zTcOv9xfooKiE985Pio0SB1iw9QhNkyJsaHqEBumpJi65wnRoXIEct9DoCWgVPkgShXge4orqrXpQN23DbcVlqug6Lj2FR/X0WPV3/u6AJvULipUiTGhSjpZvJJOKV4XRQRzfS2gmaBU+SBKFeA/Kqpqta/4hPYWHVdB0XHtPVm26p6f0Ima02/Hc6qQoIC604ixYUo6Wby+KV9hahUceIH2BEBTefPzm7/5AFqc8OBAdYuPULf4iNPWGWN09Fi1CorrCld96SooqithB0tPqLLGpR2Hj2nH4WNn3H5suOP0snWyhLWLDlGQnVOLQHNEqQKAU9hsNl0UEayLIoI1oEPMaetrnC4dKDnhLlkFJ49w7TtZvoqP16ioolpFFdXauK/0tNfbA2xqFxXiLllJsZ7lq3W4g1OLgJ+iVAFAIwTZA5TcOlzJrcPPuL68skYFRSdOP9JVfEIFRcdVVevSvuIT2ld8Qqt3fX3a60OD7J5lK+ab04pJsaEKc/DPNuCr+NsJAF4UERKknglB6plw+twMl8vo6LEq99Et99Guk4+DZZU6UeNU3qFy5R0qP+P227RyuOdztYsOUWyYQzHhDsWEORQbHnTyV4ciQ4IUEMARL+BCYqK6lzBRHUBTVdU6daCk8pSjW/WFq658lZ6oafC2AmxSdJhDMWF1RSsm3KHYMIeiw4POWMRiwhyKCqWIoeVhojoANEPBgXaltAlXSpszn1osPVHjvizE3qLjOlRWpeKKahUfr1bR8Rr378sra+Uycs/tkioa9P71RSw67JviRREDGo5SBQB+Iio0SFHto9S7fdT3jquudankRLWKK2pUfLxaxRXVKjperZKTk+gbUsR2NaGIxYQFuQsZRQwtCaUKAJoZR2CA2kaEqG1ESINf4ytFLDIkSOEOu8KCA9UqOFBhDrvCgwMV7ghUePA3vw8JCuBbkvA5lCoAgM8XsW8LsEnhjkCFBdtPFq66AtYqOFBhwYEKd5exul89l9W9rtWpr3EEcmsiNBmlCgBwTs5HESuvrFFFtVPHq2pVUeVURXWtjlc7dayqtm5Zdd3V7l1GKq+qVXlVraQqr+xPkN32TemqL2BnKm2nHDU707L6o2phjkDZOc3ZolCqAAAXzLkUsVO5XEYnaurKVkWVUxVVtaqoqitedctOWV7t1PHq2pOF7Jv17pJWXTeuqtYlSapxGpUcr1HJ8YZ/y/JsQoICFBJkV0ig3f374CC7QgJPLq9f5n5ety64/vdBASdfe+qYb34NDrR7LAsMsHFa1EKUKgCA3wgIOHk0KThQOv0uQ+ekxulyF6zj1Z6lrO7XulJ2rH59/fKqU15T/U3Bq6h2yumqu1pRZY1LlTUuSd4rat8nwCaPcuYucSfLWbC7pH1T4oJPKW7fFLUAj1IX/K0CGBvu4EK0Z8BPBADQogXZAxQVGqCo0CCvbM8Yo6rab4paZY2zrlzVOr/5fY1TVbWuk889f3/a+pO/Vp1hO1W1zpOlrY7LSMernTpe/f03BW+q/ze8h8YN7XRe38MfUaoAAPAim83mPgoUG+447+9XX+KqzlLcTitnNU5V1jpPeZ3rW2XNs7idup2QIPt53y9/RKkCAMCPnVriouSdo204N3x/FAAAwAsoVQAAAF5AqQIAAPACShUAAIAXUKoAAAC8gFIFAADgBZQqAAAAL6BUAQAAeAGlCgAAwAsoVQAAAF5AqQIAAPACShUAAIAXUKoAAAC8gFIFAADgBYFWB2gujDGSpLKyMouTAACAhqr/3K7/HG8KSpWXlJeXS5KSkpIsTgIAABrr66+/VlRUVJO2YTPeqGaQy+XSgQMHFBERIZvN5pVtlpWVKSkpSQUFBYqMjPTKNq3GPvmH5rZPzW1/JPbJXzS3fWpu+yNJpaWl6tChg4qLixUdHd2kbXGkyksCAgKUmJh4XrYdGRnZbP7jrcc++Yfmtk/NbX8k9slfNLd9am77I9V9jjd5G17IAQAA0OJRqgAAALyAUuXDgoODNWXKFAUHB1sdxWvYJ//Q3Papue2PxD75i+a2T81tfyTv7hMT1QEAALyAI1UAAABeQKkCAADwAkoVAACAF1CqAAAAvIBS5YNWrVql6667TgkJCbLZbHrvvfesjtRk06ZN06BBgxQREaG2bdtqxIgRysvLszrWOXvhhRfUp08f9wXw0tPTtWjRIqtjedX06dNls9k0ceJEq6Ocs6lTp8pms3k8unfvbnWsJtu/f79+/vOfq3Xr1goNDVVqaqrWr19vdaxz1rFjx9P+nGw2m7KysqyOdk6cTqcefvhhpaSkKDQ0VBdffLEee+wxr9xbzkrl5eWaOHGikpOTFRoaqksvvVTr1q2zOlaDne2z1RijRx55RO3atVNoaKgyMjK0ffv2Rr0HpcoHVVRUqG/fvpo9e7bVUbxm5cqVysrK0ueff66lS5eqpqZGw4YNU0VFhdXRzkliYqKmT5+u7OxsrV+/XldffbWuv/56bd682epoXrFu3Tq9+OKL6tOnj9VRmqxXr146ePCg+/HJJ59YHalJiouLddlllykoKEiLFi3Sli1b9NRTTykmJsbqaOds3bp1Hn9GS5culSTddNNNFic7NzNmzNALL7yg5557Tlu3btWMGTM0c+ZMPfvss1ZHa5Jx48Zp6dKlev3115Wbm6thw4YpIyND+/fvtzpag5zts3XmzJmaNWuW5syZozVr1ig8PFyZmZmqrKxs+JsY+DRJZv78+VbH8LrDhw8bSWblypVWR/GamJgY88orr1gdo8nKy8tNly5dzNKlS82VV15pJkyYYHWkczZlyhTTt29fq2N41YMPPmguv/xyq2OcVxMmTDAXX3yxcblcVkc5J8OHDzdjxozxWHbDDTeYUaNGWZSo6Y4fP27sdrtZsGCBx/IBAwaY3//+9xalOnff/mx1uVwmPj7ePPnkk+5lJSUlJjg42Lz11lsN3i5HqmCJ0tJSSVJsbKzFSZrO6XTq7bffVkVFhdLT062O02RZWVkaPny4MjIyrI7iFdu3b1dCQoI6deqkUaNGae/evVZHapIPPvhAaWlpuummm9S2bVv1799fL7/8stWxvKa6ulp///vfNWbMGK/dnP5Cu/TSS7Vs2TJ99dVXkqSNGzfqk08+0bXXXmtxsnNXW1srp9OpkJAQj+WhoaF+f/RXkvLz81VYWOjx715UVJSGDBmi1atXN3g73FAZF5zL5dLEiRN12WWXqXfv3lbHOWe5ublKT09XZWWlWrVqpfnz56tnz55Wx2qSt99+Wxs2bPCreRLfZ8iQIXrttdfUrVs3HTx4UH/4wx80dOhQbdq0SREREVbHOye7du3SCy+8oEmTJul3v/ud1q1bp3vvvVcOh0OjR4+2Ol6TvffeeyopKdHtt99udZRzNnnyZJWVlal79+6y2+1yOp16/PHHNWrUKKujnbOIiAilp6frscceU48ePRQXF6e33npLq1evVufOna2O12SFhYWSpLi4OI/lcXFx7nUNQanCBZeVlaVNmzb5/f/ddOvWTTk5OSotLdU//vEPjR49WitXrvTbYlVQUKAJEyZo6dKlp/3fqL869chAnz59NGTIECUnJ2vevHkaO3ashcnOncvlUlpamp544glJUv/+/bVp0ybNmTOnWZSq//u//9O1116rhIQEq6Ocs3nz5umNN97Qm2++qV69eiknJ0cTJ05UQkKCX/8Zvf766xozZozat28vu92uAQMGaOTIkcrOzrY6ms/g9B8uqPHjx2vBggVasWKFEhMTrY7TJA6HQ507d9bAgQM1bdo09e3bV88884zVsc5Zdna2Dh8+rAEDBigwMFCBgYFauXKlZs2apcDAQDmdTqsjNll0dLS6du2qHTt2WB3lnLVr1+604t6jRw+/P60pSXv27NF//vMfjRs3zuooTfLAAw9o8uTJuvXWW5WamqrbbrtN9913n6ZNm2Z1tCa5+OKLtXLlSh07dkwFBQVau3atampq1KlTJ6ujNVl8fLwk6dChQx7LDx065F7XEJQqXBDGGI0fP17z58/X8uXLlZKSYnUkr3O5XKqqqrI6xjm75pprlJubq5ycHPcjLS1No0aNUk5Ojux2u9URm+zYsWPauXOn2rVrZ3WUc3bZZZeddjmSr776SsnJyRYl8p5XX31Vbdu21fDhw62O0iTHjx9XQIDnx6vdbpfL5bIokXeFh4erXbt2Ki4u1pIlS3T99ddbHanJUlJSFB8fr2XLlrmXlZWVac2aNY2aK8vpPx907Ngxj/+Tzs/PV05OjmJjY9WhQwcLk527rKwsvfnmm3r//fcVERHhPkcdFRWl0NBQi9M13kMPPaRrr71WHTp0UHl5ud588019/PHHWrJkidXRzllERMRpc9zCw8PVunVrv5379pvf/EbXXXedkpOTdeDAAU2ZMkV2u10jR460Oto5u++++3TppZfqiSee0M0336y1a9fqpZde0ksvvWR1tCZxuVx69dVXNXr0aAUG+vdH03XXXafHH39cHTp0UK9evfTFF1/oz3/+s8aMGWN1tCZZsmSJjDHq1q2bduzYoQceeEDdu3fXHXfcYXW0BjnbZ+vEiRP1xz/+UV26dFFKSooefvhhJSQkaMSIEQ1/E+99QRHesmLFCiPptMfo0aOtjnbOzrQ/ksyrr75qdbRzMmbMGJOcnGwcDoe56KKLzDXXXGP+/e9/Wx3L6/z9kgq33HKLadeunXE4HKZ9+/bmlltuMTt27LA6VpN9+OGHpnfv3iY4ONh0797dvPTSS1ZHarIlS5YYSSYvL8/qKE1WVlZmJkyYYDp06GBCQkJMp06dzO9//3tTVVVldbQmeeedd0ynTp2Mw+Ew8fHxJisry5SUlFgdq8HO9tnqcrnMww8/bOLi4kxwcLC55pprGv3fo80YP7/EKwAAgA9gThUAAIAXUKoAAAC8gFIFAADgBZQqAAAAL6BUAQAAeAGlCgAAwAsoVQAAAF5AqQIAAPACShUAn7Z7927ZbDbl5ORYHcVt27ZtuuSSSxQSEqJ+/fo1aVs2m03vvfeeV3IBsBalCsD3uv3222Wz2TR9+nSP5e+9955sNptFqaw1ZcoUhYeHKy8vz+MGrN9WWFioe+65R506dVJwcLCSkpJ03XXXfe9rmuLjjz+WzWZTSUnJedk+gO9HqQJwViEhIZoxY4aKi4utjuI11dXV5/zanTt36vLLL1dycrJat259xjG7d+/WwIEDtXz5cj355JPKzc3V4sWLddVVVykrK+uc3/tCMMaotrbW6hiA36FUATirjIwMxcfHa9q0ad85ZurUqaedCnv66afVsWNH9/Pbb79dI0aM0BNPPKG4uDhFR0fr0UcfVW1trR544AHFxsYqMTFRr7766mnb37Ztmy699FKFhISod+/eWrlypcf6TZs26dprr1WrVq0UFxen2267TUePHnWv/8EPfqDx48dr4sSJatOmjTIzM8+4Hy6XS48++qgSExMVHBysfv36afHixe71NptN2dnZevTRR2Wz2TR16tQzbufuu++WzWbT2rVrdeONN6pr167q1auXJk2apM8///yMrznTkaacnBzZbDbt3r1bkrRnzx5dd911iomJUXh4uHr16qWPPvpIu3fv1lVXXSVJiomJkc1m0+233+7ep2nTpiklJUWhoaHq27ev/vGPf5z2vosWLdLAgQMVHBysTz75RBs3btRVV12liIgIRUZGauDAgVq/fv0ZswOgVAFoALvdrieeeELPPvus9u3b16RtLV++XAcOHNCqVav05z//WVOmTNGPf/xjxcTEaM2aNfr1r3+tO++887T3eeCBB3T//ffriy++UHp6uq677jp9/fXXkqSSkhJdffXV6t+/v9avX6/Fixfr0KFDuvnmmz22MXfuXDkcDn366aeaM2fOGfM988wzeuqpp/SnP/1JX375pTIzM/WTn/xE27dvlyQdPHhQvXr10v3336+DBw/qN7/5zWnbKCoq0uLFi5WVlaXw8PDT1kdHR5/Lj06SlJWVpaqqKq1atUq5ubmaMWOGWrVqpaSkJP3zn/+UJOXl5engwYN65plnJEnTpk3T3/72N82ZM0ebN2/Wfffdp5///OenFdPJkydr+vTp2rp1q/r06aNRo0YpMTFR69atU3Z2tiZPnqygoKBzzg40ewYAvsfo0aPN9ddfb4wx5pJLLjFjxowxxhgzf/58c+o/IVOmTDF9+/b1eO1f/vIXk5yc7LGt5ORk43Q63cu6detmhg4d6n5eW1trwsPDzVtvvWWMMSY/P99IMtOnT3ePqampMYmJiWbGjBnGGGMee+wxM2zYMI/3LigoMJJMXl6eMcaYK6+80vTv3/+s+5uQkGAef/xxj2WDBg0yd999t/t53759zZQpU75zG2vWrDGSzL/+9a+zvp8kM3/+fGOMMStWrDCSTHFxsXv9F198YSSZ/Px8Y4wxqampZurUqWfc1pleX1lZacLCwsxnn33mMXbs2LFm5MiRHq977733PMZERESY11577az7AKBOoGVtDoDfmTFjhq6++uozHp1pqF69eikg4JuD5HFxcerdu7f7ud1uV+vWrXX48GGP16Wnp7t/HxgYqLS0NG3dulWStHHjRq1YsUKtWrU67f127typrl27SpIGDhz4vdnKysp04MABXXbZZR7LL7vsMm3cuLGBe1g3J+l8uffee3XXXXfp3//+tzIyMnTjjTeqT58+3zl+x44dOn78uH74wx96LK+urlb//v09lqWlpXk8nzRpksaNG6fXX39dGRkZuummm3TxxRd7b2eAZobTfwAa7IorrlBmZqYeeuih09YFBAScViZqampOG/ft00c2m+2My1wuV4NzHTt2TNddd51ycnI8Htu3b9cVV1zhHnemU3HnQ5cuXWSz2bRt27ZGva6+bJ76c/z2z3DcuHHatWuXbrvtNuXm5iotLU3PPvvsd27z2LFjkqSFCxd6/Gy2bNniMa9KOv3nM3XqVG3evFnDhw/X8uXL1bNnT82fP79R+wS0JJQqAI0yffp0ffjhh1q9erXH8osuukiFhYUehcCb15Y6dXJ3bW2tsrOz1aNHD0nSgAEDtHnzZnXs2FGdO3f2eDSmSEVGRiohIUGffvqpx/JPP/1UPXv2bPB2YmNjlZmZqdmzZ6uiouK09d91yYOLLrpIUt28rXpn+hkmJSXp17/+tf71r3/p/vvv18svvyxJcjgckiSn0+ke27NnTwUHB2vv3r2n/WySkpLOui9du3bVfffdp3//+9+64YYbzvglAgB1KFUAGiU1NVWjRo3SrFmzPJb/4Ac/0JEjRzRz5kzt3LlTs2fP1qJFi7z2vrNnz9b8+fO1bds2ZWVlqbi4WGPGjJFUN3m7qKhII0eO1Lp167Rz504tWbJEd9xxh0fBaIgHHnhAM2bM0DvvvKO8vDxNnjxZOTk5mjBhQqPzOp1ODR48WP/85z+1fft2bd26VbNmzfI4lXmq+qIzdepUbd++XQsXLtRTTz3lMWbixIlasmSJ8vPztWHDBq1YscJdLpOTk2Wz2bRgwQIdOXJEx44dU0REhH7zm9/ovvvu09y5c7Vz505t2LBBzz77rObOnfud+U+cOKHx48fr448/1p49e/Tpp59q3bp17vcCcDpKFYBGe/TRR087PdejRw89//zzmj17tvr27au1a9c2ae7Vt02fPl3Tp09X37599cknn+iDDz5QmzZtJMl9dMnpdGrYsGFKTU3VxIkTFR0d7TF/qyHuvfdeTZo0Sffff79SU1O1ePFiffDBB+rSpUujttOpUydt2LBBV111le6//3717t1bP/zhD7Vs2TK98MILZ3xNUFCQ3nrrLW3btk19+vTRjBkz9Mc//tFjjNPpVFZWlnr06KEf/ehH6tq1q55//nlJUvv27fWHP/xBkydPVlxcnMaPHy9Jeuyxx/Twww9r2rRp7tctXLhQKSkp35nfbrfr66+/1i9+8Qt17dpVN998s6699lr94Q9/aNTPAWhJbOZ8zqgEAABoIThSBQAA4AWUKgAAAC+gVAEAAHgBpQoAAMALKFUAAABeQKkCAADwAkoVAACAF1CqAAAAvIBSBQAA4AWUKgAAAC+gVAEAAHjB/wd7yAHtlNCeaQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "# Retrieve Only the Numeric Values\n",
    "data = filter_gene_expression_data.iloc[0:,1:].values\n",
    "\n",
    "# Initialize the Matrix\n",
    "num_rows, num_cols = data.shape\n",
    "biclusters, layers = [], []\n",
    "\n",
    "# Use Elbow Method to Get the Optimal Number of Cluster\n",
    "# Create Scaled DataFrame where Each Variable has Mean of 0 and Standard Deviation of 1\n",
    "scaled_integrated_input = StandardScaler().fit_transform(data)\n",
    "\n",
    "# Create List to Hold SSE Values for Each k\n",
    "sse = []\n",
    "\n",
    "# Perform KMeans Algorithm by Different Value of K\n",
    "for k in range(1, 10):\n",
    "    kmeans = KMeans(n_clusters=k)\n",
    "    kmeans.fit(scaled_integrated_input)\n",
    "    sse.append(kmeans.inertia_)\n",
    "\n",
    "# visualize results\n",
    "plt.plot(range(1, 10), sse)\n",
    "plt.xticks(range(1, 11))\n",
    "plt.xlabel(\"Number of Clusters\")\n",
    "plt.ylabel(\"SSE\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*We can observe that the “elbow” is the number 4 which is optimal for this case.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Plaid Biclustering Model**\n",
    "\n",
    "*Plaid Biclustering Model will be applied to the dataset with the number of biclusters as 4.\n",
    "<br><br>\n",
    "The code is reference from <br>\n",
    "@article{padilha2017, <br>\n",
    "  title={A systematic comparative evaluation of biclustering techniques}, <br>\n",
    "  author={Padilha, Victor A and Campello, Ricardo J G B}, <br>\n",
    "  journal={BMC Bioinformatics}, <br>\n",
    "  volume={18}, <br>\n",
    "  number={1}, <br>\n",
    "  pages={55}, <br>\n",
    "  year={2017}, <br>\n",
    "  publisher={BioMed Central} <br>\n",
    "}*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     HSPA6\n",
       "1    GUCA1A\n",
       "2      CCL5\n",
       "3     MMP14\n",
       "4     TRADD\n",
       "Name: Gene Symbol, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stored the Gene Symbol in the Array\n",
    "# Can Be Retrieved the Genes After Biclustering Analysis\n",
    "gene_symbols = filter_gene_expression_data['Gene Symbol']\n",
    "\n",
    "# Display the First Five Rows of the Genes\n",
    "gene_symbols.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster(rows=[  22   49   51   81   98   99  115  125  137  149  150  152  158  165\n",
      "  175  188  190  191  193  194  220  224  231  234  253  276  286  294\n",
      "  295  315  320  371  373  375  389  408  426  429  495  500  517  526\n",
      "  536  547  557  571  612  653  671  692  694  740  741  772  790  797\n",
      "  819  841  849  859  862  867  874  884  890  937 1066 1131 1137 1140\n",
      " 1146 1159 1223 1266 1399 1427 1440 1607 1641 1666 1677 1705 1706 1711\n",
      " 1746 1748 1793 1843 1844 1853 1858 1859 1894 1898 1949 1960 1967 1968\n",
      " 1997 2040 2043 2058 2233 2269 2283 2288 2290 2307 2336 2345 2347 2351\n",
      " 2353 2355 2362 2385 2386 2430 2521 2594 2602 2680 2687 2704], cols=[18 23 24 26 27 29 30 31 32 33])\n",
      "Bicluster(rows=[   5    8   25   33   79   94  156  171  196  217  242  279  293  348\n",
      "  350  365  374  393  401  419  432  438  443  444  464  502  540  562\n",
      "  573  576  594  608  624  640  703  713  715  728  753  764  766  767\n",
      "  778  798  808  809  815  827  870  879  903  908  942  959  971  979\n",
      " 1023 1063 1088 1091 1102 1108 1155 1180 1207 1210 1216 1238 1268 1276\n",
      " 1283 1301 1360 1370 1385 1452 1524 1540 1542 1554 1618 1639 1650 1659\n",
      " 1660 1679 1681 1707 1721 1723 1730 1802 1806 1835 1857 1886 1889 1915\n",
      " 1933 1969 1973 1993 2010 2011 2018 2029 2047 2057 2072 2125 2169 2175\n",
      " 2210 2217 2259 2270 2299 2304 2309 2311 2325 2341 2352 2357 2371 2378\n",
      " 2390 2402 2431 2447 2464 2477 2499 2501 2520 2523 2529 2548 2591 2658\n",
      " 2684 2701 2721], cols=[17 18 19 20 23 24 25 26 28 29 30 31 32 33])\n",
      "Bicluster(rows=[ 306  521  833 1742 2171 2389], cols=[33])\n",
      "Bicluster(rows=[ 202  313  521  733  850  914 1518 1742 1778 2016 2146 2171 2528], cols=[30])\n"
     ]
    }
   ],
   "source": [
    "# Apply the Plaid Biclustering Model\n",
    "import numpy as np\n",
    "\n",
    "from biclustlib.algorithms import Plaid\n",
    "\n",
    "# creating an instance of the Plaid class and running with the parameters of the original study\n",
    "plaid = Plaid( num_biclusters=4, fit_background_layer=True, row_prunning_threshold=0.5,\n",
    "                 col_prunning_threshold=0.3, significance_tests=0, back_fitting_steps=1,\n",
    "                 initialization_iterations=6, iterations_per_layer=10)\n",
    "biclustering = plaid.run(data)\n",
    "\n",
    "print(biclustering)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Store Bicluster Data Together With Its Gene Symbol**\n",
    "*In the process below, each bicluster data will be append with the gene symbol and save it in the file.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 1 Data with Gene Symbols:\n",
      "   Gene Symbol  GSM509804_E1507T.CEL  GSM509809_E1542T.CEL  \\\n",
      "22       ABCF1              8.220648              7.910793   \n",
      "49       AP2M1             10.741044              9.977123   \n",
      "\n",
      "    GSM509810_E1546T.CEL  GSM509812_E1584T.CEL  GSM509813_E1589T.CEL  \\\n",
      "22              8.712417              8.613985              9.082743   \n",
      "49             10.212407             10.456746             10.831445   \n",
      "\n",
      "    GSM509815_E1610T.CEL  GSM509816_E1614T.CEL  GSM509817_E1635T.CEL  \\\n",
      "22              8.246391              8.755096              8.835079   \n",
      "49             10.036581             10.342455             10.531348   \n",
      "\n",
      "    GSM509818_E1709T.CEL  GSM509819_E1796T.CEL  \n",
      "22              8.682854              8.672898  \n",
      "49             10.527527             10.845476  \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Bicluster 2 Data with Gene Symbols:\n",
      "  Gene Symbol  GSM509803_E2644N.CEL  GSM509804_E1507T.CEL  \\\n",
      "5        PLD1              6.881787              5.259621   \n",
      "8      CAPNS1             11.487255             11.683634   \n",
      "\n",
      "   GSM509805_E1520T.CEL  GSM509806_E1521T.CEL  GSM509809_E1542T.CEL  \\\n",
      "5              6.110842               6.33293              6.353203   \n",
      "8              9.717346               9.44340              9.738445   \n",
      "\n",
      "   GSM509810_E1546T.CEL  GSM509811_E1566T.CEL  GSM509812_E1584T.CEL  \\\n",
      "5              5.717247              5.599669              5.322972   \n",
      "8             10.297885              9.856282             10.251217   \n",
      "\n",
      "   GSM509814_E1603T.CEL  GSM509815_E1610T.CEL  GSM509816_E1614T.CEL  \\\n",
      "5              5.563821              5.048422              5.617596   \n",
      "8             10.479287              9.159860              9.994437   \n",
      "\n",
      "   GSM509817_E1635T.CEL  GSM509818_E1709T.CEL  GSM509819_E1796T.CEL  \n",
      "5              5.885284              6.386359              4.944501  \n",
      "8              9.873228              9.720860              9.232219  \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Bicluster 3 Data with Gene Symbols:\n",
      "    Gene Symbol  GSM509819_E1796T.CEL\n",
      "306        DVL3              7.691792\n",
      "521       EPHB4              7.516879\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Bicluster 4 Data with Gene Symbols:\n",
      "    Gene Symbol  GSM509816_E1614T.CEL\n",
      "202        NREP              7.824220\n",
      "313        HEXB              8.918272\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Initialize an Empty List to Store the Data for Each Bicluster Along with Gene Symbols\n",
    "bicluster_expression_data = []\n",
    "\n",
    "# Iterate Over Each Bicluster in bicluster_data\n",
    "for bicluster in biclustering.biclusters:\n",
    "    rows = bicluster.rows\n",
    "    cols = bicluster.cols\n",
    "    \n",
    "    # Extract the Subset of Data Based on Rows and Columns\n",
    "    bicluster_data_subset = filter_gene_expression_data.iloc[rows, cols]\n",
    "    \n",
    "    # Extract Gene Symbols Corresponding to the Rows\n",
    "    gene_symbols_subset = gene_symbols.iloc[rows]\n",
    "    \n",
    "    # Concatenate Gene Symbols and Expression Data Along the Columns Axis\n",
    "    bicluster_data = pd.concat([gene_symbols_subset, bicluster_data_subset], axis=1)\n",
    "    \n",
    "    # Append the DataFrame to the List\n",
    "    bicluster_expression_data.append(bicluster_data)\n",
    "\n",
    "# Save Each Bicluster's Data to a Separate CSV File\n",
    "for i, data_with_symbols in enumerate(bicluster_expression_data, start=1):\n",
    "    data_with_symbols.to_csv(f'D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/Bicluster Expression Data {i}.csv', index=False)\n",
    "\n",
    "\n",
    "# Show the Dataset of Bicluster\n",
    "# Display the Value of the DataFrame\n",
    "for i, data_with_symbols in enumerate(bicluster_expression_data, start=1):\n",
    "    print(f\"Bicluster {i} Data with Gene Symbols:\")\n",
    "    print(data_with_symbols.head(2))\n",
    "    print(\"\\n\\n\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Apply Support Vector Machine Method for Classification**\n",
    "\n",
    "#### **First Step**\n",
    "*Gene Expression Data do not consist of the target. In order to train the data in the supervised learning process, need to insert the target in the dataset. <br>\n",
    "Hence, we load the dataset that saved in the file, then insert the target to the dataset.* <br>\n",
    "- Target means if the sample is sufferred from esophageal cancer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load Each Dataset\n",
    "bicluster1 = pd.read_csv(\"D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/Bicluster Expression Data 1.csv\")\n",
    "bicluster2 = pd.read_csv(\"D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/Bicluster Expression Data 2.csv\")\n",
    "bicluster3 = pd.read_csv(\"D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/Bicluster Expression Data 3.csv\")\n",
    "bicluster4 = pd.read_csv(\"D:/UTM/20222023 SEM 2/PSM I/FYP/dataset/Bicluster Expression Data 4.csv\")\n",
    "\n",
    "# Transpose the Dataset\n",
    "bicluster1_transpose = bicluster1.set_index('Gene Symbol').transpose()\n",
    "bicluster2_transpose = bicluster2.set_index('Gene Symbol').transpose()\n",
    "bicluster3_transpose = bicluster3.set_index('Gene Symbol').transpose()\n",
    "bicluster4_transpose = bicluster4.set_index('Gene Symbol').transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 1 Dimension:  (10, 124)\n",
      "Bicluster 1: \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>TMEM39A</th>\n",
       "      <th>NUP107</th>\n",
       "      <th>TMEM38B</th>\n",
       "      <th>PBK</th>\n",
       "      <th>ASPM</th>\n",
       "      <th>NELFCD</th>\n",
       "      <th>DNMT3B</th>\n",
       "      <th>TBL1XR1</th>\n",
       "      <th>GINS2</th>\n",
       "      <th>COL5A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509804_E1507T.CEL</th>\n",
       "      <td>6.611383</td>\n",
       "      <td>9.726017</td>\n",
       "      <td>5.445221</td>\n",
       "      <td>7.313210</td>\n",
       "      <td>7.352215</td>\n",
       "      <td>8.965812</td>\n",
       "      <td>6.870785</td>\n",
       "      <td>6.893274</td>\n",
       "      <td>7.324547</td>\n",
       "      <td>9.898811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509809_E1542T.CEL</th>\n",
       "      <td>6.389347</td>\n",
       "      <td>8.623677</td>\n",
       "      <td>4.765128</td>\n",
       "      <td>7.265171</td>\n",
       "      <td>8.088150</td>\n",
       "      <td>8.763703</td>\n",
       "      <td>5.790180</td>\n",
       "      <td>5.530707</td>\n",
       "      <td>7.380882</td>\n",
       "      <td>8.929217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509810_E1546T.CEL</th>\n",
       "      <td>6.724950</td>\n",
       "      <td>8.185211</td>\n",
       "      <td>5.477874</td>\n",
       "      <td>8.210063</td>\n",
       "      <td>7.910408</td>\n",
       "      <td>9.619201</td>\n",
       "      <td>6.787680</td>\n",
       "      <td>7.152694</td>\n",
       "      <td>7.778961</td>\n",
       "      <td>10.010738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <td>6.587234</td>\n",
       "      <td>8.930046</td>\n",
       "      <td>4.597372</td>\n",
       "      <td>8.368051</td>\n",
       "      <td>8.356823</td>\n",
       "      <td>8.400663</td>\n",
       "      <td>6.114453</td>\n",
       "      <td>6.503139</td>\n",
       "      <td>8.567287</td>\n",
       "      <td>8.953696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509813_E1589T.CEL</th>\n",
       "      <td>6.583535</td>\n",
       "      <td>9.160049</td>\n",
       "      <td>6.840389</td>\n",
       "      <td>8.215945</td>\n",
       "      <td>8.576583</td>\n",
       "      <td>9.793823</td>\n",
       "      <td>8.950951</td>\n",
       "      <td>7.235061</td>\n",
       "      <td>9.331082</td>\n",
       "      <td>9.496884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol            TMEM39A    NUP107   TMEM38B       PBK      ASPM  \\\n",
       "GSM509804_E1507T.CEL  6.611383  9.726017  5.445221  7.313210  7.352215   \n",
       "GSM509809_E1542T.CEL  6.389347  8.623677  4.765128  7.265171  8.088150   \n",
       "GSM509810_E1546T.CEL  6.724950  8.185211  5.477874  8.210063  7.910408   \n",
       "GSM509812_E1584T.CEL  6.587234  8.930046  4.597372  8.368051  8.356823   \n",
       "GSM509813_E1589T.CEL  6.583535  9.160049  6.840389  8.215945  8.576583   \n",
       "\n",
       "Gene Symbol             NELFCD    DNMT3B   TBL1XR1     GINS2     COL5A2  \n",
       "GSM509804_E1507T.CEL  8.965812  6.870785  6.893274  7.324547   9.898811  \n",
       "GSM509809_E1542T.CEL  8.763703  5.790180  5.530707  7.380882   8.929217  \n",
       "GSM509810_E1546T.CEL  9.619201  6.787680  7.152694  7.778961  10.010738  \n",
       "GSM509812_E1584T.CEL  8.400663  6.114453  6.503139  8.567287   8.953696  \n",
       "GSM509813_E1589T.CEL  9.793823  8.950951  7.235061  9.331082   9.496884  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the Dimension of the Bicluster 1\n",
    "print(\"Bicluster 1 Dimension: \", bicluster1_transpose.shape)\n",
    "print(\"Bicluster 1: \\n\")\n",
    "\n",
    "# Display the First Five Rows and 114th Column until 124th Column Dataset\n",
    "bicluster1_transpose.iloc[:,114:124].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 2 Dimension:  (14, 143)\n",
      "Bicluster 2: \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>SLC27A6</th>\n",
       "      <th>GIPC2</th>\n",
       "      <th>EPB41L4A</th>\n",
       "      <th>ADAMTSL4</th>\n",
       "      <th>RIPK4</th>\n",
       "      <th>UBAP1</th>\n",
       "      <th>VPS37B</th>\n",
       "      <th>CSNK1E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509803_E2644N.CEL</th>\n",
       "      <td>4.229821</td>\n",
       "      <td>5.761416</td>\n",
       "      <td>5.770840</td>\n",
       "      <td>5.621589</td>\n",
       "      <td>10.616623</td>\n",
       "      <td>8.803706</td>\n",
       "      <td>9.253934</td>\n",
       "      <td>7.787345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509804_E1507T.CEL</th>\n",
       "      <td>4.103199</td>\n",
       "      <td>3.446680</td>\n",
       "      <td>4.413280</td>\n",
       "      <td>5.258764</td>\n",
       "      <td>9.521838</td>\n",
       "      <td>7.682166</td>\n",
       "      <td>6.960880</td>\n",
       "      <td>7.428256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509805_E1520T.CEL</th>\n",
       "      <td>3.881914</td>\n",
       "      <td>3.485505</td>\n",
       "      <td>4.465795</td>\n",
       "      <td>5.194590</td>\n",
       "      <td>7.898191</td>\n",
       "      <td>7.109491</td>\n",
       "      <td>7.278069</td>\n",
       "      <td>7.222471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509806_E1521T.CEL</th>\n",
       "      <td>3.941313</td>\n",
       "      <td>3.378748</td>\n",
       "      <td>4.152978</td>\n",
       "      <td>5.104112</td>\n",
       "      <td>8.974263</td>\n",
       "      <td>7.553683</td>\n",
       "      <td>8.003548</td>\n",
       "      <td>6.968401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509809_E1542T.CEL</th>\n",
       "      <td>4.079763</td>\n",
       "      <td>3.801562</td>\n",
       "      <td>4.456992</td>\n",
       "      <td>5.741768</td>\n",
       "      <td>8.988443</td>\n",
       "      <td>7.181832</td>\n",
       "      <td>8.338481</td>\n",
       "      <td>7.648359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol            SLC27A6     GIPC2  EPB41L4A  ADAMTSL4      RIPK4  \\\n",
       "GSM509803_E2644N.CEL  4.229821  5.761416  5.770840  5.621589  10.616623   \n",
       "GSM509804_E1507T.CEL  4.103199  3.446680  4.413280  5.258764   9.521838   \n",
       "GSM509805_E1520T.CEL  3.881914  3.485505  4.465795  5.194590   7.898191   \n",
       "GSM509806_E1521T.CEL  3.941313  3.378748  4.152978  5.104112   8.974263   \n",
       "GSM509809_E1542T.CEL  4.079763  3.801562  4.456992  5.741768   8.988443   \n",
       "\n",
       "Gene Symbol              UBAP1    VPS37B    CSNK1E  \n",
       "GSM509803_E2644N.CEL  8.803706  9.253934  7.787345  \n",
       "GSM509804_E1507T.CEL  7.682166  6.960880  7.428256  \n",
       "GSM509805_E1520T.CEL  7.109491  7.278069  7.222471  \n",
       "GSM509806_E1521T.CEL  7.553683  8.003548  6.968401  \n",
       "GSM509809_E1542T.CEL  7.181832  8.338481  7.648359  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the Dimension of the Bicluster 2 Dataset\n",
    "print(\"Bicluster 2 Dimension: \", bicluster2_transpose.shape)\n",
    "print(\"Bicluster 2: \\n\")\n",
    "\n",
    "# Display the First Five Rows and 135th Column until 145th Column Dataset\n",
    "bicluster2_transpose.iloc[:,135:145].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 3 Dimension:  (1, 6)\n",
      "Bicluster 3: \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>DVL3</th>\n",
       "      <th>EPHB4</th>\n",
       "      <th>APOC1</th>\n",
       "      <th>LAMB3</th>\n",
       "      <th>HOXD11</th>\n",
       "      <th>ANO1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <td>7.691792</td>\n",
       "      <td>7.516879</td>\n",
       "      <td>8.240716</td>\n",
       "      <td>10.040889</td>\n",
       "      <td>4.567577</td>\n",
       "      <td>4.450293</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol               DVL3     EPHB4     APOC1      LAMB3    HOXD11  \\\n",
       "GSM509819_E1796T.CEL  7.691792  7.516879  8.240716  10.040889  4.567577   \n",
       "\n",
       "Gene Symbol               ANO1  \n",
       "GSM509819_E1796T.CEL  4.450293  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the Dimension of the Bicluster 3 Dataset\n",
    "print(\"Bicluster 3 Dimension: \", bicluster3_transpose.shape)\n",
    "print(\"Bicluster 3: \\n\")\n",
    "\n",
    "# Display the First Five Rows Dataset\n",
    "bicluster3_transpose.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 4 Dimension:  (1, 13)\n",
      "Bicluster 4: \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>NREP</th>\n",
       "      <th>HEXB</th>\n",
       "      <th>EPHB4</th>\n",
       "      <th>KIF3B</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>SAP30</th>\n",
       "      <th>PTK2</th>\n",
       "      <th>LAMB3</th>\n",
       "      <th>MBD4</th>\n",
       "      <th>CHN1</th>\n",
       "      <th>ALMS1</th>\n",
       "      <th>HOXD11</th>\n",
       "      <th>BANP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <td>7.82422</td>\n",
       "      <td>8.918272</td>\n",
       "      <td>7.471044</td>\n",
       "      <td>6.550222</td>\n",
       "      <td>6.12884</td>\n",
       "      <td>5.130646</td>\n",
       "      <td>9.079471</td>\n",
       "      <td>9.57014</td>\n",
       "      <td>8.38678</td>\n",
       "      <td>6.323455</td>\n",
       "      <td>6.883904</td>\n",
       "      <td>5.71802</td>\n",
       "      <td>7.740366</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              NREP      HEXB     EPHB4     KIF3B    BRCA1  \\\n",
       "GSM509816_E1614T.CEL  7.82422  8.918272  7.471044  6.550222  6.12884   \n",
       "\n",
       "Gene Symbol              SAP30      PTK2    LAMB3     MBD4      CHN1  \\\n",
       "GSM509816_E1614T.CEL  5.130646  9.079471  9.57014  8.38678  6.323455   \n",
       "\n",
       "Gene Symbol              ALMS1   HOXD11      BANP  \n",
       "GSM509816_E1614T.CEL  6.883904  5.71802  7.740366  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the Dimension of the Bicluster 4 Dataset\n",
    "print(\"Bicluster 4 Dimension: \", bicluster4_transpose.shape)\n",
    "print(\"Bicluster 4: \\n\")\n",
    "\n",
    "# Display the First Five Rows Dataset\n",
    "bicluster4_transpose.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Second Step**\n",
    "\n",
    "*Assign one more column as Target to identify whether the the sample is sufferred from cancer.* <br><br>\n",
    "**Target**\n",
    "- 1: True: Suffer from Cancer\n",
    "- 0: False: Not Suffer from Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 1 Dimension:  (10, 125)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\4238487191.py:2: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  bicluster1_transpose['Target'] = pd.Series(pd.np.where(bicluster1_transpose.index.isin(sample_names_array), 1, 0), index=bicluster1_transpose.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>AP2M1</th>\n",
       "      <th>PTGES3</th>\n",
       "      <th>LAMC1</th>\n",
       "      <th>PLOD1</th>\n",
       "      <th>PSMD2</th>\n",
       "      <th>SLC3A2</th>\n",
       "      <th>SSRP1</th>\n",
       "      <th>RPN1</th>\n",
       "      <th>MMP2</th>\n",
       "      <th>...</th>\n",
       "      <th>NUP107</th>\n",
       "      <th>TMEM38B</th>\n",
       "      <th>PBK</th>\n",
       "      <th>ASPM</th>\n",
       "      <th>NELFCD</th>\n",
       "      <th>DNMT3B</th>\n",
       "      <th>TBL1XR1</th>\n",
       "      <th>GINS2</th>\n",
       "      <th>COL5A2</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509804_E1507T.CEL</th>\n",
       "      <td>8.220648</td>\n",
       "      <td>10.741044</td>\n",
       "      <td>11.043840</td>\n",
       "      <td>8.875783</td>\n",
       "      <td>8.104684</td>\n",
       "      <td>9.578103</td>\n",
       "      <td>7.971561</td>\n",
       "      <td>7.599398</td>\n",
       "      <td>9.279621</td>\n",
       "      <td>9.319667</td>\n",
       "      <td>...</td>\n",
       "      <td>9.726017</td>\n",
       "      <td>5.445221</td>\n",
       "      <td>7.313210</td>\n",
       "      <td>7.352215</td>\n",
       "      <td>8.965812</td>\n",
       "      <td>6.870785</td>\n",
       "      <td>6.893274</td>\n",
       "      <td>7.324547</td>\n",
       "      <td>9.898811</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509809_E1542T.CEL</th>\n",
       "      <td>7.910793</td>\n",
       "      <td>9.977123</td>\n",
       "      <td>11.498975</td>\n",
       "      <td>7.893135</td>\n",
       "      <td>8.269134</td>\n",
       "      <td>9.720489</td>\n",
       "      <td>8.141545</td>\n",
       "      <td>8.038281</td>\n",
       "      <td>9.388059</td>\n",
       "      <td>8.331845</td>\n",
       "      <td>...</td>\n",
       "      <td>8.623677</td>\n",
       "      <td>4.765128</td>\n",
       "      <td>7.265171</td>\n",
       "      <td>8.088150</td>\n",
       "      <td>8.763703</td>\n",
       "      <td>5.790180</td>\n",
       "      <td>5.530707</td>\n",
       "      <td>7.380882</td>\n",
       "      <td>8.929217</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509810_E1546T.CEL</th>\n",
       "      <td>8.712417</td>\n",
       "      <td>10.212407</td>\n",
       "      <td>11.414298</td>\n",
       "      <td>8.340719</td>\n",
       "      <td>7.301604</td>\n",
       "      <td>10.014823</td>\n",
       "      <td>10.025367</td>\n",
       "      <td>8.580273</td>\n",
       "      <td>9.040079</td>\n",
       "      <td>8.920236</td>\n",
       "      <td>...</td>\n",
       "      <td>8.185211</td>\n",
       "      <td>5.477874</td>\n",
       "      <td>8.210063</td>\n",
       "      <td>7.910408</td>\n",
       "      <td>9.619201</td>\n",
       "      <td>6.787680</td>\n",
       "      <td>7.152694</td>\n",
       "      <td>7.778961</td>\n",
       "      <td>10.010738</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <td>8.613985</td>\n",
       "      <td>10.456746</td>\n",
       "      <td>10.792888</td>\n",
       "      <td>7.912544</td>\n",
       "      <td>7.783395</td>\n",
       "      <td>9.653568</td>\n",
       "      <td>8.551611</td>\n",
       "      <td>8.408757</td>\n",
       "      <td>9.475060</td>\n",
       "      <td>9.526403</td>\n",
       "      <td>...</td>\n",
       "      <td>8.930046</td>\n",
       "      <td>4.597372</td>\n",
       "      <td>8.368051</td>\n",
       "      <td>8.356823</td>\n",
       "      <td>8.400663</td>\n",
       "      <td>6.114453</td>\n",
       "      <td>6.503139</td>\n",
       "      <td>8.567287</td>\n",
       "      <td>8.953696</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509813_E1589T.CEL</th>\n",
       "      <td>9.082743</td>\n",
       "      <td>10.831445</td>\n",
       "      <td>11.875117</td>\n",
       "      <td>8.940777</td>\n",
       "      <td>8.667018</td>\n",
       "      <td>11.177005</td>\n",
       "      <td>9.461601</td>\n",
       "      <td>8.845955</td>\n",
       "      <td>9.455267</td>\n",
       "      <td>9.426712</td>\n",
       "      <td>...</td>\n",
       "      <td>9.160049</td>\n",
       "      <td>6.840389</td>\n",
       "      <td>8.215945</td>\n",
       "      <td>8.576583</td>\n",
       "      <td>9.793823</td>\n",
       "      <td>8.950951</td>\n",
       "      <td>7.235061</td>\n",
       "      <td>9.331082</td>\n",
       "      <td>9.496884</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509815_E1610T.CEL</th>\n",
       "      <td>8.246391</td>\n",
       "      <td>10.036581</td>\n",
       "      <td>11.821380</td>\n",
       "      <td>9.998583</td>\n",
       "      <td>8.720222</td>\n",
       "      <td>10.774002</td>\n",
       "      <td>8.292859</td>\n",
       "      <td>8.535826</td>\n",
       "      <td>9.330133</td>\n",
       "      <td>8.804083</td>\n",
       "      <td>...</td>\n",
       "      <td>10.200163</td>\n",
       "      <td>6.290956</td>\n",
       "      <td>7.325244</td>\n",
       "      <td>9.338510</td>\n",
       "      <td>9.725768</td>\n",
       "      <td>6.838957</td>\n",
       "      <td>8.715199</td>\n",
       "      <td>9.070411</td>\n",
       "      <td>8.867885</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <td>8.755096</td>\n",
       "      <td>10.342455</td>\n",
       "      <td>12.138041</td>\n",
       "      <td>8.796925</td>\n",
       "      <td>7.875919</td>\n",
       "      <td>10.366143</td>\n",
       "      <td>8.922290</td>\n",
       "      <td>8.523105</td>\n",
       "      <td>9.117200</td>\n",
       "      <td>8.346820</td>\n",
       "      <td>...</td>\n",
       "      <td>9.186521</td>\n",
       "      <td>5.895004</td>\n",
       "      <td>8.165070</td>\n",
       "      <td>9.464792</td>\n",
       "      <td>8.760688</td>\n",
       "      <td>6.958920</td>\n",
       "      <td>7.352432</td>\n",
       "      <td>8.375379</td>\n",
       "      <td>10.165669</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509817_E1635T.CEL</th>\n",
       "      <td>8.835079</td>\n",
       "      <td>10.531348</td>\n",
       "      <td>11.979479</td>\n",
       "      <td>8.921878</td>\n",
       "      <td>8.026332</td>\n",
       "      <td>9.837584</td>\n",
       "      <td>9.854667</td>\n",
       "      <td>8.972070</td>\n",
       "      <td>9.360248</td>\n",
       "      <td>9.664517</td>\n",
       "      <td>...</td>\n",
       "      <td>9.105222</td>\n",
       "      <td>5.922400</td>\n",
       "      <td>8.535196</td>\n",
       "      <td>8.996614</td>\n",
       "      <td>9.897316</td>\n",
       "      <td>6.844505</td>\n",
       "      <td>7.660548</td>\n",
       "      <td>8.854069</td>\n",
       "      <td>11.291970</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509818_E1709T.CEL</th>\n",
       "      <td>8.682854</td>\n",
       "      <td>10.527527</td>\n",
       "      <td>11.417330</td>\n",
       "      <td>8.482972</td>\n",
       "      <td>7.609727</td>\n",
       "      <td>10.443150</td>\n",
       "      <td>9.206129</td>\n",
       "      <td>8.702681</td>\n",
       "      <td>9.495476</td>\n",
       "      <td>8.629193</td>\n",
       "      <td>...</td>\n",
       "      <td>9.104914</td>\n",
       "      <td>5.666086</td>\n",
       "      <td>7.930660</td>\n",
       "      <td>8.449101</td>\n",
       "      <td>9.843952</td>\n",
       "      <td>7.864205</td>\n",
       "      <td>6.703331</td>\n",
       "      <td>8.901941</td>\n",
       "      <td>9.793506</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <td>8.672898</td>\n",
       "      <td>10.845476</td>\n",
       "      <td>11.840404</td>\n",
       "      <td>9.446835</td>\n",
       "      <td>8.704273</td>\n",
       "      <td>11.238388</td>\n",
       "      <td>8.894755</td>\n",
       "      <td>8.551992</td>\n",
       "      <td>9.809486</td>\n",
       "      <td>9.194378</td>\n",
       "      <td>...</td>\n",
       "      <td>9.671777</td>\n",
       "      <td>6.246486</td>\n",
       "      <td>7.543486</td>\n",
       "      <td>9.275292</td>\n",
       "      <td>9.300892</td>\n",
       "      <td>7.719712</td>\n",
       "      <td>6.968967</td>\n",
       "      <td>8.497313</td>\n",
       "      <td>9.826365</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 125 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              ABCF1      AP2M1     PTGES3     LAMC1     PLOD1  \\\n",
       "GSM509804_E1507T.CEL  8.220648  10.741044  11.043840  8.875783  8.104684   \n",
       "GSM509809_E1542T.CEL  7.910793   9.977123  11.498975  7.893135  8.269134   \n",
       "GSM509810_E1546T.CEL  8.712417  10.212407  11.414298  8.340719  7.301604   \n",
       "GSM509812_E1584T.CEL  8.613985  10.456746  10.792888  7.912544  7.783395   \n",
       "GSM509813_E1589T.CEL  9.082743  10.831445  11.875117  8.940777  8.667018   \n",
       "GSM509815_E1610T.CEL  8.246391  10.036581  11.821380  9.998583  8.720222   \n",
       "GSM509816_E1614T.CEL  8.755096  10.342455  12.138041  8.796925  7.875919   \n",
       "GSM509817_E1635T.CEL  8.835079  10.531348  11.979479  8.921878  8.026332   \n",
       "GSM509818_E1709T.CEL  8.682854  10.527527  11.417330  8.482972  7.609727   \n",
       "GSM509819_E1796T.CEL  8.672898  10.845476  11.840404  9.446835  8.704273   \n",
       "\n",
       "Gene Symbol               PSMD2     SLC3A2     SSRP1      RPN1      MMP2  ...  \\\n",
       "GSM509804_E1507T.CEL   9.578103   7.971561  7.599398  9.279621  9.319667  ...   \n",
       "GSM509809_E1542T.CEL   9.720489   8.141545  8.038281  9.388059  8.331845  ...   \n",
       "GSM509810_E1546T.CEL  10.014823  10.025367  8.580273  9.040079  8.920236  ...   \n",
       "GSM509812_E1584T.CEL   9.653568   8.551611  8.408757  9.475060  9.526403  ...   \n",
       "GSM509813_E1589T.CEL  11.177005   9.461601  8.845955  9.455267  9.426712  ...   \n",
       "GSM509815_E1610T.CEL  10.774002   8.292859  8.535826  9.330133  8.804083  ...   \n",
       "GSM509816_E1614T.CEL  10.366143   8.922290  8.523105  9.117200  8.346820  ...   \n",
       "GSM509817_E1635T.CEL   9.837584   9.854667  8.972070  9.360248  9.664517  ...   \n",
       "GSM509818_E1709T.CEL  10.443150   9.206129  8.702681  9.495476  8.629193  ...   \n",
       "GSM509819_E1796T.CEL  11.238388   8.894755  8.551992  9.809486  9.194378  ...   \n",
       "\n",
       "Gene Symbol              NUP107   TMEM38B       PBK      ASPM    NELFCD  \\\n",
       "GSM509804_E1507T.CEL   9.726017  5.445221  7.313210  7.352215  8.965812   \n",
       "GSM509809_E1542T.CEL   8.623677  4.765128  7.265171  8.088150  8.763703   \n",
       "GSM509810_E1546T.CEL   8.185211  5.477874  8.210063  7.910408  9.619201   \n",
       "GSM509812_E1584T.CEL   8.930046  4.597372  8.368051  8.356823  8.400663   \n",
       "GSM509813_E1589T.CEL   9.160049  6.840389  8.215945  8.576583  9.793823   \n",
       "GSM509815_E1610T.CEL  10.200163  6.290956  7.325244  9.338510  9.725768   \n",
       "GSM509816_E1614T.CEL   9.186521  5.895004  8.165070  9.464792  8.760688   \n",
       "GSM509817_E1635T.CEL   9.105222  5.922400  8.535196  8.996614  9.897316   \n",
       "GSM509818_E1709T.CEL   9.104914  5.666086  7.930660  8.449101  9.843952   \n",
       "GSM509819_E1796T.CEL   9.671777  6.246486  7.543486  9.275292  9.300892   \n",
       "\n",
       "Gene Symbol             DNMT3B   TBL1XR1     GINS2     COL5A2  Target  \n",
       "GSM509804_E1507T.CEL  6.870785  6.893274  7.324547   9.898811       1  \n",
       "GSM509809_E1542T.CEL  5.790180  5.530707  7.380882   8.929217       1  \n",
       "GSM509810_E1546T.CEL  6.787680  7.152694  7.778961  10.010738       1  \n",
       "GSM509812_E1584T.CEL  6.114453  6.503139  8.567287   8.953696       1  \n",
       "GSM509813_E1589T.CEL  8.950951  7.235061  9.331082   9.496884       1  \n",
       "GSM509815_E1610T.CEL  6.838957  8.715199  9.070411   8.867885       1  \n",
       "GSM509816_E1614T.CEL  6.958920  7.352432  8.375379  10.165669       1  \n",
       "GSM509817_E1635T.CEL  6.844505  7.660548  8.854069  11.291970       1  \n",
       "GSM509818_E1709T.CEL  7.864205  6.703331  8.901941   9.793506       1  \n",
       "GSM509819_E1796T.CEL  7.719712  6.968967  8.497313   9.826365       1  \n",
       "\n",
       "[10 rows x 125 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "bicluster1_transpose['Target'] = pd.Series(pd.np.where(bicluster1_transpose.index.isin(sample_names_array), 1, 0), index=bicluster1_transpose.index)\n",
    "\n",
    "# Check the Dimension of the Bicluster 1 Dataset and Display it\n",
    "print(\"Bicluster 1 Dimension: \", bicluster1_transpose.shape)\n",
    "bicluster1_transpose\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>NUP107</th>\n",
       "      <th>TMEM38B</th>\n",
       "      <th>PBK</th>\n",
       "      <th>ASPM</th>\n",
       "      <th>NELFCD</th>\n",
       "      <th>DNMT3B</th>\n",
       "      <th>TBL1XR1</th>\n",
       "      <th>GINS2</th>\n",
       "      <th>COL5A2</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509804_E1507T.CEL</th>\n",
       "      <td>9.726017</td>\n",
       "      <td>5.445221</td>\n",
       "      <td>7.313210</td>\n",
       "      <td>7.352215</td>\n",
       "      <td>8.965812</td>\n",
       "      <td>6.870785</td>\n",
       "      <td>6.893274</td>\n",
       "      <td>7.324547</td>\n",
       "      <td>9.898811</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509809_E1542T.CEL</th>\n",
       "      <td>8.623677</td>\n",
       "      <td>4.765128</td>\n",
       "      <td>7.265171</td>\n",
       "      <td>8.088150</td>\n",
       "      <td>8.763703</td>\n",
       "      <td>5.790180</td>\n",
       "      <td>5.530707</td>\n",
       "      <td>7.380882</td>\n",
       "      <td>8.929217</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509810_E1546T.CEL</th>\n",
       "      <td>8.185211</td>\n",
       "      <td>5.477874</td>\n",
       "      <td>8.210063</td>\n",
       "      <td>7.910408</td>\n",
       "      <td>9.619201</td>\n",
       "      <td>6.787680</td>\n",
       "      <td>7.152694</td>\n",
       "      <td>7.778961</td>\n",
       "      <td>10.010738</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <td>8.930046</td>\n",
       "      <td>4.597372</td>\n",
       "      <td>8.368051</td>\n",
       "      <td>8.356823</td>\n",
       "      <td>8.400663</td>\n",
       "      <td>6.114453</td>\n",
       "      <td>6.503139</td>\n",
       "      <td>8.567287</td>\n",
       "      <td>8.953696</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509813_E1589T.CEL</th>\n",
       "      <td>9.160049</td>\n",
       "      <td>6.840389</td>\n",
       "      <td>8.215945</td>\n",
       "      <td>8.576583</td>\n",
       "      <td>9.793823</td>\n",
       "      <td>8.950951</td>\n",
       "      <td>7.235061</td>\n",
       "      <td>9.331082</td>\n",
       "      <td>9.496884</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509815_E1610T.CEL</th>\n",
       "      <td>10.200163</td>\n",
       "      <td>6.290956</td>\n",
       "      <td>7.325244</td>\n",
       "      <td>9.338510</td>\n",
       "      <td>9.725768</td>\n",
       "      <td>6.838957</td>\n",
       "      <td>8.715199</td>\n",
       "      <td>9.070411</td>\n",
       "      <td>8.867885</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <td>9.186521</td>\n",
       "      <td>5.895004</td>\n",
       "      <td>8.165070</td>\n",
       "      <td>9.464792</td>\n",
       "      <td>8.760688</td>\n",
       "      <td>6.958920</td>\n",
       "      <td>7.352432</td>\n",
       "      <td>8.375379</td>\n",
       "      <td>10.165669</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509817_E1635T.CEL</th>\n",
       "      <td>9.105222</td>\n",
       "      <td>5.922400</td>\n",
       "      <td>8.535196</td>\n",
       "      <td>8.996614</td>\n",
       "      <td>9.897316</td>\n",
       "      <td>6.844505</td>\n",
       "      <td>7.660548</td>\n",
       "      <td>8.854069</td>\n",
       "      <td>11.291970</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509818_E1709T.CEL</th>\n",
       "      <td>9.104914</td>\n",
       "      <td>5.666086</td>\n",
       "      <td>7.930660</td>\n",
       "      <td>8.449101</td>\n",
       "      <td>9.843952</td>\n",
       "      <td>7.864205</td>\n",
       "      <td>6.703331</td>\n",
       "      <td>8.901941</td>\n",
       "      <td>9.793506</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <td>9.671777</td>\n",
       "      <td>6.246486</td>\n",
       "      <td>7.543486</td>\n",
       "      <td>9.275292</td>\n",
       "      <td>9.300892</td>\n",
       "      <td>7.719712</td>\n",
       "      <td>6.968967</td>\n",
       "      <td>8.497313</td>\n",
       "      <td>9.826365</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              NUP107   TMEM38B       PBK      ASPM    NELFCD  \\\n",
       "GSM509804_E1507T.CEL   9.726017  5.445221  7.313210  7.352215  8.965812   \n",
       "GSM509809_E1542T.CEL   8.623677  4.765128  7.265171  8.088150  8.763703   \n",
       "GSM509810_E1546T.CEL   8.185211  5.477874  8.210063  7.910408  9.619201   \n",
       "GSM509812_E1584T.CEL   8.930046  4.597372  8.368051  8.356823  8.400663   \n",
       "GSM509813_E1589T.CEL   9.160049  6.840389  8.215945  8.576583  9.793823   \n",
       "GSM509815_E1610T.CEL  10.200163  6.290956  7.325244  9.338510  9.725768   \n",
       "GSM509816_E1614T.CEL   9.186521  5.895004  8.165070  9.464792  8.760688   \n",
       "GSM509817_E1635T.CEL   9.105222  5.922400  8.535196  8.996614  9.897316   \n",
       "GSM509818_E1709T.CEL   9.104914  5.666086  7.930660  8.449101  9.843952   \n",
       "GSM509819_E1796T.CEL   9.671777  6.246486  7.543486  9.275292  9.300892   \n",
       "\n",
       "Gene Symbol             DNMT3B   TBL1XR1     GINS2     COL5A2  Target  \n",
       "GSM509804_E1507T.CEL  6.870785  6.893274  7.324547   9.898811       1  \n",
       "GSM509809_E1542T.CEL  5.790180  5.530707  7.380882   8.929217       1  \n",
       "GSM509810_E1546T.CEL  6.787680  7.152694  7.778961  10.010738       1  \n",
       "GSM509812_E1584T.CEL  6.114453  6.503139  8.567287   8.953696       1  \n",
       "GSM509813_E1589T.CEL  8.950951  7.235061  9.331082   9.496884       1  \n",
       "GSM509815_E1610T.CEL  6.838957  8.715199  9.070411   8.867885       1  \n",
       "GSM509816_E1614T.CEL  6.958920  7.352432  8.375379  10.165669       1  \n",
       "GSM509817_E1635T.CEL  6.844505  7.660548  8.854069  11.291970       1  \n",
       "GSM509818_E1709T.CEL  7.864205  6.703331  8.901941   9.793506       1  \n",
       "GSM509819_E1796T.CEL  7.719712  6.968967  8.497313   9.826365       1  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bicluster1_transpose.iloc[:,115:125]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 2 Dimension:  (14, 144)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\2589442151.py:2: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  bicluster2_transpose['Target'] = pd.Series(pd.np.where(bicluster2_transpose.index.isin(sample_names_array), 1, 0), index=bicluster2_transpose.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>PLD1</th>\n",
       "      <th>CAPNS1</th>\n",
       "      <th>RHOA</th>\n",
       "      <th>GUK1</th>\n",
       "      <th>CAPN1</th>\n",
       "      <th>ATP5O</th>\n",
       "      <th>USP9X</th>\n",
       "      <th>GNAI3</th>\n",
       "      <th>RAB5B</th>\n",
       "      <th>ALDH2</th>\n",
       "      <th>...</th>\n",
       "      <th>RNF39</th>\n",
       "      <th>SLC27A6</th>\n",
       "      <th>GIPC2</th>\n",
       "      <th>EPB41L4A</th>\n",
       "      <th>ADAMTSL4</th>\n",
       "      <th>RIPK4</th>\n",
       "      <th>UBAP1</th>\n",
       "      <th>VPS37B</th>\n",
       "      <th>CSNK1E</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509803_E2644N.CEL</th>\n",
       "      <td>6.881787</td>\n",
       "      <td>11.487255</td>\n",
       "      <td>11.498704</td>\n",
       "      <td>8.070603</td>\n",
       "      <td>9.974100</td>\n",
       "      <td>10.315923</td>\n",
       "      <td>9.258463</td>\n",
       "      <td>9.336316</td>\n",
       "      <td>9.404959</td>\n",
       "      <td>9.865762</td>\n",
       "      <td>...</td>\n",
       "      <td>8.580828</td>\n",
       "      <td>4.229821</td>\n",
       "      <td>5.761416</td>\n",
       "      <td>5.770840</td>\n",
       "      <td>5.621589</td>\n",
       "      <td>10.616623</td>\n",
       "      <td>8.803706</td>\n",
       "      <td>9.253934</td>\n",
       "      <td>7.787345</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509804_E1507T.CEL</th>\n",
       "      <td>5.259621</td>\n",
       "      <td>11.683634</td>\n",
       "      <td>9.443730</td>\n",
       "      <td>6.925202</td>\n",
       "      <td>8.369743</td>\n",
       "      <td>9.552479</td>\n",
       "      <td>8.801151</td>\n",
       "      <td>7.465446</td>\n",
       "      <td>8.300172</td>\n",
       "      <td>8.940402</td>\n",
       "      <td>...</td>\n",
       "      <td>6.716919</td>\n",
       "      <td>4.103199</td>\n",
       "      <td>3.446680</td>\n",
       "      <td>4.413280</td>\n",
       "      <td>5.258764</td>\n",
       "      <td>9.521838</td>\n",
       "      <td>7.682166</td>\n",
       "      <td>6.960880</td>\n",
       "      <td>7.428256</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509805_E1520T.CEL</th>\n",
       "      <td>6.110842</td>\n",
       "      <td>9.717346</td>\n",
       "      <td>10.077566</td>\n",
       "      <td>7.342639</td>\n",
       "      <td>8.177127</td>\n",
       "      <td>9.663726</td>\n",
       "      <td>7.916811</td>\n",
       "      <td>7.574082</td>\n",
       "      <td>7.407111</td>\n",
       "      <td>8.524075</td>\n",
       "      <td>...</td>\n",
       "      <td>5.681840</td>\n",
       "      <td>3.881914</td>\n",
       "      <td>3.485505</td>\n",
       "      <td>4.465795</td>\n",
       "      <td>5.194590</td>\n",
       "      <td>7.898191</td>\n",
       "      <td>7.109491</td>\n",
       "      <td>7.278069</td>\n",
       "      <td>7.222471</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509806_E1521T.CEL</th>\n",
       "      <td>6.332930</td>\n",
       "      <td>9.443400</td>\n",
       "      <td>9.096021</td>\n",
       "      <td>7.187553</td>\n",
       "      <td>8.717012</td>\n",
       "      <td>9.768274</td>\n",
       "      <td>8.767959</td>\n",
       "      <td>7.793492</td>\n",
       "      <td>7.691034</td>\n",
       "      <td>9.435296</td>\n",
       "      <td>...</td>\n",
       "      <td>5.902807</td>\n",
       "      <td>3.941313</td>\n",
       "      <td>3.378748</td>\n",
       "      <td>4.152978</td>\n",
       "      <td>5.104112</td>\n",
       "      <td>8.974263</td>\n",
       "      <td>7.553683</td>\n",
       "      <td>8.003548</td>\n",
       "      <td>6.968401</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509809_E1542T.CEL</th>\n",
       "      <td>6.353203</td>\n",
       "      <td>9.738445</td>\n",
       "      <td>9.459102</td>\n",
       "      <td>7.484665</td>\n",
       "      <td>8.800069</td>\n",
       "      <td>9.412405</td>\n",
       "      <td>8.629101</td>\n",
       "      <td>7.579606</td>\n",
       "      <td>8.580519</td>\n",
       "      <td>8.151028</td>\n",
       "      <td>...</td>\n",
       "      <td>6.493334</td>\n",
       "      <td>4.079763</td>\n",
       "      <td>3.801562</td>\n",
       "      <td>4.456992</td>\n",
       "      <td>5.741768</td>\n",
       "      <td>8.988443</td>\n",
       "      <td>7.181832</td>\n",
       "      <td>8.338481</td>\n",
       "      <td>7.648359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509810_E1546T.CEL</th>\n",
       "      <td>5.717247</td>\n",
       "      <td>10.297885</td>\n",
       "      <td>10.361707</td>\n",
       "      <td>7.254365</td>\n",
       "      <td>9.434099</td>\n",
       "      <td>9.599810</td>\n",
       "      <td>8.760746</td>\n",
       "      <td>7.900863</td>\n",
       "      <td>7.341538</td>\n",
       "      <td>9.132426</td>\n",
       "      <td>...</td>\n",
       "      <td>6.252399</td>\n",
       "      <td>3.730487</td>\n",
       "      <td>3.906003</td>\n",
       "      <td>4.328748</td>\n",
       "      <td>5.218253</td>\n",
       "      <td>9.324331</td>\n",
       "      <td>7.611860</td>\n",
       "      <td>6.452536</td>\n",
       "      <td>7.171389</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509811_E1566T.CEL</th>\n",
       "      <td>5.599669</td>\n",
       "      <td>9.856282</td>\n",
       "      <td>10.434216</td>\n",
       "      <td>7.455711</td>\n",
       "      <td>7.886432</td>\n",
       "      <td>9.479757</td>\n",
       "      <td>9.130925</td>\n",
       "      <td>7.563768</td>\n",
       "      <td>8.248977</td>\n",
       "      <td>8.715156</td>\n",
       "      <td>...</td>\n",
       "      <td>5.428539</td>\n",
       "      <td>3.798422</td>\n",
       "      <td>3.706377</td>\n",
       "      <td>4.186752</td>\n",
       "      <td>4.925162</td>\n",
       "      <td>8.660770</td>\n",
       "      <td>7.062897</td>\n",
       "      <td>6.387506</td>\n",
       "      <td>7.272552</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <td>5.322972</td>\n",
       "      <td>10.251217</td>\n",
       "      <td>10.142577</td>\n",
       "      <td>7.391777</td>\n",
       "      <td>9.743674</td>\n",
       "      <td>9.707477</td>\n",
       "      <td>8.814641</td>\n",
       "      <td>7.150979</td>\n",
       "      <td>8.148660</td>\n",
       "      <td>9.505728</td>\n",
       "      <td>...</td>\n",
       "      <td>5.950275</td>\n",
       "      <td>3.698107</td>\n",
       "      <td>3.992634</td>\n",
       "      <td>4.286096</td>\n",
       "      <td>5.046298</td>\n",
       "      <td>9.978480</td>\n",
       "      <td>7.064077</td>\n",
       "      <td>7.659996</td>\n",
       "      <td>7.070365</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509814_E1603T.CEL</th>\n",
       "      <td>5.563821</td>\n",
       "      <td>10.479287</td>\n",
       "      <td>9.889366</td>\n",
       "      <td>7.522050</td>\n",
       "      <td>8.880853</td>\n",
       "      <td>9.696858</td>\n",
       "      <td>8.969091</td>\n",
       "      <td>8.196075</td>\n",
       "      <td>8.098303</td>\n",
       "      <td>9.295092</td>\n",
       "      <td>...</td>\n",
       "      <td>5.662010</td>\n",
       "      <td>3.787020</td>\n",
       "      <td>4.987507</td>\n",
       "      <td>4.451226</td>\n",
       "      <td>4.990375</td>\n",
       "      <td>10.032413</td>\n",
       "      <td>8.168227</td>\n",
       "      <td>8.026121</td>\n",
       "      <td>7.377919</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509815_E1610T.CEL</th>\n",
       "      <td>5.048422</td>\n",
       "      <td>9.159860</td>\n",
       "      <td>10.187868</td>\n",
       "      <td>6.831804</td>\n",
       "      <td>8.183162</td>\n",
       "      <td>9.409882</td>\n",
       "      <td>8.863978</td>\n",
       "      <td>8.332505</td>\n",
       "      <td>7.472178</td>\n",
       "      <td>9.383773</td>\n",
       "      <td>...</td>\n",
       "      <td>5.830913</td>\n",
       "      <td>4.241176</td>\n",
       "      <td>3.991872</td>\n",
       "      <td>3.721478</td>\n",
       "      <td>4.925162</td>\n",
       "      <td>8.234292</td>\n",
       "      <td>6.871717</td>\n",
       "      <td>6.611458</td>\n",
       "      <td>6.552709</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <td>5.617596</td>\n",
       "      <td>9.994437</td>\n",
       "      <td>10.154259</td>\n",
       "      <td>7.371730</td>\n",
       "      <td>8.592445</td>\n",
       "      <td>10.153807</td>\n",
       "      <td>9.080246</td>\n",
       "      <td>8.209631</td>\n",
       "      <td>8.335656</td>\n",
       "      <td>9.125454</td>\n",
       "      <td>...</td>\n",
       "      <td>6.049930</td>\n",
       "      <td>3.574875</td>\n",
       "      <td>3.648740</td>\n",
       "      <td>4.196909</td>\n",
       "      <td>5.024340</td>\n",
       "      <td>8.930225</td>\n",
       "      <td>7.779759</td>\n",
       "      <td>7.651306</td>\n",
       "      <td>7.266086</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509817_E1635T.CEL</th>\n",
       "      <td>5.885284</td>\n",
       "      <td>9.873228</td>\n",
       "      <td>10.915969</td>\n",
       "      <td>7.243826</td>\n",
       "      <td>8.603635</td>\n",
       "      <td>9.531908</td>\n",
       "      <td>8.422629</td>\n",
       "      <td>8.518374</td>\n",
       "      <td>7.847211</td>\n",
       "      <td>9.370655</td>\n",
       "      <td>...</td>\n",
       "      <td>6.446707</td>\n",
       "      <td>3.693587</td>\n",
       "      <td>3.531805</td>\n",
       "      <td>4.062639</td>\n",
       "      <td>5.198883</td>\n",
       "      <td>8.526820</td>\n",
       "      <td>8.272057</td>\n",
       "      <td>7.553935</td>\n",
       "      <td>6.904186</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509818_E1709T.CEL</th>\n",
       "      <td>6.386359</td>\n",
       "      <td>9.720860</td>\n",
       "      <td>9.871655</td>\n",
       "      <td>6.886891</td>\n",
       "      <td>8.435622</td>\n",
       "      <td>9.536607</td>\n",
       "      <td>8.516104</td>\n",
       "      <td>7.736674</td>\n",
       "      <td>8.392854</td>\n",
       "      <td>7.786360</td>\n",
       "      <td>...</td>\n",
       "      <td>6.393301</td>\n",
       "      <td>3.741121</td>\n",
       "      <td>4.004577</td>\n",
       "      <td>4.267456</td>\n",
       "      <td>4.916629</td>\n",
       "      <td>8.669479</td>\n",
       "      <td>7.874860</td>\n",
       "      <td>7.918919</td>\n",
       "      <td>6.861825</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <td>4.944501</td>\n",
       "      <td>9.232219</td>\n",
       "      <td>10.537199</td>\n",
       "      <td>7.685813</td>\n",
       "      <td>8.125677</td>\n",
       "      <td>9.735254</td>\n",
       "      <td>9.177735</td>\n",
       "      <td>7.944521</td>\n",
       "      <td>7.694214</td>\n",
       "      <td>9.192314</td>\n",
       "      <td>...</td>\n",
       "      <td>5.434447</td>\n",
       "      <td>3.656964</td>\n",
       "      <td>5.455840</td>\n",
       "      <td>4.181800</td>\n",
       "      <td>5.377791</td>\n",
       "      <td>7.570866</td>\n",
       "      <td>7.222414</td>\n",
       "      <td>7.007866</td>\n",
       "      <td>6.941757</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 144 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol               PLD1     CAPNS1       RHOA      GUK1     CAPN1  \\\n",
       "GSM509803_E2644N.CEL  6.881787  11.487255  11.498704  8.070603  9.974100   \n",
       "GSM509804_E1507T.CEL  5.259621  11.683634   9.443730  6.925202  8.369743   \n",
       "GSM509805_E1520T.CEL  6.110842   9.717346  10.077566  7.342639  8.177127   \n",
       "GSM509806_E1521T.CEL  6.332930   9.443400   9.096021  7.187553  8.717012   \n",
       "GSM509809_E1542T.CEL  6.353203   9.738445   9.459102  7.484665  8.800069   \n",
       "GSM509810_E1546T.CEL  5.717247  10.297885  10.361707  7.254365  9.434099   \n",
       "GSM509811_E1566T.CEL  5.599669   9.856282  10.434216  7.455711  7.886432   \n",
       "GSM509812_E1584T.CEL  5.322972  10.251217  10.142577  7.391777  9.743674   \n",
       "GSM509814_E1603T.CEL  5.563821  10.479287   9.889366  7.522050  8.880853   \n",
       "GSM509815_E1610T.CEL  5.048422   9.159860  10.187868  6.831804  8.183162   \n",
       "GSM509816_E1614T.CEL  5.617596   9.994437  10.154259  7.371730  8.592445   \n",
       "GSM509817_E1635T.CEL  5.885284   9.873228  10.915969  7.243826  8.603635   \n",
       "GSM509818_E1709T.CEL  6.386359   9.720860   9.871655  6.886891  8.435622   \n",
       "GSM509819_E1796T.CEL  4.944501   9.232219  10.537199  7.685813  8.125677   \n",
       "\n",
       "Gene Symbol               ATP5O     USP9X     GNAI3     RAB5B     ALDH2  ...  \\\n",
       "GSM509803_E2644N.CEL  10.315923  9.258463  9.336316  9.404959  9.865762  ...   \n",
       "GSM509804_E1507T.CEL   9.552479  8.801151  7.465446  8.300172  8.940402  ...   \n",
       "GSM509805_E1520T.CEL   9.663726  7.916811  7.574082  7.407111  8.524075  ...   \n",
       "GSM509806_E1521T.CEL   9.768274  8.767959  7.793492  7.691034  9.435296  ...   \n",
       "GSM509809_E1542T.CEL   9.412405  8.629101  7.579606  8.580519  8.151028  ...   \n",
       "GSM509810_E1546T.CEL   9.599810  8.760746  7.900863  7.341538  9.132426  ...   \n",
       "GSM509811_E1566T.CEL   9.479757  9.130925  7.563768  8.248977  8.715156  ...   \n",
       "GSM509812_E1584T.CEL   9.707477  8.814641  7.150979  8.148660  9.505728  ...   \n",
       "GSM509814_E1603T.CEL   9.696858  8.969091  8.196075  8.098303  9.295092  ...   \n",
       "GSM509815_E1610T.CEL   9.409882  8.863978  8.332505  7.472178  9.383773  ...   \n",
       "GSM509816_E1614T.CEL  10.153807  9.080246  8.209631  8.335656  9.125454  ...   \n",
       "GSM509817_E1635T.CEL   9.531908  8.422629  8.518374  7.847211  9.370655  ...   \n",
       "GSM509818_E1709T.CEL   9.536607  8.516104  7.736674  8.392854  7.786360  ...   \n",
       "GSM509819_E1796T.CEL   9.735254  9.177735  7.944521  7.694214  9.192314  ...   \n",
       "\n",
       "Gene Symbol              RNF39   SLC27A6     GIPC2  EPB41L4A  ADAMTSL4  \\\n",
       "GSM509803_E2644N.CEL  8.580828  4.229821  5.761416  5.770840  5.621589   \n",
       "GSM509804_E1507T.CEL  6.716919  4.103199  3.446680  4.413280  5.258764   \n",
       "GSM509805_E1520T.CEL  5.681840  3.881914  3.485505  4.465795  5.194590   \n",
       "GSM509806_E1521T.CEL  5.902807  3.941313  3.378748  4.152978  5.104112   \n",
       "GSM509809_E1542T.CEL  6.493334  4.079763  3.801562  4.456992  5.741768   \n",
       "GSM509810_E1546T.CEL  6.252399  3.730487  3.906003  4.328748  5.218253   \n",
       "GSM509811_E1566T.CEL  5.428539  3.798422  3.706377  4.186752  4.925162   \n",
       "GSM509812_E1584T.CEL  5.950275  3.698107  3.992634  4.286096  5.046298   \n",
       "GSM509814_E1603T.CEL  5.662010  3.787020  4.987507  4.451226  4.990375   \n",
       "GSM509815_E1610T.CEL  5.830913  4.241176  3.991872  3.721478  4.925162   \n",
       "GSM509816_E1614T.CEL  6.049930  3.574875  3.648740  4.196909  5.024340   \n",
       "GSM509817_E1635T.CEL  6.446707  3.693587  3.531805  4.062639  5.198883   \n",
       "GSM509818_E1709T.CEL  6.393301  3.741121  4.004577  4.267456  4.916629   \n",
       "GSM509819_E1796T.CEL  5.434447  3.656964  5.455840  4.181800  5.377791   \n",
       "\n",
       "Gene Symbol               RIPK4     UBAP1    VPS37B    CSNK1E  Target  \n",
       "GSM509803_E2644N.CEL  10.616623  8.803706  9.253934  7.787345       0  \n",
       "GSM509804_E1507T.CEL   9.521838  7.682166  6.960880  7.428256       1  \n",
       "GSM509805_E1520T.CEL   7.898191  7.109491  7.278069  7.222471       1  \n",
       "GSM509806_E1521T.CEL   8.974263  7.553683  8.003548  6.968401       1  \n",
       "GSM509809_E1542T.CEL   8.988443  7.181832  8.338481  7.648359       1  \n",
       "GSM509810_E1546T.CEL   9.324331  7.611860  6.452536  7.171389       1  \n",
       "GSM509811_E1566T.CEL   8.660770  7.062897  6.387506  7.272552       1  \n",
       "GSM509812_E1584T.CEL   9.978480  7.064077  7.659996  7.070365       1  \n",
       "GSM509814_E1603T.CEL  10.032413  8.168227  8.026121  7.377919       1  \n",
       "GSM509815_E1610T.CEL   8.234292  6.871717  6.611458  6.552709       1  \n",
       "GSM509816_E1614T.CEL   8.930225  7.779759  7.651306  7.266086       1  \n",
       "GSM509817_E1635T.CEL   8.526820  8.272057  7.553935  6.904186       1  \n",
       "GSM509818_E1709T.CEL   8.669479  7.874860  7.918919  6.861825       1  \n",
       "GSM509819_E1796T.CEL   7.570866  7.222414  7.007866  6.941757       1  \n",
       "\n",
       "[14 rows x 144 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "bicluster2_transpose['Target'] = pd.Series(pd.np.where(bicluster2_transpose.index.isin(sample_names_array), 1, 0), index=bicluster2_transpose.index)\n",
    "\n",
    "# Check the Dimension of the Bicluster 2 Dataset and Display it\n",
    "print(\"Bicluster 2 Dimension: \", bicluster2_transpose.shape)\n",
    "bicluster2_transpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>GIPC2</th>\n",
       "      <th>EPB41L4A</th>\n",
       "      <th>ADAMTSL4</th>\n",
       "      <th>RIPK4</th>\n",
       "      <th>UBAP1</th>\n",
       "      <th>VPS37B</th>\n",
       "      <th>CSNK1E</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509803_E2644N.CEL</th>\n",
       "      <td>5.761416</td>\n",
       "      <td>5.770840</td>\n",
       "      <td>5.621589</td>\n",
       "      <td>10.616623</td>\n",
       "      <td>8.803706</td>\n",
       "      <td>9.253934</td>\n",
       "      <td>7.787345</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509804_E1507T.CEL</th>\n",
       "      <td>3.446680</td>\n",
       "      <td>4.413280</td>\n",
       "      <td>5.258764</td>\n",
       "      <td>9.521838</td>\n",
       "      <td>7.682166</td>\n",
       "      <td>6.960880</td>\n",
       "      <td>7.428256</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509805_E1520T.CEL</th>\n",
       "      <td>3.485505</td>\n",
       "      <td>4.465795</td>\n",
       "      <td>5.194590</td>\n",
       "      <td>7.898191</td>\n",
       "      <td>7.109491</td>\n",
       "      <td>7.278069</td>\n",
       "      <td>7.222471</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509806_E1521T.CEL</th>\n",
       "      <td>3.378748</td>\n",
       "      <td>4.152978</td>\n",
       "      <td>5.104112</td>\n",
       "      <td>8.974263</td>\n",
       "      <td>7.553683</td>\n",
       "      <td>8.003548</td>\n",
       "      <td>6.968401</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509809_E1542T.CEL</th>\n",
       "      <td>3.801562</td>\n",
       "      <td>4.456992</td>\n",
       "      <td>5.741768</td>\n",
       "      <td>8.988443</td>\n",
       "      <td>7.181832</td>\n",
       "      <td>8.338481</td>\n",
       "      <td>7.648359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509810_E1546T.CEL</th>\n",
       "      <td>3.906003</td>\n",
       "      <td>4.328748</td>\n",
       "      <td>5.218253</td>\n",
       "      <td>9.324331</td>\n",
       "      <td>7.611860</td>\n",
       "      <td>6.452536</td>\n",
       "      <td>7.171389</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509811_E1566T.CEL</th>\n",
       "      <td>3.706377</td>\n",
       "      <td>4.186752</td>\n",
       "      <td>4.925162</td>\n",
       "      <td>8.660770</td>\n",
       "      <td>7.062897</td>\n",
       "      <td>6.387506</td>\n",
       "      <td>7.272552</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509812_E1584T.CEL</th>\n",
       "      <td>3.992634</td>\n",
       "      <td>4.286096</td>\n",
       "      <td>5.046298</td>\n",
       "      <td>9.978480</td>\n",
       "      <td>7.064077</td>\n",
       "      <td>7.659996</td>\n",
       "      <td>7.070365</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509814_E1603T.CEL</th>\n",
       "      <td>4.987507</td>\n",
       "      <td>4.451226</td>\n",
       "      <td>4.990375</td>\n",
       "      <td>10.032413</td>\n",
       "      <td>8.168227</td>\n",
       "      <td>8.026121</td>\n",
       "      <td>7.377919</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509815_E1610T.CEL</th>\n",
       "      <td>3.991872</td>\n",
       "      <td>3.721478</td>\n",
       "      <td>4.925162</td>\n",
       "      <td>8.234292</td>\n",
       "      <td>6.871717</td>\n",
       "      <td>6.611458</td>\n",
       "      <td>6.552709</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <td>3.648740</td>\n",
       "      <td>4.196909</td>\n",
       "      <td>5.024340</td>\n",
       "      <td>8.930225</td>\n",
       "      <td>7.779759</td>\n",
       "      <td>7.651306</td>\n",
       "      <td>7.266086</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509817_E1635T.CEL</th>\n",
       "      <td>3.531805</td>\n",
       "      <td>4.062639</td>\n",
       "      <td>5.198883</td>\n",
       "      <td>8.526820</td>\n",
       "      <td>8.272057</td>\n",
       "      <td>7.553935</td>\n",
       "      <td>6.904186</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509818_E1709T.CEL</th>\n",
       "      <td>4.004577</td>\n",
       "      <td>4.267456</td>\n",
       "      <td>4.916629</td>\n",
       "      <td>8.669479</td>\n",
       "      <td>7.874860</td>\n",
       "      <td>7.918919</td>\n",
       "      <td>6.861825</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <td>5.455840</td>\n",
       "      <td>4.181800</td>\n",
       "      <td>5.377791</td>\n",
       "      <td>7.570866</td>\n",
       "      <td>7.222414</td>\n",
       "      <td>7.007866</td>\n",
       "      <td>6.941757</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              GIPC2  EPB41L4A  ADAMTSL4      RIPK4     UBAP1  \\\n",
       "GSM509803_E2644N.CEL  5.761416  5.770840  5.621589  10.616623  8.803706   \n",
       "GSM509804_E1507T.CEL  3.446680  4.413280  5.258764   9.521838  7.682166   \n",
       "GSM509805_E1520T.CEL  3.485505  4.465795  5.194590   7.898191  7.109491   \n",
       "GSM509806_E1521T.CEL  3.378748  4.152978  5.104112   8.974263  7.553683   \n",
       "GSM509809_E1542T.CEL  3.801562  4.456992  5.741768   8.988443  7.181832   \n",
       "GSM509810_E1546T.CEL  3.906003  4.328748  5.218253   9.324331  7.611860   \n",
       "GSM509811_E1566T.CEL  3.706377  4.186752  4.925162   8.660770  7.062897   \n",
       "GSM509812_E1584T.CEL  3.992634  4.286096  5.046298   9.978480  7.064077   \n",
       "GSM509814_E1603T.CEL  4.987507  4.451226  4.990375  10.032413  8.168227   \n",
       "GSM509815_E1610T.CEL  3.991872  3.721478  4.925162   8.234292  6.871717   \n",
       "GSM509816_E1614T.CEL  3.648740  4.196909  5.024340   8.930225  7.779759   \n",
       "GSM509817_E1635T.CEL  3.531805  4.062639  5.198883   8.526820  8.272057   \n",
       "GSM509818_E1709T.CEL  4.004577  4.267456  4.916629   8.669479  7.874860   \n",
       "GSM509819_E1796T.CEL  5.455840  4.181800  5.377791   7.570866  7.222414   \n",
       "\n",
       "Gene Symbol             VPS37B    CSNK1E  Target  \n",
       "GSM509803_E2644N.CEL  9.253934  7.787345       0  \n",
       "GSM509804_E1507T.CEL  6.960880  7.428256       1  \n",
       "GSM509805_E1520T.CEL  7.278069  7.222471       1  \n",
       "GSM509806_E1521T.CEL  8.003548  6.968401       1  \n",
       "GSM509809_E1542T.CEL  8.338481  7.648359       1  \n",
       "GSM509810_E1546T.CEL  6.452536  7.171389       1  \n",
       "GSM509811_E1566T.CEL  6.387506  7.272552       1  \n",
       "GSM509812_E1584T.CEL  7.659996  7.070365       1  \n",
       "GSM509814_E1603T.CEL  8.026121  7.377919       1  \n",
       "GSM509815_E1610T.CEL  6.611458  6.552709       1  \n",
       "GSM509816_E1614T.CEL  7.651306  7.266086       1  \n",
       "GSM509817_E1635T.CEL  7.553935  6.904186       1  \n",
       "GSM509818_E1709T.CEL  7.918919  6.861825       1  \n",
       "GSM509819_E1796T.CEL  7.007866  6.941757       1  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bicluster2_transpose.iloc[:,136:146]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 3 Dimension:  (1, 7)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\785164208.py:2: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  bicluster3_transpose['Target'] = pd.Series(pd.np.where(bicluster3_transpose.index.isin(sample_names_array), 1, 0), index=bicluster3_transpose.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>DVL3</th>\n",
       "      <th>EPHB4</th>\n",
       "      <th>APOC1</th>\n",
       "      <th>LAMB3</th>\n",
       "      <th>HOXD11</th>\n",
       "      <th>ANO1</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509819_E1796T.CEL</th>\n",
       "      <td>7.691792</td>\n",
       "      <td>7.516879</td>\n",
       "      <td>8.240716</td>\n",
       "      <td>10.040889</td>\n",
       "      <td>4.567577</td>\n",
       "      <td>4.450293</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol               DVL3     EPHB4     APOC1      LAMB3    HOXD11  \\\n",
       "GSM509819_E1796T.CEL  7.691792  7.516879  8.240716  10.040889  4.567577   \n",
       "\n",
       "Gene Symbol               ANO1  Target  \n",
       "GSM509819_E1796T.CEL  4.450293       1  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "bicluster3_transpose['Target'] = pd.Series(pd.np.where(bicluster3_transpose.index.isin(sample_names_array), 1, 0), index=bicluster3_transpose.index)\n",
    "\n",
    "# Check the Dimension of the Bicluster 3 Dataset and Display it\n",
    "print(\"Bicluster 3 Dimension: \", bicluster3_transpose.shape)\n",
    "bicluster3_transpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicluster 4 Dimension:  (1, 14)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\3889919974.py:2: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  bicluster4_transpose['Target'] = pd.Series(pd.np.where(bicluster4_transpose.index.isin(sample_names_array), 1, 0), index=bicluster4_transpose.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>NREP</th>\n",
       "      <th>HEXB</th>\n",
       "      <th>EPHB4</th>\n",
       "      <th>KIF3B</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>SAP30</th>\n",
       "      <th>PTK2</th>\n",
       "      <th>LAMB3</th>\n",
       "      <th>MBD4</th>\n",
       "      <th>CHN1</th>\n",
       "      <th>ALMS1</th>\n",
       "      <th>HOXD11</th>\n",
       "      <th>BANP</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509816_E1614T.CEL</th>\n",
       "      <td>7.82422</td>\n",
       "      <td>8.918272</td>\n",
       "      <td>7.471044</td>\n",
       "      <td>6.550222</td>\n",
       "      <td>6.12884</td>\n",
       "      <td>5.130646</td>\n",
       "      <td>9.079471</td>\n",
       "      <td>9.57014</td>\n",
       "      <td>8.38678</td>\n",
       "      <td>6.323455</td>\n",
       "      <td>6.883904</td>\n",
       "      <td>5.71802</td>\n",
       "      <td>7.740366</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              NREP      HEXB     EPHB4     KIF3B    BRCA1  \\\n",
       "GSM509816_E1614T.CEL  7.82422  8.918272  7.471044  6.550222  6.12884   \n",
       "\n",
       "Gene Symbol              SAP30      PTK2    LAMB3     MBD4      CHN1  \\\n",
       "GSM509816_E1614T.CEL  5.130646  9.079471  9.57014  8.38678  6.323455   \n",
       "\n",
       "Gene Symbol              ALMS1   HOXD11      BANP  Target  \n",
       "GSM509816_E1614T.CEL  6.883904  5.71802  7.740366       1  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "bicluster4_transpose['Target'] = pd.Series(pd.np.where(bicluster4_transpose.index.isin(sample_names_array), 1, 0), index=bicluster4_transpose.index)\n",
    "\n",
    "# Check the Dimension of the Bicluster 4 Dataset and Display it\n",
    "print(\"Bicluster 4 Dimension: \", bicluster4_transpose.shape)\n",
    "bicluster4_transpose"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Classfication the Gene To Identify the Potential Biomarkers**\n",
    "\n",
    "*From the biclustering, a group of cancer sample had been retrieved. In order to know which gene is the potential biomarker, the gene in all biclusters will be retrieved out for classification process.*\n",
    "<br><br><br>\n",
    "\n",
    "##### **First Step**\n",
    "*Gene Symbol in the biclusters will be retrieved out. Then, the occurrences of the gene will be calculated. As from the findings, some of the genes appears once and some genes appear twice. Hence, the classification process will be implement to all the genes and genes that occurred twice.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieved the Gene Symbol from Biclusters\n",
    "gene_bicluster1 = bicluster1[\"Gene Symbol\"]\n",
    "gene_bicluster2 = bicluster2[\"Gene Symbol\"]\n",
    "gene_bicluster3 = bicluster3[\"Gene Symbol\"]\n",
    "gene_bicluster4 = bicluster4[\"Gene Symbol\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene symbol: ABCF1, Occurrences: 1\n",
      "Gene symbol: AP2M1, Occurrences: 1\n",
      "Gene symbol: PTGES3, Occurrences: 1\n",
      "Gene symbol: LAMC1, Occurrences: 1\n",
      "Gene symbol: PLOD1, Occurrences: 1\n",
      "Gene symbol: PSMD2, Occurrences: 1\n",
      "Gene symbol: SLC3A2, Occurrences: 1\n",
      "Gene symbol: SSRP1, Occurrences: 1\n",
      "Gene symbol: RPN1, Occurrences: 1\n",
      "Gene symbol: MMP2, Occurrences: 1\n",
      "Gene symbol: KPNA2, Occurrences: 1\n",
      "Gene symbol: TUBA1B, Occurrences: 1\n",
      "Gene symbol: THBS1, Occurrences: 1\n",
      "Gene symbol: SSB, Occurrences: 1\n",
      "Gene symbol: ITPR3, Occurrences: 1\n",
      "Gene symbol: COX7A2L, Occurrences: 1\n",
      "Gene symbol: BGN, Occurrences: 1\n",
      "Gene symbol: TXNRD1, Occurrences: 1\n",
      "Gene symbol: RALY, Occurrences: 1\n",
      "Gene symbol: AKR1B1, Occurrences: 1\n",
      "Gene symbol: PTDSS1, Occurrences: 1\n",
      "Gene symbol: SHC1, Occurrences: 1\n",
      "Gene symbol: TGFBI, Occurrences: 1\n",
      "Gene symbol: NCBP2, Occurrences: 1\n",
      "Gene symbol: FXR1, Occurrences: 1\n",
      "Gene symbol: LUM, Occurrences: 1\n",
      "Gene symbol: AEBP1, Occurrences: 1\n",
      "Gene symbol: COL3A1, Occurrences: 1\n",
      "Gene symbol: CDC25B, Occurrences: 1\n",
      "Gene symbol: ARPC1B, Occurrences: 1\n",
      "Gene symbol: MYO10, Occurrences: 1\n",
      "Gene symbol: SLC16A1, Occurrences: 1\n",
      "Gene symbol: PSMB4, Occurrences: 1\n",
      "Gene symbol: CDK4, Occurrences: 1\n",
      "Gene symbol: COL1A1, Occurrences: 1\n",
      "Gene symbol: COL1A2, Occurrences: 1\n",
      "Gene symbol: CPSF6, Occurrences: 1\n",
      "Gene symbol: RANBP1, Occurrences: 1\n",
      "Gene symbol: FBN1, Occurrences: 1\n",
      "Gene symbol: UBE2S, Occurrences: 1\n",
      "Gene symbol: CDC20, Occurrences: 1\n",
      "Gene symbol: MSH6, Occurrences: 1\n",
      "Gene symbol: UBE2C, Occurrences: 1\n",
      "Gene symbol: PTPRK, Occurrences: 1\n",
      "Gene symbol: THBS2, Occurrences: 1\n",
      "Gene symbol: SLC33A1, Occurrences: 1\n",
      "Gene symbol: COL5A1, Occurrences: 1\n",
      "Gene symbol: MNAT1, Occurrences: 1\n",
      "Gene symbol: ERF, Occurrences: 1\n",
      "Gene symbol: ZKSCAN5, Occurrences: 1\n",
      "Gene symbol: HMGB3, Occurrences: 1\n",
      "Gene symbol: FZD6, Occurrences: 1\n",
      "Gene symbol: FUT8, Occurrences: 1\n",
      "Gene symbol: COL7A1, Occurrences: 1\n",
      "Gene symbol: EIF2AK2, Occurrences: 1\n",
      "Gene symbol: SMC2, Occurrences: 1\n",
      "Gene symbol: COL11A1, Occurrences: 1\n",
      "Gene symbol: MMP1, Occurrences: 1\n",
      "Gene symbol: MYO5A, Occurrences: 1\n",
      "Gene symbol: EXO1, Occurrences: 1\n",
      "Gene symbol: VCAN, Occurrences: 1\n",
      "Gene symbol: NEK2, Occurrences: 1\n",
      "Gene symbol: IRS1, Occurrences: 1\n",
      "Gene symbol: WDHD1, Occurrences: 1\n",
      "Gene symbol: FEN1, Occurrences: 1\n",
      "Gene symbol: RAD51, Occurrences: 1\n",
      "Gene symbol: FLAD1, Occurrences: 1\n",
      "Gene symbol: POLE2, Occurrences: 1\n",
      "Gene symbol: GTF2E1, Occurrences: 1\n",
      "Gene symbol: COL10A1, Occurrences: 1\n",
      "Gene symbol: PAK2, Occurrences: 1\n",
      "Gene symbol: TNFAIP6, Occurrences: 1\n",
      "Gene symbol: GALNS, Occurrences: 1\n",
      "Gene symbol: NUP155, Occurrences: 1\n",
      "Gene symbol: CDH11, Occurrences: 1\n",
      "Gene symbol: TFRC, Occurrences: 1\n",
      "Gene symbol: ALG3, Occurrences: 1\n",
      "Gene symbol: PRPS1, Occurrences: 1\n",
      "Gene symbol: PARP1, Occurrences: 1\n",
      "Gene symbol: XPO1, Occurrences: 1\n",
      "Gene symbol: ATP1B3, Occurrences: 1\n",
      "Gene symbol: LONP1, Occurrences: 1\n",
      "Gene symbol: TUBB, Occurrences: 1\n",
      "Gene symbol: SF3B4, Occurrences: 1\n",
      "Gene symbol: POLR2H, Occurrences: 1\n",
      "Gene symbol: ADCY3, Occurrences: 1\n",
      "Gene symbol: NTHL1, Occurrences: 1\n",
      "Gene symbol: RPL39L, Occurrences: 1\n",
      "Gene symbol: BANF1, Occurrences: 1\n",
      "Gene symbol: GCFC2, Occurrences: 1\n",
      "Gene symbol: HSP90AA1, Occurrences: 1\n",
      "Gene symbol: FZD2, Occurrences: 1\n",
      "Gene symbol: FN1, Occurrences: 1\n",
      "Gene symbol: TRIM27, Occurrences: 1\n",
      "Gene symbol: SLC25A17, Occurrences: 1\n",
      "Gene symbol: GANAB, Occurrences: 1\n",
      "Gene symbol: COL4A2, Occurrences: 1\n",
      "Gene symbol: COL4A1, Occurrences: 1\n",
      "Gene symbol: PSMD14, Occurrences: 1\n",
      "Gene symbol: FTL, Occurrences: 1\n",
      "Gene symbol: CKAP5, Occurrences: 1\n",
      "Gene symbol: FANCI, Occurrences: 1\n",
      "Gene symbol: TCEA1, Occurrences: 1\n",
      "Gene symbol: SUPT16H, Occurrences: 1\n",
      "Gene symbol: POMGNT1, Occurrences: 1\n",
      "Gene symbol: NUP85, Occurrences: 1\n",
      "Gene symbol: USB1, Occurrences: 1\n",
      "Gene symbol: ARMC1, Occurrences: 1\n",
      "Gene symbol: GREM1, Occurrences: 1\n",
      "Gene symbol: WDR12, Occurrences: 1\n",
      "Gene symbol: CD320, Occurrences: 1\n",
      "Gene symbol: NIT2, Occurrences: 1\n",
      "Gene symbol: RFWD3, Occurrences: 1\n",
      "Gene symbol: AGK, Occurrences: 1\n",
      "Gene symbol: TMEM39A, Occurrences: 1\n",
      "Gene symbol: NUP107, Occurrences: 1\n",
      "Gene symbol: TMEM38B, Occurrences: 1\n",
      "Gene symbol: PBK, Occurrences: 1\n",
      "Gene symbol: ASPM, Occurrences: 1\n",
      "Gene symbol: NELFCD, Occurrences: 1\n",
      "Gene symbol: DNMT3B, Occurrences: 1\n",
      "Gene symbol: TBL1XR1, Occurrences: 1\n",
      "Gene symbol: GINS2, Occurrences: 1\n",
      "Gene symbol: COL5A2, Occurrences: 1\n",
      "Gene symbol: PLD1, Occurrences: 1\n",
      "Gene symbol: CAPNS1, Occurrences: 1\n",
      "Gene symbol: RHOA, Occurrences: 1\n",
      "Gene symbol: GUK1, Occurrences: 1\n",
      "Gene symbol: CAPN1, Occurrences: 1\n",
      "Gene symbol: ATP5O, Occurrences: 1\n",
      "Gene symbol: USP9X, Occurrences: 1\n",
      "Gene symbol: GNAI3, Occurrences: 1\n",
      "Gene symbol: RAB5B, Occurrences: 1\n",
      "Gene symbol: ALDH2, Occurrences: 1\n",
      "Gene symbol: GOLGA4, Occurrences: 1\n",
      "Gene symbol: TSG101, Occurrences: 1\n",
      "Gene symbol: SH3GL1, Occurrences: 1\n",
      "Gene symbol: VWF, Occurrences: 1\n",
      "Gene symbol: CHMP2A, Occurrences: 1\n",
      "Gene symbol: GAS7, Occurrences: 1\n",
      "Gene symbol: LSS, Occurrences: 1\n",
      "Gene symbol: CSK, Occurrences: 1\n",
      "Gene symbol: RAP1A, Occurrences: 1\n",
      "Gene symbol: NOTCH2, Occurrences: 1\n",
      "Gene symbol: TRIM29, Occurrences: 1\n",
      "Gene symbol: GALE, Occurrences: 1\n",
      "Gene symbol: CHMP2B, Occurrences: 1\n",
      "Gene symbol: PRKCD, Occurrences: 1\n",
      "Gene symbol: ITPR2, Occurrences: 1\n",
      "Gene symbol: MAPKAPK3, Occurrences: 1\n",
      "Gene symbol: C7, Occurrences: 1\n",
      "Gene symbol: SLC11A2, Occurrences: 1\n",
      "Gene symbol: TIMP2, Occurrences: 1\n",
      "Gene symbol: RHOG, Occurrences: 1\n",
      "Gene symbol: PDLIM5, Occurrences: 1\n",
      "Gene symbol: DYNLT3, Occurrences: 1\n",
      "Gene symbol: HES1, Occurrences: 1\n",
      "Gene symbol: EPHA2, Occurrences: 1\n",
      "Gene symbol: SSBP2, Occurrences: 1\n",
      "Gene symbol: KAT2B, Occurrences: 1\n",
      "Gene symbol: GAB2, Occurrences: 1\n",
      "Gene symbol: ATP5D, Occurrences: 1\n",
      "Gene symbol: SUOX, Occurrences: 1\n",
      "Gene symbol: HNMT, Occurrences: 1\n",
      "Gene symbol: GNG11, Occurrences: 1\n",
      "Gene symbol: ADK, Occurrences: 1\n",
      "Gene symbol: BTD, Occurrences: 1\n",
      "Gene symbol: ACOX3, Occurrences: 1\n",
      "Gene symbol: PPP1R3C, Occurrences: 1\n",
      "Gene symbol: SYNGR1, Occurrences: 1\n",
      "Gene symbol: CYP11A1, Occurrences: 1\n",
      "Gene symbol: GOLGA2, Occurrences: 1\n",
      "Gene symbol: DPYD, Occurrences: 1\n",
      "Gene symbol: ABCA8, Occurrences: 1\n",
      "Gene symbol: PRKAR2A, Occurrences: 1\n",
      "Gene symbol: GMDS, Occurrences: 1\n",
      "Gene symbol: GNE, Occurrences: 1\n",
      "Gene symbol: TMPRSS2, Occurrences: 1\n",
      "Gene symbol: CAPN5, Occurrences: 1\n",
      "Gene symbol: RP2, Occurrences: 1\n",
      "Gene symbol: AP1B1, Occurrences: 1\n",
      "Gene symbol: CTSG, Occurrences: 1\n",
      "Gene symbol: ENTPD5, Occurrences: 1\n",
      "Gene symbol: EREG, Occurrences: 1\n",
      "Gene symbol: WASL, Occurrences: 1\n",
      "Gene symbol: JAK2, Occurrences: 1\n",
      "Gene symbol: TGM3, Occurrences: 1\n",
      "Gene symbol: RAB5A, Occurrences: 1\n",
      "Gene symbol: DLG2, Occurrences: 1\n",
      "Gene symbol: TLR3, Occurrences: 1\n",
      "Gene symbol: IL18, Occurrences: 1\n",
      "Gene symbol: CNGA1, Occurrences: 1\n",
      "Gene symbol: SIM2, Occurrences: 1\n",
      "Gene symbol: OVOL1, Occurrences: 1\n",
      "Gene symbol: DKK4, Occurrences: 1\n",
      "Gene symbol: EPB41L3, Occurrences: 1\n",
      "Gene symbol: PLAGL1, Occurrences: 1\n",
      "Gene symbol: SLC16A7, Occurrences: 1\n",
      "Gene symbol: GAB1, Occurrences: 1\n",
      "Gene symbol: CAST, Occurrences: 1\n",
      "Gene symbol: MUC1, Occurrences: 1\n",
      "Gene symbol: CITED2, Occurrences: 1\n",
      "Gene symbol: ARPC2, Occurrences: 1\n",
      "Gene symbol: IL1RL2, Occurrences: 1\n",
      "Gene symbol: PITX1, Occurrences: 1\n",
      "Gene symbol: RAC1, Occurrences: 1\n",
      "Gene symbol: CAPN2, Occurrences: 1\n",
      "Gene symbol: ARPC3, Occurrences: 1\n",
      "Gene symbol: SAP18, Occurrences: 1\n",
      "Gene symbol: STK24, Occurrences: 1\n",
      "Gene symbol: TXN, Occurrences: 1\n",
      "Gene symbol: ABI1, Occurrences: 1\n",
      "Gene symbol: MYD88, Occurrences: 1\n",
      "Gene symbol: MSMO1, Occurrences: 1\n",
      "Gene symbol: PIM1, Occurrences: 1\n",
      "Gene symbol: TRIP11, Occurrences: 1\n",
      "Gene symbol: IL33, Occurrences: 1\n",
      "Gene symbol: MAPK13, Occurrences: 1\n",
      "Gene symbol: PLP1, Occurrences: 1\n",
      "Gene symbol: FUT6, Occurrences: 1\n",
      "Gene symbol: RORA, Occurrences: 1\n",
      "Gene symbol: ERBB2, Occurrences: 1\n",
      "Gene symbol: CDK7, Occurrences: 1\n",
      "Gene symbol: AHNAK, Occurrences: 1\n",
      "Gene symbol: MAPK3, Occurrences: 1\n",
      "Gene symbol: LPIN1, Occurrences: 1\n",
      "Gene symbol: UBE2E1, Occurrences: 1\n",
      "Gene symbol: DOCK9, Occurrences: 1\n",
      "Gene symbol: IL1RN, Occurrences: 1\n",
      "Gene symbol: PLEKHM1, Occurrences: 1\n",
      "Gene symbol: COL14A1, Occurrences: 1\n",
      "Gene symbol: KANK1, Occurrences: 1\n",
      "Gene symbol: TICAM1, Occurrences: 1\n",
      "Gene symbol: ABHD5, Occurrences: 1\n",
      "Gene symbol: TEAD1, Occurrences: 1\n",
      "Gene symbol: GYS2, Occurrences: 1\n",
      "Gene symbol: LAMB4, Occurrences: 1\n",
      "Gene symbol: FLG, Occurrences: 1\n",
      "Gene symbol: PDCD6IP, Occurrences: 1\n",
      "Gene symbol: ARPC4, Occurrences: 1\n",
      "Gene symbol: SMAP1, Occurrences: 1\n",
      "Gene symbol: VPS4B, Occurrences: 1\n",
      "Gene symbol: LMBRD1, Occurrences: 1\n",
      "Gene symbol: FYCO1, Occurrences: 1\n",
      "Gene symbol: GOLPH3L, Occurrences: 1\n",
      "Gene symbol: SLC35C1, Occurrences: 1\n",
      "Gene symbol: NDUFA3, Occurrences: 1\n",
      "Gene symbol: FAM114A2, Occurrences: 1\n",
      "Gene symbol: RHBDF1, Occurrences: 1\n",
      "Gene symbol: SNX24, Occurrences: 1\n",
      "Gene symbol: VAV3, Occurrences: 1\n",
      "Gene symbol: GALNT12, Occurrences: 1\n",
      "Gene symbol: THSD4, Occurrences: 1\n",
      "Gene symbol: MSRA, Occurrences: 1\n",
      "Gene symbol: C1GALT1, Occurrences: 1\n",
      "Gene symbol: ELOVL4, Occurrences: 1\n",
      "Gene symbol: SAMD9, Occurrences: 1\n",
      "Gene symbol: DENND1B, Occurrences: 1\n",
      "Gene symbol: RNF39, Occurrences: 1\n",
      "Gene symbol: SLC27A6, Occurrences: 1\n",
      "Gene symbol: GIPC2, Occurrences: 1\n",
      "Gene symbol: EPB41L4A, Occurrences: 1\n",
      "Gene symbol: ADAMTSL4, Occurrences: 1\n",
      "Gene symbol: RIPK4, Occurrences: 1\n",
      "Gene symbol: UBAP1, Occurrences: 1\n",
      "Gene symbol: VPS37B, Occurrences: 1\n",
      "Gene symbol: CSNK1E, Occurrences: 1\n",
      "Gene symbol: DVL3, Occurrences: 1\n",
      "Gene symbol: EPHB4, Occurrences: 2\n",
      "Gene symbol: APOC1, Occurrences: 1\n",
      "Gene symbol: LAMB3, Occurrences: 2\n",
      "Gene symbol: HOXD11, Occurrences: 2\n",
      "Gene symbol: ANO1, Occurrences: 1\n",
      "Gene symbol: NREP, Occurrences: 1\n",
      "Gene symbol: HEXB, Occurrences: 1\n",
      "Gene symbol: KIF3B, Occurrences: 1\n",
      "Gene symbol: BRCA1, Occurrences: 1\n",
      "Gene symbol: SAP30, Occurrences: 1\n",
      "Gene symbol: PTK2, Occurrences: 1\n",
      "Gene symbol: MBD4, Occurrences: 1\n",
      "Gene symbol: CHN1, Occurrences: 1\n",
      "Gene symbol: ALMS1, Occurrences: 1\n",
      "Gene symbol: BANP, Occurrences: 1\n"
     ]
    }
   ],
   "source": [
    "# Concatenate the Gene Together\n",
    "# From This Part, May Observed that Some Genes Occur Once while Some Genes Occur Twice\n",
    "gene_symbols = np.concatenate((gene_bicluster1, gene_bicluster2, gene_bicluster3, gene_bicluster4))\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "# Initialize an Empty List to Store Gene Symbols Occurring Exactly Twice\n",
    "gene_symbols_count_2 = []\n",
    "\n",
    "# Count the Occurrence of the Gene\n",
    "counts = Counter(symbol for symbol in gene_symbols)\n",
    "\n",
    "# Print the Symbols and Their Counts\n",
    "for symbol, count in counts.items():\n",
    "    print(f\"Gene symbol: {symbol}, Occurrences: {count}\")\n",
    "    if count > 1:\n",
    "        gene_symbols_count_2.append(symbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['EPHB4', 'LAMB3', 'HOXD11']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_symbols_count_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### **Second Step**\n",
    "*The genes that have been retrieved will be used to filter the data from the original dataset. Then, one more column named as \"Target\" will be added for classification purpose.*\n",
    "<br><br><br>\n",
    "\n",
    "**Notes:** Below is the dataset that involved all the genes in the biclusters data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The dimension of the dataset is:  (34, 284)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\310653734.py:8: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  important_gene_data['Target'] = pd.Series(pd.np.where(important_gene_data.index.isin(sample_names_array), 1, 0), index=important_gene_data.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ADAMTSL4</th>\n",
       "      <th>NELFCD</th>\n",
       "      <th>DNMT3B</th>\n",
       "      <th>RIPK4</th>\n",
       "      <th>TBL1XR1</th>\n",
       "      <th>UBAP1</th>\n",
       "      <th>GINS2</th>\n",
       "      <th>VPS37B</th>\n",
       "      <th>COL5A2</th>\n",
       "      <th>CSNK1E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>5.864863</td>\n",
       "      <td>8.847780</td>\n",
       "      <td>5.429582</td>\n",
       "      <td>10.528422</td>\n",
       "      <td>6.103538</td>\n",
       "      <td>8.576635</td>\n",
       "      <td>6.942437</td>\n",
       "      <td>9.204626</td>\n",
       "      <td>5.803623</td>\n",
       "      <td>7.330217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>5.416042</td>\n",
       "      <td>8.486673</td>\n",
       "      <td>5.233733</td>\n",
       "      <td>10.102096</td>\n",
       "      <td>5.805960</td>\n",
       "      <td>8.265671</td>\n",
       "      <td>6.760712</td>\n",
       "      <td>7.639002</td>\n",
       "      <td>5.855633</td>\n",
       "      <td>7.400418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>5.300279</td>\n",
       "      <td>8.597308</td>\n",
       "      <td>5.480147</td>\n",
       "      <td>10.313087</td>\n",
       "      <td>5.571494</td>\n",
       "      <td>8.563540</td>\n",
       "      <td>6.950697</td>\n",
       "      <td>8.970822</td>\n",
       "      <td>6.742893</td>\n",
       "      <td>7.682231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>5.341540</td>\n",
       "      <td>8.602311</td>\n",
       "      <td>5.228284</td>\n",
       "      <td>10.123108</td>\n",
       "      <td>4.935344</td>\n",
       "      <td>8.577154</td>\n",
       "      <td>6.846777</td>\n",
       "      <td>9.346425</td>\n",
       "      <td>5.990346</td>\n",
       "      <td>7.277551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>6.309677</td>\n",
       "      <td>8.247174</td>\n",
       "      <td>5.252625</td>\n",
       "      <td>10.470718</td>\n",
       "      <td>4.991667</td>\n",
       "      <td>8.115302</td>\n",
       "      <td>6.561701</td>\n",
       "      <td>8.833978</td>\n",
       "      <td>5.726995</td>\n",
       "      <td>7.696482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol           ADAMTSL4    NELFCD    DNMT3B      RIPK4   TBL1XR1  \\\n",
       "GSM509787_E1507N.CEL  5.864863  8.847780  5.429582  10.528422  6.103538   \n",
       "GSM509788_E1520N.CEL  5.416042  8.486673  5.233733  10.102096  5.805960   \n",
       "GSM509789_E1521N.CEL  5.300279  8.597308  5.480147  10.313087  5.571494   \n",
       "GSM509790_E1532N.CEL  5.341540  8.602311  5.228284  10.123108  4.935344   \n",
       "GSM509791_E1535N.CEL  6.309677  8.247174  5.252625  10.470718  4.991667   \n",
       "\n",
       "Gene Symbol              UBAP1     GINS2    VPS37B    COL5A2    CSNK1E  \n",
       "GSM509787_E1507N.CEL  8.576635  6.942437  9.204626  5.803623  7.330217  \n",
       "GSM509788_E1520N.CEL  8.265671  6.760712  7.639002  5.855633  7.400418  \n",
       "GSM509789_E1521N.CEL  8.563540  6.950697  8.970822  6.742893  7.682231  \n",
       "GSM509790_E1532N.CEL  8.577154  6.846777  9.346425  5.990346  7.277551  \n",
       "GSM509791_E1535N.CEL  8.115302  6.561701  8.833978  5.726995  7.696482  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter Based on gene symbols\n",
    "important_gene_data = filter_gene_expression_data[filter_gene_expression_data['Gene Symbol'].isin(gene_symbols)]\n",
    "\n",
    "# Transpose the DataFrame\n",
    "important_gene_data = important_gene_data.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "important_gene_data['Target'] = pd.Series(pd.np.where(important_gene_data.index.isin(sample_names_array), 1, 0), index=important_gene_data.index)\n",
    "print(\"The dimension of the dataset is: \", important_gene_data.shape)\n",
    "important_gene_data.iloc[:,273:283].head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Notes:** Below is dataset that involved the Genes that occurred twice in the biclusters data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The dimension of the dataset is:  (34, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\787545725.py:8: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  important_gene_data_2['Target'] = pd.Series(pd.np.where(important_gene_data_2.index.isin(sample_names_array), 1, 0), index=important_gene_data_2.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>EPHB4</th>\n",
       "      <th>LAMB3</th>\n",
       "      <th>HOXD11</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>6.892630</td>\n",
       "      <td>8.098309</td>\n",
       "      <td>4.216160</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>6.778301</td>\n",
       "      <td>6.880918</td>\n",
       "      <td>4.863400</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>7.016927</td>\n",
       "      <td>8.011573</td>\n",
       "      <td>4.607442</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>6.984915</td>\n",
       "      <td>7.639520</td>\n",
       "      <td>4.455296</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>7.114578</td>\n",
       "      <td>7.014760</td>\n",
       "      <td>4.367265</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              EPHB4     LAMB3    HOXD11  Target\n",
       "GSM509787_E1507N.CEL  6.892630  8.098309  4.216160       0\n",
       "GSM509788_E1520N.CEL  6.778301  6.880918  4.863400       0\n",
       "GSM509789_E1521N.CEL  7.016927  8.011573  4.607442       0\n",
       "GSM509790_E1532N.CEL  6.984915  7.639520  4.455296       0\n",
       "GSM509791_E1535N.CEL  7.114578  7.014760  4.367265       0"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter based on gene symbols\n",
    "important_gene_data_2 = filter_gene_expression_data[filter_gene_expression_data['Gene Symbol'].isin(gene_symbols_count_2)]\n",
    "\n",
    "# Transpose the DataFrame\n",
    "important_gene_data_2 = important_gene_data_2.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "important_gene_data_2['Target'] = pd.Series(pd.np.where(important_gene_data_2.index.isin(sample_names_array), 1, 0), index=important_gene_data_2.index)\n",
    "print(\"The dimension of the dataset is: \", important_gene_data_2.shape)\n",
    "important_gene_data_2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Notes:** Below is the Original Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The dimension of the dataset is:  (34, 2736)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\2588962534.py:5: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  original_gene_expression_data['Target'] = pd.Series(pd.np.where(original_gene_expression_data.index.isin(sample_names_array), 1, 0), index=original_gene_expression_data.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>TBC1D2</th>\n",
       "      <th>B4GALT7</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>NACA2</th>\n",
       "      <th>TRDN</th>\n",
       "      <th>SCAF4</th>\n",
       "      <th>LAMA1</th>\n",
       "      <th>FBXO31</th>\n",
       "      <th>SLC44A1</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>7.765013</td>\n",
       "      <td>5.949763</td>\n",
       "      <td>7.054392</td>\n",
       "      <td>3.451194</td>\n",
       "      <td>3.389328</td>\n",
       "      <td>6.060083</td>\n",
       "      <td>4.591470</td>\n",
       "      <td>6.167321</td>\n",
       "      <td>6.157059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>7.297379</td>\n",
       "      <td>5.828337</td>\n",
       "      <td>6.838957</td>\n",
       "      <td>3.677271</td>\n",
       "      <td>3.281392</td>\n",
       "      <td>5.524395</td>\n",
       "      <td>4.637121</td>\n",
       "      <td>6.635602</td>\n",
       "      <td>6.376391</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>7.498661</td>\n",
       "      <td>5.811177</td>\n",
       "      <td>7.229369</td>\n",
       "      <td>3.575415</td>\n",
       "      <td>3.498269</td>\n",
       "      <td>6.353812</td>\n",
       "      <td>4.658779</td>\n",
       "      <td>6.359670</td>\n",
       "      <td>5.889888</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>7.672872</td>\n",
       "      <td>5.988737</td>\n",
       "      <td>7.151789</td>\n",
       "      <td>3.647739</td>\n",
       "      <td>3.266002</td>\n",
       "      <td>6.034053</td>\n",
       "      <td>4.482420</td>\n",
       "      <td>6.408367</td>\n",
       "      <td>6.167377</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>7.479567</td>\n",
       "      <td>5.652972</td>\n",
       "      <td>6.965022</td>\n",
       "      <td>3.517448</td>\n",
       "      <td>3.465066</td>\n",
       "      <td>6.884892</td>\n",
       "      <td>4.618102</td>\n",
       "      <td>6.108703</td>\n",
       "      <td>6.840216</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol             TBC1D2   B4GALT7  CASP8AP2     NACA2      TRDN  \\\n",
       "GSM509787_E1507N.CEL  7.765013  5.949763  7.054392  3.451194  3.389328   \n",
       "GSM509788_E1520N.CEL  7.297379  5.828337  6.838957  3.677271  3.281392   \n",
       "GSM509789_E1521N.CEL  7.498661  5.811177  7.229369  3.575415  3.498269   \n",
       "GSM509790_E1532N.CEL  7.672872  5.988737  7.151789  3.647739  3.266002   \n",
       "GSM509791_E1535N.CEL  7.479567  5.652972  6.965022  3.517448  3.465066   \n",
       "\n",
       "Gene Symbol              SCAF4     LAMA1    FBXO31   SLC44A1  Target  \n",
       "GSM509787_E1507N.CEL  6.060083  4.591470  6.167321  6.157059       0  \n",
       "GSM509788_E1520N.CEL  5.524395  4.637121  6.635602  6.376391       0  \n",
       "GSM509789_E1521N.CEL  6.353812  4.658779  6.359670  5.889888       0  \n",
       "GSM509790_E1532N.CEL  6.034053  4.482420  6.408367  6.167377       0  \n",
       "GSM509791_E1535N.CEL  6.884892  4.618102  6.108703  6.840216       0  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transpose the Gene Expression Dataset\n",
    "original_gene_expression_data = filter_gene_expression_data.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "original_gene_expression_data['Target'] = pd.Series(pd.np.where(original_gene_expression_data.index.isin(sample_names_array), 1, 0), index=original_gene_expression_data.index)\n",
    "print(\"The dimension of the dataset is: \", original_gene_expression_data.shape)\n",
    "original_gene_expression_data.iloc[:,2726:2736].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes In All Biclusters :\n",
      "\n",
      "Cross-validation scores: [1.         1.         1.         1.         1.         0.66666667\n",
      " 1.         1.         1.         1.        ]\n",
      "Mean accuracy: 0.9666666666666666\n",
      "Standard deviation of accuracy: 0.1\n",
      "Confusion Matrix:\n",
      "[[17  0]\n",
      " [ 1 16]]\n",
      "Accuracy: 0.9705882352941176\n",
      "Precision: 1.0\n",
      "Recall: 0.9411764705882353\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.9696969696969697 \n",
      "\n",
      "\n",
      "Genes Occur in Multiple Biclusters :\n",
      "\n",
      "Cross-validation scores: [1.   1.   0.75 1.   1.   1.   1.   1.   1.   1.  ]\n",
      "Mean accuracy: 0.975\n",
      "Standard deviation of accuracy: 0.075\n",
      "Confusion Matrix:\n",
      "[[16  1]\n",
      " [ 0 17]]\n",
      "Accuracy: 0.9705882352941176\n",
      "Precision: 0.9444444444444444\n",
      "Recall: 1.0\n",
      "Specificity: 0.9411764705882353\n",
      "F1 Score: 0.9714285714285714 \n",
      "\n",
      "\n",
      "Original Dataset :\n",
      "\n",
      "Cross-validation scores: [1.         1.         1.         1.         1.         0.66666667\n",
      " 1.         1.         1.         1.        ]\n",
      "Mean accuracy: 0.9666666666666666\n",
      "Standard deviation of accuracy: 0.1\n",
      "Confusion Matrix:\n",
      "[[17  0]\n",
      " [ 1 16]]\n",
      "Accuracy: 0.9705882352941176\n",
      "Precision: 1.0\n",
      "Recall: 0.9411764705882353\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.9696969696969697 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import KFold, cross_val_score, cross_val_predict\n",
    "from sklearn import svm\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "\n",
    "datasets = [important_gene_data, important_gene_data_2, original_gene_expression_data]\n",
    "labels = ['Genes In All Biclusters', 'Genes Occur in Multiple Biclusters', 'Original Dataset']\n",
    "\n",
    "cross_validation = []\n",
    "\n",
    "for i, data in enumerate(datasets):\n",
    "    X = data.iloc[:, :-1]\n",
    "    y = data.iloc[:, -1]\n",
    "    \n",
    "    # Split dataset into training set and test set\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, stratify=y) \n",
    "    \n",
    "    # Create a SVM Classifier\n",
    "    clf = svm.SVC(kernel='linear', probability=True)\n",
    "    clf.fit(X_train, y_train)\n",
    "\n",
    "    # Obtain Labels\n",
    "    print(labels[i], \":\\n\")\n",
    "\n",
    "    # Confusion Matrix and Cross Validation\n",
    "    # Perform 10 Fold Validation\n",
    "    k_folds = KFold(n_splits = 10)\n",
    "    # Calculate the Cross Validation Score\n",
    "    scores = cross_val_score(clf, X, y, cv = k_folds)\n",
    "    # Store the scores\n",
    "    cross_validation.append(scores)\n",
    "    # Perform cross-validation and obtain predicted labels for each fold\n",
    "    predicted_labels = cross_val_predict(clf, X, y, cv=k_folds)\n",
    "\n",
    "    # Compute the confusion matrix\n",
    "    conf_matrix = confusion_matrix(y, predicted_labels)\n",
    "    # Print the cross-validation scores\n",
    "    print(\"Cross-validation scores:\", scores)\n",
    "    # Print the mean and standard deviation of the scores\n",
    "    print(\"Mean accuracy:\", scores.mean())\n",
    "    print(\"Standard deviation of accuracy:\", scores.std())\n",
    "    # Print the confusion matrix\n",
    "    print(\"Confusion Matrix:\")\n",
    "    print(conf_matrix)\n",
    "\n",
    "    # Extract true positives (TP), true negatives (TN), false positives (FP), and false negatives (FN)\n",
    "    TP = conf_matrix[1, 1]\n",
    "    TN = conf_matrix[0, 0]\n",
    "    FP = conf_matrix[0, 1]\n",
    "    FN = conf_matrix[1, 0]\n",
    "\n",
    "    # Calculate accuracy\n",
    "    accuracy = (TP + TN) / (TP + TN + FP + FN)\n",
    "    # Calculate precision\n",
    "    precision = TP / (TP + FP)\n",
    "    # Calculate recall\n",
    "    recall = TP / (TP + FN)\n",
    "    # Calculate specificity\n",
    "    specificity = TN / (TN + FP)\n",
    "    # Calculate F1 score\n",
    "    f1_score = 2 * (precision * recall) / (precision + recall)\n",
    "\n",
    "    print(\"Accuracy:\", accuracy)\n",
    "    print(\"Precision:\", precision)\n",
    "    print(\"Recall:\", recall)\n",
    "    print(\"Specificity:\", specificity)\n",
    "    print(\"F1 Score:\", f1_score, \"\\n\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### **Third Step**\n",
    "*Apply Different Random State to the Gene Expression Dataset* <br><br>\n",
    "\n",
    "**Performance Measurement** <br>\n",
    "*Accuracy of each random state will be obtained to measure the performance of SVM classifier for each dataset* <br>\n",
    "*Confusion Matrix will be generated to make the comparison* <br>\n",
    "*T-Test will be performed to check the significance differences between 3 datasets.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset:  Genes In All Biclusters\n",
      "Random State: 5, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 10, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 15, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 20, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 25, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 30, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 35, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 40, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 45, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 50, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Dataset:  Genes Occur in Multiple Biclusters\n",
      "Random State: 5, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 10, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 15, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 20, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[6 1]\n",
      " [0 7]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 0.875\n",
      "Recall: 1.0\n",
      "Specificity: 0.8571428571428571\n",
      "F1 Score: 0.9333333333333333 \n",
      "\n",
      "\n",
      "Random State: 25, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[6 1]\n",
      " [0 7]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 0.875\n",
      "Recall: 1.0\n",
      "Specificity: 0.8571428571428571\n",
      "F1 Score: 0.9333333333333333 \n",
      "\n",
      "\n",
      "Random State: 30, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 35, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 40, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 45, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 50, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Dataset:  Original Dataset\n",
      "Random State: 5, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 10, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 15, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 20, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 25, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 30, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 35, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 40, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 45, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 50, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import svm\n",
    "from sklearn.metrics import confusion_matrix, roc_curve, auc\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "datasets = [important_gene_data, important_gene_data_2, original_gene_expression_data]\n",
    "labels = ['Genes In All Biclusters', 'Genes Occur in Multiple Biclusters', 'Original Dataset']\n",
    "\n",
    "# Initialize dictionaries to store FPR, TPR, AUC values, and accuracy scores for each dataset\n",
    "results = {label: {'fprs': [], 'tprs': [], 'aucs': [], 'scores': []} for label in labels}\n",
    "random_states = [5, 10, 15, 20, 25, 30, 35, 40, 45, 50]\n",
    "\n",
    "for i, data in enumerate(datasets):\n",
    "    X = data.iloc[:, :-1]\n",
    "    y = data.iloc[:, -1]\n",
    "    label = labels[i]\n",
    "\n",
    "    print(\"Dataset: \", label)\n",
    "    \n",
    "    for rs in random_states:\n",
    "        # Split dataset into training set and test set\n",
    "        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=rs, stratify=y) \n",
    "\n",
    "        # Create a SVM Classifier\n",
    "        clf = svm.SVC(kernel='linear', probability=True)\n",
    "        clf.fit(X_train, y_train)\n",
    "        y_pred = clf.predict(X_test)\n",
    "        accuracy = np.mean(y_pred == y_test)\n",
    "        results[label]['scores'].append(accuracy)\n",
    "        print(f\"Random State: {rs}, Accuracy: {accuracy}\")     \n",
    "\n",
    "        # ROC Curve\n",
    "        # Get predicted probabilities\n",
    "        y_pred_proba = clf.predict_proba(X_test)[:, 1]\n",
    "\n",
    "        # Calculate ROC curve\n",
    "        fpr, tpr, _ = roc_curve(y_test, y_pred_proba)\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "\n",
    "        # Store the results\n",
    "        results[label]['fprs'].append(fpr)\n",
    "        results[label]['tprs'].append(tpr)\n",
    "        results[label]['aucs'].append(roc_auc)\n",
    "\n",
    "        # Confusion Matrix\n",
    "        conf_matrix = confusion_matrix(y_test, y_pred)\n",
    "        print(\"Confusion Matrix:\")\n",
    "        print(conf_matrix)\n",
    "\n",
    "        # Extract true positives (TP), true negatives (TN), false positives (FP), and false negatives (FN)\n",
    "        TP = conf_matrix[1, 1]\n",
    "        TN = conf_matrix[0, 0]\n",
    "        FP = conf_matrix[0, 1]\n",
    "        FN = conf_matrix[1, 0]\n",
    "\n",
    "        # Calculate accuracy\n",
    "        accuracy = (TP + TN) / (TP + TN + FP + FN)\n",
    "        # Calculate precision\n",
    "        precision = TP / (TP + FP)\n",
    "        # Calculate recall\n",
    "        recall = TP / (TP + FN)\n",
    "        # Calculate specificity\n",
    "        specificity = TN / (TN + FP)\n",
    "        # Calculate F1 score\n",
    "        f1_score = 2 * (precision * recall) / (precision + recall)\n",
    "\n",
    "        print(\"Accuracy:\", accuracy)\n",
    "        print(\"Precision:\", precision)\n",
    "        print(\"Recall:\", recall)\n",
    "        print(\"Specificity:\", specificity)\n",
    "        print(\"F1 Score:\", f1_score, \"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T-test between Genes In All Biclusters and Genes Occur in Multiple Biclusters:\n",
      "T-statistic: 0.8847, P-value: 0.3880\n",
      "\n",
      "\n",
      "T-test between Genes In All Biclusters and Original Dataset:\n",
      "T-statistic: 0.0000, P-value: 1.0000\n",
      "\n",
      "\n",
      "T-test between Genes Occur in Multiple Biclusters and Original Dataset:\n",
      "T-statistic: -0.8847, P-value: 0.3880\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import ttest_ind\n",
    "\n",
    "# Perform t-tests\n",
    "for i, label1 in enumerate(labels):\n",
    "    for j, label2 in enumerate(labels):\n",
    "        if i < j:\n",
    "            scores1 = results[label1]['scores']\n",
    "            scores2 = results[label2]['scores']\n",
    "            t_stat, p_value = ttest_ind(scores1, scores2)\n",
    "            print(f\"T-test between {label1} and {label2}:\")\n",
    "            print(f\"T-statistic: {t_stat:.4f}, P-value: {p_value:.4f}\")\n",
    "            print(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### **Additional Testing on Biclusters**\n",
    "\n",
    "*Retrieved the data from Gene Expression Dataset based on each bicluster.* <br>\n",
    "*Then, applying the SVM classifier to obtain the performance of each bicluster.* <br>\n",
    "*Generate Confusion Matrix for performance measurement.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The dimension of the dataset is:  (34, 125)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\3742791526.py:8: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  bicluster1_data['Target'] = pd.Series(pd.np.where(bicluster1_data.index.isin(sample_names_array), 1, 0), index=bicluster1_data.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>AP2M1</th>\n",
       "      <th>PTGES3</th>\n",
       "      <th>LAMC1</th>\n",
       "      <th>PLOD1</th>\n",
       "      <th>PSMD2</th>\n",
       "      <th>SLC3A2</th>\n",
       "      <th>SSRP1</th>\n",
       "      <th>RPN1</th>\n",
       "      <th>MMP2</th>\n",
       "      <th>...</th>\n",
       "      <th>NUP107</th>\n",
       "      <th>TMEM38B</th>\n",
       "      <th>PBK</th>\n",
       "      <th>ASPM</th>\n",
       "      <th>NELFCD</th>\n",
       "      <th>DNMT3B</th>\n",
       "      <th>TBL1XR1</th>\n",
       "      <th>GINS2</th>\n",
       "      <th>COL5A2</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>8.206662</td>\n",
       "      <td>9.908055</td>\n",
       "      <td>11.148693</td>\n",
       "      <td>7.226272</td>\n",
       "      <td>6.972069</td>\n",
       "      <td>9.350621</td>\n",
       "      <td>7.982148</td>\n",
       "      <td>7.758097</td>\n",
       "      <td>8.403526</td>\n",
       "      <td>6.238408</td>\n",
       "      <td>...</td>\n",
       "      <td>7.782242</td>\n",
       "      <td>4.488508</td>\n",
       "      <td>6.419918</td>\n",
       "      <td>7.595809</td>\n",
       "      <td>8.847780</td>\n",
       "      <td>5.429582</td>\n",
       "      <td>6.103538</td>\n",
       "      <td>6.942437</td>\n",
       "      <td>5.803623</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>8.013936</td>\n",
       "      <td>9.763084</td>\n",
       "      <td>10.984160</td>\n",
       "      <td>7.778692</td>\n",
       "      <td>6.949944</td>\n",
       "      <td>9.397088</td>\n",
       "      <td>7.546967</td>\n",
       "      <td>7.551768</td>\n",
       "      <td>8.198948</td>\n",
       "      <td>6.496422</td>\n",
       "      <td>...</td>\n",
       "      <td>7.510356</td>\n",
       "      <td>4.629209</td>\n",
       "      <td>5.386169</td>\n",
       "      <td>6.355480</td>\n",
       "      <td>8.486673</td>\n",
       "      <td>5.233733</td>\n",
       "      <td>5.805960</td>\n",
       "      <td>6.760712</td>\n",
       "      <td>5.855633</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>8.252299</td>\n",
       "      <td>9.902845</td>\n",
       "      <td>11.243377</td>\n",
       "      <td>8.077243</td>\n",
       "      <td>7.212330</td>\n",
       "      <td>9.611524</td>\n",
       "      <td>8.310192</td>\n",
       "      <td>7.641995</td>\n",
       "      <td>8.529213</td>\n",
       "      <td>6.828133</td>\n",
       "      <td>...</td>\n",
       "      <td>7.751299</td>\n",
       "      <td>4.886045</td>\n",
       "      <td>5.924394</td>\n",
       "      <td>6.598339</td>\n",
       "      <td>8.597308</td>\n",
       "      <td>5.480147</td>\n",
       "      <td>5.571494</td>\n",
       "      <td>6.950697</td>\n",
       "      <td>6.742893</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>8.221171</td>\n",
       "      <td>9.670848</td>\n",
       "      <td>11.055280</td>\n",
       "      <td>7.429716</td>\n",
       "      <td>7.152230</td>\n",
       "      <td>9.173008</td>\n",
       "      <td>7.575227</td>\n",
       "      <td>7.443549</td>\n",
       "      <td>8.431940</td>\n",
       "      <td>6.049753</td>\n",
       "      <td>...</td>\n",
       "      <td>7.738510</td>\n",
       "      <td>4.557542</td>\n",
       "      <td>5.948984</td>\n",
       "      <td>7.157977</td>\n",
       "      <td>8.602311</td>\n",
       "      <td>5.228284</td>\n",
       "      <td>4.935344</td>\n",
       "      <td>6.846777</td>\n",
       "      <td>5.990346</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>7.794273</td>\n",
       "      <td>9.378916</td>\n",
       "      <td>10.693799</td>\n",
       "      <td>7.010852</td>\n",
       "      <td>7.072170</td>\n",
       "      <td>9.587929</td>\n",
       "      <td>7.597609</td>\n",
       "      <td>7.333652</td>\n",
       "      <td>8.861096</td>\n",
       "      <td>6.618955</td>\n",
       "      <td>...</td>\n",
       "      <td>7.853255</td>\n",
       "      <td>4.601803</td>\n",
       "      <td>5.698980</td>\n",
       "      <td>7.150963</td>\n",
       "      <td>8.247174</td>\n",
       "      <td>5.252625</td>\n",
       "      <td>4.991667</td>\n",
       "      <td>6.561701</td>\n",
       "      <td>5.726995</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 125 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              ABCF1     AP2M1     PTGES3     LAMC1     PLOD1  \\\n",
       "GSM509787_E1507N.CEL  8.206662  9.908055  11.148693  7.226272  6.972069   \n",
       "GSM509788_E1520N.CEL  8.013936  9.763084  10.984160  7.778692  6.949944   \n",
       "GSM509789_E1521N.CEL  8.252299  9.902845  11.243377  8.077243  7.212330   \n",
       "GSM509790_E1532N.CEL  8.221171  9.670848  11.055280  7.429716  7.152230   \n",
       "GSM509791_E1535N.CEL  7.794273  9.378916  10.693799  7.010852  7.072170   \n",
       "\n",
       "Gene Symbol              PSMD2    SLC3A2     SSRP1      RPN1      MMP2  ...  \\\n",
       "GSM509787_E1507N.CEL  9.350621  7.982148  7.758097  8.403526  6.238408  ...   \n",
       "GSM509788_E1520N.CEL  9.397088  7.546967  7.551768  8.198948  6.496422  ...   \n",
       "GSM509789_E1521N.CEL  9.611524  8.310192  7.641995  8.529213  6.828133  ...   \n",
       "GSM509790_E1532N.CEL  9.173008  7.575227  7.443549  8.431940  6.049753  ...   \n",
       "GSM509791_E1535N.CEL  9.587929  7.597609  7.333652  8.861096  6.618955  ...   \n",
       "\n",
       "Gene Symbol             NUP107   TMEM38B       PBK      ASPM    NELFCD  \\\n",
       "GSM509787_E1507N.CEL  7.782242  4.488508  6.419918  7.595809  8.847780   \n",
       "GSM509788_E1520N.CEL  7.510356  4.629209  5.386169  6.355480  8.486673   \n",
       "GSM509789_E1521N.CEL  7.751299  4.886045  5.924394  6.598339  8.597308   \n",
       "GSM509790_E1532N.CEL  7.738510  4.557542  5.948984  7.157977  8.602311   \n",
       "GSM509791_E1535N.CEL  7.853255  4.601803  5.698980  7.150963  8.247174   \n",
       "\n",
       "Gene Symbol             DNMT3B   TBL1XR1     GINS2    COL5A2  Target  \n",
       "GSM509787_E1507N.CEL  5.429582  6.103538  6.942437  5.803623       0  \n",
       "GSM509788_E1520N.CEL  5.233733  5.805960  6.760712  5.855633       0  \n",
       "GSM509789_E1521N.CEL  5.480147  5.571494  6.950697  6.742893       0  \n",
       "GSM509790_E1532N.CEL  5.228284  4.935344  6.846777  5.990346       0  \n",
       "GSM509791_E1535N.CEL  5.252625  4.991667  6.561701  5.726995       0  \n",
       "\n",
       "[5 rows x 125 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter based on gene symbols in Bicluster 1\n",
    "bicluster1_data = filter_gene_expression_data[filter_gene_expression_data['Gene Symbol'].isin(gene_bicluster1)]\n",
    "\n",
    "# Transpose the DataFrame\n",
    "bicluster1_data = bicluster1_data.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "bicluster1_data['Target'] = pd.Series(pd.np.where(bicluster1_data.index.isin(sample_names_array), 1, 0), index=bicluster1_data.index)\n",
    "print(\"The dimension of the dataset is: \", bicluster1_data.shape)\n",
    "bicluster1_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>NUP107</th>\n",
       "      <th>TMEM38B</th>\n",
       "      <th>PBK</th>\n",
       "      <th>ASPM</th>\n",
       "      <th>NELFCD</th>\n",
       "      <th>DNMT3B</th>\n",
       "      <th>TBL1XR1</th>\n",
       "      <th>GINS2</th>\n",
       "      <th>COL5A2</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>7.782242</td>\n",
       "      <td>4.488508</td>\n",
       "      <td>6.419918</td>\n",
       "      <td>7.595809</td>\n",
       "      <td>8.847780</td>\n",
       "      <td>5.429582</td>\n",
       "      <td>6.103538</td>\n",
       "      <td>6.942437</td>\n",
       "      <td>5.803623</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>7.510356</td>\n",
       "      <td>4.629209</td>\n",
       "      <td>5.386169</td>\n",
       "      <td>6.355480</td>\n",
       "      <td>8.486673</td>\n",
       "      <td>5.233733</td>\n",
       "      <td>5.805960</td>\n",
       "      <td>6.760712</td>\n",
       "      <td>5.855633</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>7.751299</td>\n",
       "      <td>4.886045</td>\n",
       "      <td>5.924394</td>\n",
       "      <td>6.598339</td>\n",
       "      <td>8.597308</td>\n",
       "      <td>5.480147</td>\n",
       "      <td>5.571494</td>\n",
       "      <td>6.950697</td>\n",
       "      <td>6.742893</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>7.738510</td>\n",
       "      <td>4.557542</td>\n",
       "      <td>5.948984</td>\n",
       "      <td>7.157977</td>\n",
       "      <td>8.602311</td>\n",
       "      <td>5.228284</td>\n",
       "      <td>4.935344</td>\n",
       "      <td>6.846777</td>\n",
       "      <td>5.990346</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>7.853255</td>\n",
       "      <td>4.601803</td>\n",
       "      <td>5.698980</td>\n",
       "      <td>7.150963</td>\n",
       "      <td>8.247174</td>\n",
       "      <td>5.252625</td>\n",
       "      <td>4.991667</td>\n",
       "      <td>6.561701</td>\n",
       "      <td>5.726995</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol             NUP107   TMEM38B       PBK      ASPM    NELFCD  \\\n",
       "GSM509787_E1507N.CEL  7.782242  4.488508  6.419918  7.595809  8.847780   \n",
       "GSM509788_E1520N.CEL  7.510356  4.629209  5.386169  6.355480  8.486673   \n",
       "GSM509789_E1521N.CEL  7.751299  4.886045  5.924394  6.598339  8.597308   \n",
       "GSM509790_E1532N.CEL  7.738510  4.557542  5.948984  7.157977  8.602311   \n",
       "GSM509791_E1535N.CEL  7.853255  4.601803  5.698980  7.150963  8.247174   \n",
       "\n",
       "Gene Symbol             DNMT3B   TBL1XR1     GINS2    COL5A2  Target  \n",
       "GSM509787_E1507N.CEL  5.429582  6.103538  6.942437  5.803623       0  \n",
       "GSM509788_E1520N.CEL  5.233733  5.805960  6.760712  5.855633       0  \n",
       "GSM509789_E1521N.CEL  5.480147  5.571494  6.950697  6.742893       0  \n",
       "GSM509790_E1532N.CEL  5.228284  4.935344  6.846777  5.990346       0  \n",
       "GSM509791_E1535N.CEL  5.252625  4.991667  6.561701  5.726995       0  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bicluster1_data.iloc[:, 115:125].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The dimension of the dataset is:  (34, 144)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\3679862082.py:8: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  bicluster2_data['Target'] = pd.Series(pd.np.where(bicluster2_data.index.isin(sample_names_array), 1, 0), index=bicluster2_data.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>PLD1</th>\n",
       "      <th>CAPNS1</th>\n",
       "      <th>RHOA</th>\n",
       "      <th>GUK1</th>\n",
       "      <th>CAPN1</th>\n",
       "      <th>ATP5O</th>\n",
       "      <th>USP9X</th>\n",
       "      <th>GNAI3</th>\n",
       "      <th>RAB5B</th>\n",
       "      <th>ALDH2</th>\n",
       "      <th>...</th>\n",
       "      <th>RNF39</th>\n",
       "      <th>SLC27A6</th>\n",
       "      <th>GIPC2</th>\n",
       "      <th>EPB41L4A</th>\n",
       "      <th>ADAMTSL4</th>\n",
       "      <th>RIPK4</th>\n",
       "      <th>UBAP1</th>\n",
       "      <th>VPS37B</th>\n",
       "      <th>CSNK1E</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>6.905594</td>\n",
       "      <td>11.424412</td>\n",
       "      <td>11.511128</td>\n",
       "      <td>7.950930</td>\n",
       "      <td>9.831494</td>\n",
       "      <td>10.322505</td>\n",
       "      <td>9.396548</td>\n",
       "      <td>9.151744</td>\n",
       "      <td>8.969629</td>\n",
       "      <td>10.020181</td>\n",
       "      <td>...</td>\n",
       "      <td>7.291863</td>\n",
       "      <td>5.072733</td>\n",
       "      <td>5.386074</td>\n",
       "      <td>5.103955</td>\n",
       "      <td>5.864863</td>\n",
       "      <td>10.528422</td>\n",
       "      <td>8.576635</td>\n",
       "      <td>9.204626</td>\n",
       "      <td>7.330217</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>6.819023</td>\n",
       "      <td>11.364217</td>\n",
       "      <td>11.384112</td>\n",
       "      <td>7.901695</td>\n",
       "      <td>9.465135</td>\n",
       "      <td>10.254164</td>\n",
       "      <td>9.138960</td>\n",
       "      <td>9.054295</td>\n",
       "      <td>8.793115</td>\n",
       "      <td>10.029048</td>\n",
       "      <td>...</td>\n",
       "      <td>6.748005</td>\n",
       "      <td>5.104316</td>\n",
       "      <td>4.523464</td>\n",
       "      <td>5.749728</td>\n",
       "      <td>5.416042</td>\n",
       "      <td>10.102096</td>\n",
       "      <td>8.265671</td>\n",
       "      <td>7.639002</td>\n",
       "      <td>7.400418</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>6.570245</td>\n",
       "      <td>10.941655</td>\n",
       "      <td>11.016838</td>\n",
       "      <td>7.745549</td>\n",
       "      <td>9.293055</td>\n",
       "      <td>10.307529</td>\n",
       "      <td>9.306884</td>\n",
       "      <td>9.199137</td>\n",
       "      <td>8.831414</td>\n",
       "      <td>10.133004</td>\n",
       "      <td>...</td>\n",
       "      <td>7.146375</td>\n",
       "      <td>6.155529</td>\n",
       "      <td>4.660901</td>\n",
       "      <td>5.114664</td>\n",
       "      <td>5.300279</td>\n",
       "      <td>10.313087</td>\n",
       "      <td>8.563540</td>\n",
       "      <td>8.970822</td>\n",
       "      <td>7.682231</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>7.091794</td>\n",
       "      <td>11.279280</td>\n",
       "      <td>11.090685</td>\n",
       "      <td>8.068325</td>\n",
       "      <td>9.657434</td>\n",
       "      <td>10.335877</td>\n",
       "      <td>9.502077</td>\n",
       "      <td>9.378688</td>\n",
       "      <td>9.056482</td>\n",
       "      <td>9.560176</td>\n",
       "      <td>...</td>\n",
       "      <td>7.831722</td>\n",
       "      <td>5.230182</td>\n",
       "      <td>4.351468</td>\n",
       "      <td>5.803941</td>\n",
       "      <td>5.341540</td>\n",
       "      <td>10.123108</td>\n",
       "      <td>8.577154</td>\n",
       "      <td>9.346425</td>\n",
       "      <td>7.277551</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>6.670091</td>\n",
       "      <td>11.086393</td>\n",
       "      <td>10.388703</td>\n",
       "      <td>7.572464</td>\n",
       "      <td>9.602016</td>\n",
       "      <td>10.166136</td>\n",
       "      <td>9.215495</td>\n",
       "      <td>9.146276</td>\n",
       "      <td>9.040620</td>\n",
       "      <td>10.172073</td>\n",
       "      <td>...</td>\n",
       "      <td>6.753784</td>\n",
       "      <td>4.348867</td>\n",
       "      <td>4.193768</td>\n",
       "      <td>4.942595</td>\n",
       "      <td>6.309677</td>\n",
       "      <td>10.470718</td>\n",
       "      <td>8.115302</td>\n",
       "      <td>8.833978</td>\n",
       "      <td>7.696482</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 144 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol               PLD1     CAPNS1       RHOA      GUK1     CAPN1  \\\n",
       "GSM509787_E1507N.CEL  6.905594  11.424412  11.511128  7.950930  9.831494   \n",
       "GSM509788_E1520N.CEL  6.819023  11.364217  11.384112  7.901695  9.465135   \n",
       "GSM509789_E1521N.CEL  6.570245  10.941655  11.016838  7.745549  9.293055   \n",
       "GSM509790_E1532N.CEL  7.091794  11.279280  11.090685  8.068325  9.657434   \n",
       "GSM509791_E1535N.CEL  6.670091  11.086393  10.388703  7.572464  9.602016   \n",
       "\n",
       "Gene Symbol               ATP5O     USP9X     GNAI3     RAB5B      ALDH2  ...  \\\n",
       "GSM509787_E1507N.CEL  10.322505  9.396548  9.151744  8.969629  10.020181  ...   \n",
       "GSM509788_E1520N.CEL  10.254164  9.138960  9.054295  8.793115  10.029048  ...   \n",
       "GSM509789_E1521N.CEL  10.307529  9.306884  9.199137  8.831414  10.133004  ...   \n",
       "GSM509790_E1532N.CEL  10.335877  9.502077  9.378688  9.056482   9.560176  ...   \n",
       "GSM509791_E1535N.CEL  10.166136  9.215495  9.146276  9.040620  10.172073  ...   \n",
       "\n",
       "Gene Symbol              RNF39   SLC27A6     GIPC2  EPB41L4A  ADAMTSL4  \\\n",
       "GSM509787_E1507N.CEL  7.291863  5.072733  5.386074  5.103955  5.864863   \n",
       "GSM509788_E1520N.CEL  6.748005  5.104316  4.523464  5.749728  5.416042   \n",
       "GSM509789_E1521N.CEL  7.146375  6.155529  4.660901  5.114664  5.300279   \n",
       "GSM509790_E1532N.CEL  7.831722  5.230182  4.351468  5.803941  5.341540   \n",
       "GSM509791_E1535N.CEL  6.753784  4.348867  4.193768  4.942595  6.309677   \n",
       "\n",
       "Gene Symbol               RIPK4     UBAP1    VPS37B    CSNK1E  Target  \n",
       "GSM509787_E1507N.CEL  10.528422  8.576635  9.204626  7.330217       0  \n",
       "GSM509788_E1520N.CEL  10.102096  8.265671  7.639002  7.400418       0  \n",
       "GSM509789_E1521N.CEL  10.313087  8.563540  8.970822  7.682231       0  \n",
       "GSM509790_E1532N.CEL  10.123108  8.577154  9.346425  7.277551       0  \n",
       "GSM509791_E1535N.CEL  10.470718  8.115302  8.833978  7.696482       0  \n",
       "\n",
       "[5 rows x 144 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter based on gene symbols in Bicluster 2\n",
    "bicluster2_data = filter_gene_expression_data[filter_gene_expression_data['Gene Symbol'].isin(gene_bicluster2)]\n",
    "\n",
    "# Transpose the DataFrame\n",
    "bicluster2_data = bicluster2_data.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "bicluster2_data['Target'] = pd.Series(pd.np.where(bicluster2_data.index.isin(sample_names_array), 1, 0), index=bicluster2_data.index)\n",
    "print(\"The dimension of the dataset is: \", bicluster2_data.shape)\n",
    "bicluster2_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>GIPC2</th>\n",
       "      <th>EPB41L4A</th>\n",
       "      <th>ADAMTSL4</th>\n",
       "      <th>RIPK4</th>\n",
       "      <th>UBAP1</th>\n",
       "      <th>VPS37B</th>\n",
       "      <th>CSNK1E</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>5.386074</td>\n",
       "      <td>5.103955</td>\n",
       "      <td>5.864863</td>\n",
       "      <td>10.528422</td>\n",
       "      <td>8.576635</td>\n",
       "      <td>9.204626</td>\n",
       "      <td>7.330217</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>4.523464</td>\n",
       "      <td>5.749728</td>\n",
       "      <td>5.416042</td>\n",
       "      <td>10.102096</td>\n",
       "      <td>8.265671</td>\n",
       "      <td>7.639002</td>\n",
       "      <td>7.400418</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>4.660901</td>\n",
       "      <td>5.114664</td>\n",
       "      <td>5.300279</td>\n",
       "      <td>10.313087</td>\n",
       "      <td>8.563540</td>\n",
       "      <td>8.970822</td>\n",
       "      <td>7.682231</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>4.351468</td>\n",
       "      <td>5.803941</td>\n",
       "      <td>5.341540</td>\n",
       "      <td>10.123108</td>\n",
       "      <td>8.577154</td>\n",
       "      <td>9.346425</td>\n",
       "      <td>7.277551</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>4.193768</td>\n",
       "      <td>4.942595</td>\n",
       "      <td>6.309677</td>\n",
       "      <td>10.470718</td>\n",
       "      <td>8.115302</td>\n",
       "      <td>8.833978</td>\n",
       "      <td>7.696482</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol              GIPC2  EPB41L4A  ADAMTSL4      RIPK4     UBAP1  \\\n",
       "GSM509787_E1507N.CEL  5.386074  5.103955  5.864863  10.528422  8.576635   \n",
       "GSM509788_E1520N.CEL  4.523464  5.749728  5.416042  10.102096  8.265671   \n",
       "GSM509789_E1521N.CEL  4.660901  5.114664  5.300279  10.313087  8.563540   \n",
       "GSM509790_E1532N.CEL  4.351468  5.803941  5.341540  10.123108  8.577154   \n",
       "GSM509791_E1535N.CEL  4.193768  4.942595  6.309677  10.470718  8.115302   \n",
       "\n",
       "Gene Symbol             VPS37B    CSNK1E  Target  \n",
       "GSM509787_E1507N.CEL  9.204626  7.330217       0  \n",
       "GSM509788_E1520N.CEL  7.639002  7.400418       0  \n",
       "GSM509789_E1521N.CEL  8.970822  7.682231       0  \n",
       "GSM509790_E1532N.CEL  9.346425  7.277551       0  \n",
       "GSM509791_E1535N.CEL  8.833978  7.696482       0  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bicluster2_data.iloc[:, 136:146].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The dimension of the dataset is:  (34, 7)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\2426611975.py:8: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  bicluster3_data['Target'] = pd.Series(pd.np.where(bicluster3_data.index.isin(sample_names_array), 1, 0), index=bicluster3_data.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>DVL3</th>\n",
       "      <th>EPHB4</th>\n",
       "      <th>APOC1</th>\n",
       "      <th>LAMB3</th>\n",
       "      <th>HOXD11</th>\n",
       "      <th>ANO1</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>6.855224</td>\n",
       "      <td>6.892630</td>\n",
       "      <td>6.086037</td>\n",
       "      <td>8.098309</td>\n",
       "      <td>4.216160</td>\n",
       "      <td>4.267156</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>6.884612</td>\n",
       "      <td>6.778301</td>\n",
       "      <td>5.985619</td>\n",
       "      <td>6.880918</td>\n",
       "      <td>4.863400</td>\n",
       "      <td>4.328800</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>6.897960</td>\n",
       "      <td>7.016927</td>\n",
       "      <td>6.242063</td>\n",
       "      <td>8.011573</td>\n",
       "      <td>4.607442</td>\n",
       "      <td>4.879896</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>6.986769</td>\n",
       "      <td>6.984915</td>\n",
       "      <td>5.904467</td>\n",
       "      <td>7.639520</td>\n",
       "      <td>4.455296</td>\n",
       "      <td>3.938380</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>6.735594</td>\n",
       "      <td>7.114578</td>\n",
       "      <td>6.424061</td>\n",
       "      <td>7.014760</td>\n",
       "      <td>4.367265</td>\n",
       "      <td>5.173945</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol               DVL3     EPHB4     APOC1     LAMB3    HOXD11  \\\n",
       "GSM509787_E1507N.CEL  6.855224  6.892630  6.086037  8.098309  4.216160   \n",
       "GSM509788_E1520N.CEL  6.884612  6.778301  5.985619  6.880918  4.863400   \n",
       "GSM509789_E1521N.CEL  6.897960  7.016927  6.242063  8.011573  4.607442   \n",
       "GSM509790_E1532N.CEL  6.986769  6.984915  5.904467  7.639520  4.455296   \n",
       "GSM509791_E1535N.CEL  6.735594  7.114578  6.424061  7.014760  4.367265   \n",
       "\n",
       "Gene Symbol               ANO1  Target  \n",
       "GSM509787_E1507N.CEL  4.267156       0  \n",
       "GSM509788_E1520N.CEL  4.328800       0  \n",
       "GSM509789_E1521N.CEL  4.879896       0  \n",
       "GSM509790_E1532N.CEL  3.938380       0  \n",
       "GSM509791_E1535N.CEL  5.173945       0  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter based on gene symbols in Bicluster 3\n",
    "bicluster3_data = filter_gene_expression_data[filter_gene_expression_data['Gene Symbol'].isin(gene_bicluster3)]\n",
    "\n",
    "# Transpose the DataFrame\n",
    "bicluster3_data = bicluster3_data.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "bicluster3_data['Target'] = pd.Series(pd.np.where(bicluster3_data.index.isin(sample_names_array), 1, 0), index=bicluster3_data.index)\n",
    "print(\"The dimension of the dataset is: \", bicluster3_data.shape)\n",
    "bicluster3_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The dimension of the dataset is:  (34, 14)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_10816\\662358430.py:8: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.\n",
      "  bicluster4_data['Target'] = pd.Series(pd.np.where(bicluster4_data.index.isin(sample_names_array), 1, 0), index=bicluster4_data.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>NREP</th>\n",
       "      <th>HEXB</th>\n",
       "      <th>EPHB4</th>\n",
       "      <th>KIF3B</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>SAP30</th>\n",
       "      <th>PTK2</th>\n",
       "      <th>LAMB3</th>\n",
       "      <th>MBD4</th>\n",
       "      <th>CHN1</th>\n",
       "      <th>ALMS1</th>\n",
       "      <th>HOXD11</th>\n",
       "      <th>BANP</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM509787_E1507N.CEL</th>\n",
       "      <td>6.793849</td>\n",
       "      <td>8.409774</td>\n",
       "      <td>6.892630</td>\n",
       "      <td>6.450461</td>\n",
       "      <td>5.465491</td>\n",
       "      <td>5.056697</td>\n",
       "      <td>7.247609</td>\n",
       "      <td>8.098309</td>\n",
       "      <td>7.479309</td>\n",
       "      <td>5.479397</td>\n",
       "      <td>6.168217</td>\n",
       "      <td>4.216160</td>\n",
       "      <td>7.477664</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509788_E1520N.CEL</th>\n",
       "      <td>6.467512</td>\n",
       "      <td>8.476937</td>\n",
       "      <td>6.778301</td>\n",
       "      <td>6.250886</td>\n",
       "      <td>5.454997</td>\n",
       "      <td>5.165094</td>\n",
       "      <td>7.124530</td>\n",
       "      <td>6.880918</td>\n",
       "      <td>7.229357</td>\n",
       "      <td>5.859875</td>\n",
       "      <td>6.209464</td>\n",
       "      <td>4.863400</td>\n",
       "      <td>7.672460</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509789_E1521N.CEL</th>\n",
       "      <td>6.276003</td>\n",
       "      <td>8.624108</td>\n",
       "      <td>7.016927</td>\n",
       "      <td>6.523165</td>\n",
       "      <td>5.501618</td>\n",
       "      <td>4.986756</td>\n",
       "      <td>7.136149</td>\n",
       "      <td>8.011573</td>\n",
       "      <td>7.436556</td>\n",
       "      <td>5.813864</td>\n",
       "      <td>6.329830</td>\n",
       "      <td>4.607442</td>\n",
       "      <td>7.517760</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509790_E1532N.CEL</th>\n",
       "      <td>6.308079</td>\n",
       "      <td>8.348051</td>\n",
       "      <td>6.984915</td>\n",
       "      <td>6.066030</td>\n",
       "      <td>5.208072</td>\n",
       "      <td>5.074219</td>\n",
       "      <td>7.136199</td>\n",
       "      <td>7.639520</td>\n",
       "      <td>7.605766</td>\n",
       "      <td>5.904994</td>\n",
       "      <td>6.186083</td>\n",
       "      <td>4.455296</td>\n",
       "      <td>7.783826</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM509791_E1535N.CEL</th>\n",
       "      <td>5.957888</td>\n",
       "      <td>8.073806</td>\n",
       "      <td>7.114578</td>\n",
       "      <td>6.370263</td>\n",
       "      <td>5.548058</td>\n",
       "      <td>4.929455</td>\n",
       "      <td>7.090199</td>\n",
       "      <td>7.014760</td>\n",
       "      <td>7.422373</td>\n",
       "      <td>5.421429</td>\n",
       "      <td>6.490425</td>\n",
       "      <td>4.367265</td>\n",
       "      <td>7.863359</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol               NREP      HEXB     EPHB4     KIF3B     BRCA1  \\\n",
       "GSM509787_E1507N.CEL  6.793849  8.409774  6.892630  6.450461  5.465491   \n",
       "GSM509788_E1520N.CEL  6.467512  8.476937  6.778301  6.250886  5.454997   \n",
       "GSM509789_E1521N.CEL  6.276003  8.624108  7.016927  6.523165  5.501618   \n",
       "GSM509790_E1532N.CEL  6.308079  8.348051  6.984915  6.066030  5.208072   \n",
       "GSM509791_E1535N.CEL  5.957888  8.073806  7.114578  6.370263  5.548058   \n",
       "\n",
       "Gene Symbol              SAP30      PTK2     LAMB3      MBD4      CHN1  \\\n",
       "GSM509787_E1507N.CEL  5.056697  7.247609  8.098309  7.479309  5.479397   \n",
       "GSM509788_E1520N.CEL  5.165094  7.124530  6.880918  7.229357  5.859875   \n",
       "GSM509789_E1521N.CEL  4.986756  7.136149  8.011573  7.436556  5.813864   \n",
       "GSM509790_E1532N.CEL  5.074219  7.136199  7.639520  7.605766  5.904994   \n",
       "GSM509791_E1535N.CEL  4.929455  7.090199  7.014760  7.422373  5.421429   \n",
       "\n",
       "Gene Symbol              ALMS1    HOXD11      BANP  Target  \n",
       "GSM509787_E1507N.CEL  6.168217  4.216160  7.477664       0  \n",
       "GSM509788_E1520N.CEL  6.209464  4.863400  7.672460       0  \n",
       "GSM509789_E1521N.CEL  6.329830  4.607442  7.517760       0  \n",
       "GSM509790_E1532N.CEL  6.186083  4.455296  7.783826       0  \n",
       "GSM509791_E1535N.CEL  6.490425  4.367265  7.863359       0  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter based on gene symbols in Bicluster 4\n",
    "bicluster4_data = filter_gene_expression_data[filter_gene_expression_data['Gene Symbol'].isin(gene_bicluster4)]\n",
    "\n",
    "# Transpose the DataFrame\n",
    "bicluster4_data = bicluster4_data.set_index('Gene Symbol').transpose()\n",
    "\n",
    "# Add a New Column 'Target' Based on the Condition (Check if the Sample Fall Between sample_names_array)\n",
    "bicluster4_data['Target'] = pd.Series(pd.np.where(bicluster4_data.index.isin(sample_names_array), 1, 0), index=bicluster4_data.index)\n",
    "print(\"The dimension of the dataset is: \", bicluster4_data.shape)\n",
    "bicluster4_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset Based Bicluster 1 :\n",
      "\n",
      "Cross-validation scores: [1.         1.         1.         1.         1.         0.33333333\n",
      " 1.         1.         1.         1.        ]\n",
      "Mean accuracy: 0.9333333333333332\n",
      "Standard deviation of accuracy: 0.19999999999999998\n",
      "Confusion Matrix:\n",
      "[[17  0]\n",
      " [ 2 15]]\n",
      "Accuracy: 0.9411764705882353\n",
      "Precision: 1.0\n",
      "Recall: 0.8823529411764706\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.9375 \n",
      "\n",
      "\n",
      "Dataset Based Bicluster 2 :\n",
      "\n",
      "Cross-validation scores: [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]\n",
      "Mean accuracy: 1.0\n",
      "Standard deviation of accuracy: 0.0\n",
      "Confusion Matrix:\n",
      "[[17  0]\n",
      " [ 0 17]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Dataset Based Bicluster 3 :\n",
      "\n",
      "Cross-validation scores: [1.         1.         0.75       1.         1.         0.33333333\n",
      " 1.         1.         1.         1.        ]\n",
      "Mean accuracy: 0.9083333333333332\n",
      "Standard deviation of accuracy: 0.20564937798755112\n",
      "Confusion Matrix:\n",
      "[[16  1]\n",
      " [ 2 15]]\n",
      "Accuracy: 0.9117647058823529\n",
      "Precision: 0.9375\n",
      "Recall: 0.8823529411764706\n",
      "Specificity: 0.9411764705882353\n",
      "F1 Score: 0.9090909090909091 \n",
      "\n",
      "\n",
      "Dataset Based Bicluster 4 :\n",
      "\n",
      "Cross-validation scores: [1.         1.         1.         1.         1.         0.66666667\n",
      " 1.         1.         1.         1.        ]\n",
      "Mean accuracy: 0.9666666666666666\n",
      "Standard deviation of accuracy: 0.1\n",
      "Confusion Matrix:\n",
      "[[17  0]\n",
      " [ 1 16]]\n",
      "Accuracy: 0.9705882352941176\n",
      "Precision: 1.0\n",
      "Recall: 0.9411764705882353\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.9696969696969697 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import svm\n",
    "from sklearn.metrics import confusion_matrix, roc_curve, auc\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "datasets = [bicluster1_data, bicluster2_data, bicluster3_data, bicluster4_data]\n",
    "labels = ['Dataset Based Bicluster 1', 'Dataset Based Bicluster 2', 'Dataset Based Bicluster 3', 'Dataset Based Bicluster 4']\n",
    "\n",
    "for i, data in enumerate(datasets):\n",
    "    X = data.iloc[:, :-1]\n",
    "    y = data.iloc[:, -1]\n",
    "    \n",
    "    # Split dataset into training set and test set\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, stratify=y) \n",
    "    \n",
    "    # Create a SVM Classifier\n",
    "    clf = svm.SVC(kernel='linear', probability=True)\n",
    "    clf.fit(X_train, y_train)\n",
    "\n",
    "    # Obtain Labels\n",
    "    print(labels[i], \":\\n\")\n",
    "\n",
    "    # Confusion Matrix and Cross Validation\n",
    "    # Perform 10 Fold Validation\n",
    "    k_folds = KFold(n_splits = 10)\n",
    "    # Calculate the Cross Validation Score\n",
    "    scores = cross_val_score(clf, X, y, cv = k_folds)\n",
    "    # Store the scores\n",
    "    cross_validation.append(scores)\n",
    "    # Perform cross-validation and obtain predicted labels for each fold\n",
    "    predicted_labels = cross_val_predict(clf, X, y, cv=k_folds)\n",
    "\n",
    "    # Compute the confusion matrix\n",
    "    conf_matrix = confusion_matrix(y, predicted_labels)\n",
    "    # Print the cross-validation scores\n",
    "    print(\"Cross-validation scores:\", scores)\n",
    "    # Print the mean and standard deviation of the scores\n",
    "    print(\"Mean accuracy:\", scores.mean())\n",
    "    print(\"Standard deviation of accuracy:\", scores.std())\n",
    "    # Print the confusion matrix\n",
    "    print(\"Confusion Matrix:\")\n",
    "    print(conf_matrix)\n",
    "\n",
    "    # Extract true positives (TP), true negatives (TN), false positives (FP), and false negatives (FN)\n",
    "    TP = conf_matrix[1, 1]\n",
    "    TN = conf_matrix[0, 0]\n",
    "    FP = conf_matrix[0, 1]\n",
    "    FN = conf_matrix[1, 0]\n",
    "\n",
    "    # Calculate accuracy\n",
    "    accuracy = (TP + TN) / (TP + TN + FP + FN)\n",
    "    # Calculate precision\n",
    "    precision = TP / (TP + FP)\n",
    "    # Calculate recall\n",
    "    recall = TP / (TP + FN)\n",
    "    # Calculate specificity\n",
    "    specificity = TN / (TN + FP)\n",
    "    # Calculate F1 score\n",
    "    f1_score = 2 * (precision * recall) / (precision + recall)\n",
    "\n",
    "    print(\"Accuracy:\", accuracy)\n",
    "    print(\"Precision:\", precision)\n",
    "    print(\"Recall:\", recall)\n",
    "    print(\"Specificity:\", specificity)\n",
    "    print(\"F1 Score:\", f1_score, \"\\n\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset:  Dataset Based Bicluster 1\n",
      "Random State: 5, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 10, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 15, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 20, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 25, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 30, Accuracy: 0.8571428571428571\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [2 5]]\n",
      "Accuracy: 0.8571428571428571\n",
      "Precision: 1.0\n",
      "Recall: 0.7142857142857143\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.8333333333333333 \n",
      "\n",
      "\n",
      "Random State: 35, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 40, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 45, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 50, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 55, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Dataset:  Dataset Based Bicluster 2\n",
      "Random State: 5, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 10, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 15, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 20, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 25, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 30, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 35, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 40, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 45, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 50, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 55, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Dataset:  Dataset Based Bicluster 3\n",
      "Random State: 5, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 10, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 15, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 20, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[6 1]\n",
      " [0 7]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 0.875\n",
      "Recall: 1.0\n",
      "Specificity: 0.8571428571428571\n",
      "F1 Score: 0.9333333333333333 \n",
      "\n",
      "\n",
      "Random State: 25, Accuracy: 0.8571428571428571\n",
      "Confusion Matrix:\n",
      "[[6 1]\n",
      " [1 6]]\n",
      "Accuracy: 0.8571428571428571\n",
      "Precision: 0.8571428571428571\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 0.8571428571428571\n",
      "F1 Score: 0.8571428571428571 \n",
      "\n",
      "\n",
      "Random State: 30, Accuracy: 0.8571428571428571\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [2 5]]\n",
      "Accuracy: 0.8571428571428571\n",
      "Precision: 1.0\n",
      "Recall: 0.7142857142857143\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.8333333333333333 \n",
      "\n",
      "\n",
      "Random State: 35, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 40, Accuracy: 0.8571428571428571\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [2 5]]\n",
      "Accuracy: 0.8571428571428571\n",
      "Precision: 1.0\n",
      "Recall: 0.7142857142857143\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.8333333333333333 \n",
      "\n",
      "\n",
      "Random State: 45, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 50, Accuracy: 0.8571428571428571\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [2 5]]\n",
      "Accuracy: 0.8571428571428571\n",
      "Precision: 1.0\n",
      "Recall: 0.7142857142857143\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.8333333333333333 \n",
      "\n",
      "\n",
      "Random State: 55, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[6 1]\n",
      " [0 7]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 0.875\n",
      "Recall: 1.0\n",
      "Specificity: 0.8571428571428571\n",
      "F1 Score: 0.9333333333333333 \n",
      "\n",
      "\n",
      "Dataset:  Dataset Based Bicluster 4\n",
      "Random State: 5, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 10, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 15, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 20, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 25, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 30, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 35, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n",
      "Random State: 40, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 45, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 50, Accuracy: 0.9285714285714286\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [1 6]]\n",
      "Accuracy: 0.9285714285714286\n",
      "Precision: 1.0\n",
      "Recall: 0.8571428571428571\n",
      "Specificity: 1.0\n",
      "F1 Score: 0.923076923076923 \n",
      "\n",
      "\n",
      "Random State: 55, Accuracy: 1.0\n",
      "Confusion Matrix:\n",
      "[[7 0]\n",
      " [0 7]]\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "Specificity: 1.0\n",
      "F1 Score: 1.0 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import svm\n",
    "from sklearn.metrics import confusion_matrix, roc_curve, auc\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "datasets = [bicluster1_data, bicluster2_data, bicluster3_data, bicluster4_data]\n",
    "labels = ['Dataset Based Bicluster 1', 'Dataset Based Bicluster 2', 'Dataset Based Bicluster 3', 'Dataset Based Bicluster 4']\n",
    "\n",
    "# Initialize dictionaries to store FPR, TPR, AUC values, and accuracy scores for each dataset\n",
    "results = {label: {'fprs': [], 'tprs': [], 'aucs': [], 'scores': []} for label in labels}\n",
    "random_states = [5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55]\n",
    "\n",
    "for i, data in enumerate(datasets):\n",
    "    X = data.iloc[:, :-1]\n",
    "    y = data.iloc[:, -1]\n",
    "    label = labels[i]\n",
    "\n",
    "    print(\"Dataset: \", label)\n",
    "    \n",
    "    for rs in random_states:\n",
    "        # Split dataset into training set and test set\n",
    "        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=rs, stratify=y) \n",
    "\n",
    "        # Create a SVM Classifier\n",
    "        clf = svm.SVC(kernel='linear', probability=True)\n",
    "        clf.fit(X_train, y_train)\n",
    "        y_pred = clf.predict(X_test)\n",
    "        accuracy = np.mean(y_pred == y_test)\n",
    "        results[label]['scores'].append(accuracy)\n",
    "        print(f\"Random State: {rs}, Accuracy: {accuracy}\")     \n",
    "\n",
    "        # ROC Curve\n",
    "        # Get predicted probabilities\n",
    "        y_pred_proba = clf.predict_proba(X_test)[:, 1]\n",
    "\n",
    "        # Calculate ROC curve\n",
    "        fpr, tpr, _ = roc_curve(y_test, y_pred_proba)\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "\n",
    "        # Store the results\n",
    "        results[label]['fprs'].append(fpr)\n",
    "        results[label]['tprs'].append(tpr)\n",
    "        results[label]['aucs'].append(roc_auc)\n",
    "\n",
    "        # Confusion Matrix\n",
    "        conf_matrix = confusion_matrix(y_test, y_pred)\n",
    "        print(\"Confusion Matrix:\")\n",
    "        print(conf_matrix)\n",
    "\n",
    "        # Extract true positives (TP), true negatives (TN), false positives (FP), and false negatives (FN)\n",
    "        TP = conf_matrix[1, 1]\n",
    "        TN = conf_matrix[0, 0]\n",
    "        FP = conf_matrix[0, 1]\n",
    "        FN = conf_matrix[1, 0]\n",
    "\n",
    "        # Calculate accuracy\n",
    "        accuracy = (TP + TN) / (TP + TN + FP + FN)\n",
    "        # Calculate precision\n",
    "        precision = TP / (TP + FP)\n",
    "        # Calculate recall\n",
    "        recall = TP / (TP + FN)\n",
    "        # Calculate specificity\n",
    "        specificity = TN / (TN + FP)\n",
    "        # Calculate F1 score\n",
    "        f1_score = 2 * (precision * recall) / (precision + recall)\n",
    "\n",
    "        print(\"Accuracy:\", accuracy)\n",
    "        print(\"Precision:\", precision)\n",
    "        print(\"Recall:\", recall)\n",
    "        print(\"Specificity:\", specificity)\n",
    "        print(\"F1 Score:\", f1_score, \"\\n\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
